---
title: "TMRC3 `r Sys.getenv('VERSION')`: Creating Data Structures"
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: zenburn
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body .main-container {
  max-width: 1600px;
}
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
  font-size: 16px
}
</style>

```{r options, include = FALSE}
library(hpgltools)
library(dplyr)
library(forcats)
library(glue)
library(EuPathDB)

knitr::opts_knit$set(progress = TRUE, verbose = TRUE, width = 90, echo = TRUE)
knitr::opts_chunk$set(
  error = TRUE, fig.width = 8, fig.height = 8, fig.retina = 2,
  out.width = "100%", dev = "png",
  dev.args = list(png = list(type = "cairo-png")))
old_options <- options(digits = 4, stringsAsFactors = FALSE, knitr.duplicate.label = "allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size = 12))
ver <- Sys.getenv("VERSION")
previous_file <- ""
rundate <- format(Sys.Date(), format = "%Y%m%d")

##tmp <- try(sm(loadme(filename=gsub(pattern="\\.Rmd", replace="\\.rda\\.xz", x=previous_file))))
rmd_file <- "tmrc3_datasets.Rmd"
savefile <- gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = rmd_file)
data_structures <- c()
```

# Introduction

This document takes the various outputs produced by
salmon/hisat/htseq/etc and generates the data which will be used in
all of the following analyses.

## Metadata Sources

Periodically, the shared metadata sheet was downloaded locally to the
sample_sheets/ directory, given a suffix corresponding to the current
worksheet revision, and the various trimming/alignment/etc statistics
were manually added.  Finally, some demographics were
appended to the sample metadata.  The resulting modified metadata
was used as the input for all following analyses.

```{r}
samplesheet <- "sample_sheets/tmrc3_samples.xlsx"
testthat::expect_true(file.exists(samplesheet))
demographics <- "sample_sheets/tmrc3_demographics.xlsx"
hslp_samplesheet <- "sample_sheets/tmrc3_samples_pruned.xlsx"
species_identities <- "sample_sheets/identified_parasite_species.xlsx"
```

# Annotation Collection

The primary annotation sources are:

1.  Ensembl's biomart archive from 2020 for human annotations.
2.  The TriTrypDB release 36 for parasite annotations.

Both provide GO data.  They also provide helpful links to other data
sources.  For the moment, we are focusing on the human annotations.

## Gene annotations

These analyses have focused on gene-level abundances/differences.
Thus, when htseq-count was invoked against the hisat2-based mappings,
parameters were chosen to count genes rather than transcripts.  In this
context, a gene refers to the non-redundant union of the transcripts' exons.
Similarly, when salmon counts were used via tximport, a mapping of
genes to transcripts was used to collapse the matrix to gene-level
abundances.  This decision may be revisited.

The parasite annotations were downloaded from the tritrypdb via the
provided REST interface, exported into an OrgDB instance, and
extracted via the EuPathDB package.

```{r}
hs_annot <- load_biomart_annotations(year = "2020", month = "jan")
hs_annot <- hs_annot[["annotation"]]
## The next two lines make a bridge between the gff file used by hisat2
## and the gene IDs.
hs_annot[["transcript"]] <- paste0(rownames(hs_annot), ".", hs_annot[["version"]])
hs_annot <- group_mean_cds_length(hs_annot)
hs_tx_annot <- hs_annot ## Make a copy so I don't lose transcript IDs for salmon
rownames(hs_annot) <- make.names(hs_annot[["ensembl_gene_id"]], unique = TRUE)
not_unique_idx <- grepl(x = rownames(hs_annot), pattern = "\\.\\d+$")
hs_annot <- hs_annot[!not_unique_idx, ]
```

### Also collect parasite annotations

Given that I recently included the parasite transcripts to the data, I
should also spend a moment and figure out how to include the
annotations.

As a starting point, I will just grab the EuPathDB panamensis
annotations.  I think I will need to limit them to just the shared
columns and drop the rest.

```{r, eval=FALSE}
orgdb <- "org.Lpanamensis.MHOMCOL81L13.v46.eg.db"
if (! orgdb %in% installed.packages()) {
  entries <- EuPathDB::download_eupath_metadata(webservice = "tritrypdb")
  entry <- EuPathDB::get_eupath_entry(species = "MHOMCOL", metadata = entries)
  orgdb_names <- EuPathDB::make_eupath_orgdb(entry, install = TRUE)
  orgdb <- orgdb_names[["pkgname"]]
}

tt <- sm(library(orgdb, character.only = TRUE))
pan_db <- org.Lpanamensis.MHOMCOL81L13.v46.eg.db
all_fields <- columns(pan_db)
all_lp_annot <- load_orgdb_annotations(
    pan_db,
    keytype = "gid",
    fields = c("annot_gene_entrez_id", "annot_gene_name",
               "annot_strand", "annot_gene_location_text", "annot_chromosome",
               "annot_cds_length", "annot_gene_type",
               "annot_gene_product"))$genes
all_lp_annot[["end"]] <- ""
all_lp_annot[["mean_cds_len"]] <- all_lp_annot[["annot_cds_length"]]
all_lp_annot[["version"]] <- 1
all_lp_annot[["transcript_version"]] <- 1
all_lp_annot[["hgnc_symbol"]] <- ""
all_lp_annot[["transcript"]] <- all_lp_annot[["gid"]]
all_lp_annot[["ensembl_transcript_id"]] <- all_lp_annot[["gid"]]

wanted_columns <- c("gid", "ensembl_transcript_id", "version",
                    "transcript_version", "hgnc_symbol", "annot_gene_product",
                    "annot_gene_type", "annot_cds_length", "annot_chromosome", "annot_strand",
                    "annot_gene_location_text", "end", "transcript", "mean_cds_len")
all_lp_annot <- all_lp_annot[, wanted_columns]
colnames(all_lp_annot) <- colnames(hs_tx_annot)

hslp_annot <- rbind(hs_tx_annot, all_lp_annot)
```

### Set up gene->transcript mappings

One should pay careful attention to the columns used to concatenate
the gene/version and transcript/version.  If these do not match
exactly, many genes will be dropped by tximport without any output to
signify something is wrong.

```{r}
## The tx_gene_map is used for tximport and salmon quantifications.
hs_annot[["transcript"]] <- paste0(rownames(hs_annot), ".",
                                   hs_annot[["transcript_version"]])
rownames(hs_annot) <- make.names(hs_annot[["ensembl_gene_id"]], unique = TRUE)
hs_tx_gene_map <- hs_annot[, c("transcript", "ensembl_gene_id")]
```

## Gene ontology data

The set of GO categories has not been limited to the 2020 data at the
time of this writing.  The GO categories is collected along with
lengths for goseq.  The other methods either have built-in databases
of human data (gProfiler) or support orgDB data (org.Hs.eg.db)
(clusterProfiler/topGO/gostats).

```{r, eval=FALSE}
hs_go <- try(load_biomart_go()[["go"]])
if ("try-error" %in% class(hs_go)) {
  hs_go <- NULL
}

hs_length <- hs_annot[, c("ensembl_gene_id", "cds_length")]
colnames(hs_length) <- c("ID", "length")
data_structures <- c(data_structures, "hs_length", "hs_go", "hs_annot")
```

# Dataset: Create the parent data structure of all samples

Before we do any of the following subsets/analyses of the data, we
need to collect it all in one place.  Let's do that here and name it
'hs_expt', it will comprise the full set of human data.  When we cull
some samples it will be renamed to 'hs_valid'.

The variable 'data_structures' will contain the names of every
variable I create in this document which I wish to save to disk.

Similarly, the character vector 'sanitize_columns' is the list of
metadata columns over which I will iterate and sanitize the data,
removing any spurious spaces, capitalization, etc.

## create_expt

The primary function used is 'create_expt()', which is responsible for
combining the metadata (from the first block of code), the gene
annotations (from the 2nd/3rd), and count tables (included in the
metadata as a column of filenames) into a single expressionset.  The
sanitize_columns variable defines the metadata columns which will be
explicitly sanitized beyond the general checks
performed. 'subset_genes()' is responsible for including _only_ the
protein coding genes in the data; it records the amount of information
lost to the metadata as column 'ncrna_lost'. 'sanitize_metadata()'
will perform some sanity checks of the metadata to ensure that excel
did not do anything untoward to the metadata. The 'set_expt_*()'
functions modify the metadata to make it more friendly for future
tasks, including setting some columns to factors, defining the column
of interest vs. known surrogate(s).  Finally, 'subset_expt()' is used
to drop samples which are known to not be of interest, like the few
samples from people who presented as healthy in the clinic (when the
clinicalpresentation column of the metadata is 'h').

```{r}
sanitize_columns <- c("visitnumber", "finaloutcome", "donor",
                      "typeofcells", "clinicalpresentation", "drug",
                      "condition", "batch")
hs_expt <- create_expt(samplesheet,
                       file_column = "hg38100hisatfile",
                       savefile = glue("rda/tmrc3_hs_expt_all_raw-v{ver}.rda"),
                       gene_info = hs_annot) %>%
  subset_genes(column = "gene_biotype", method = "keep",
               pattern = "protein_coding", meta_column = "ncrna_lost") %>%
  sanitize_expt_pData(columns = sanitize_columns)  %>%
  set_expt_factors(columns = sanitize_columns, class = "factor") %>%
  set_expt_conditions(fact = "finaloutcome") %>%
  set_expt_batches(fact = "visitnumber") %>%
  subset_expt(subset = "typeofcells!='pbmcs'") %>%
  subset_expt(subset = "typeofcells!='macrophages'") %>%
  subset_expt(subset = "clinicalpresentation!='h'")
data_structures <- c(data_structures, "hs_expt")

## The following should make visit 1 the largest if one uses that column as a size factor when plotting.
meta <- pData(hs_expt) %>%
  mutate(visitnumber = fct_relevel(visitnumber, c("3", "2", "1")))
start_order <- rownames(meta)
## Incorporate the identified strains
identified <- openxlsx::read.xlsx(species_identities)
no_data_idx <- identified[["Species"]] == "No hay dato"
identified <- identified[!no_data_idx, ]
identified <- identified[, c("Patient", "Species")]
colnames(identified) <- c("Patient", "ParasiteSpecies")
meta <- merge(meta, identified, by.x = "tubelabelorigin", by.y = "Patient", all.x = TRUE)
meta <- sanitize_metadata(meta, columns = "ParasiteSpecies", spaces = TRUE)
## Note that the merge removes the rownames and maybe changes sample order.
rownames(meta) <- meta[["tmrcidentifier"]]
meta <- meta[start_order, ]

pData(hs_expt) <- meta
save(list = "hs_expt", file = glue("rda/tmrc3_hs_expt_all_sanitized-v{ver}.rda"))
```

## Add the patient demographics

```{r}
hs_pd <- pData(hs_expt)
hs_demographics <- openxlsx::read.xlsx(demographics)
hs_demographics[["join"]] <- tolower(hs_demographics[["codigo_paciente"]])
hs_pd[["join"]] <- hs_pd[["tubelabelorigin"]]
hs_pd_demographics <- plyr::join(hs_pd, hs_demographics, by = "join")
hs_pd_demographics[["join"]] <- NULL
rownames(hs_pd_demographics) <- rownames(hs_pd)
na_idx <- is.na(hs_pd_demographics)
sum(na_idx)
hs_pd_demographics[na_idx] <- "undefined"
pData(hs_expt) <- hs_pd_demographics

## Now add in the ethnicity and sex
## Add a combination of the ethnicity and clinic
## While I am at it, do sex+clinic
table(pData(hs_expt)$eb_lc_sexo)
table(pData(hs_expt)$eb_lc_etnia)
table(pData(hs_expt)$clinic)
etnia_names <- pData(hs_expt)[["eb_lc_etnia"]]
etnia_idx <- etnia_names == 1
etnia_names[etnia_idx] <- "afrocol"
etnia_idx <- etnia_names == 2
etnia_names[etnia_idx] <- "indigena"
etnia_idx <- etnia_names == 3
etnia_names[etnia_idx] <- "mestiza"
pData(hs_expt)[["etnia"]] <- as.factor(etnia_names)
table(pData(hs_expt)[["etnia"]])
etnia_factor <- paste0(pData(hs_expt)[["clinic"]], "_",
                       etnia_names)
pData(hs_expt)[["clinic_etnia"]] <- as.factor(etnia_factor)
table(pData(hs_expt)[["clinic_etnia"]])

pData(hs_expt)[["sex"]] <- "female"
male_idx <- pData(hs_expt)[["eb_lc_sexo"]] == 1
pData(hs_expt)[male_idx, "sex"] <- "male"
sex_factor <- paste0(pData(hs_expt)[["clinic"]], "_",
                     pData(hs_expt)[["sex"]])
pData(hs_expt)[["clinic_sex"]] <- as.factor(sex_factor)
table(pData(hs_expt)[["clinic_sex"]])

pData(hs_expt)[["etnia_sex"]] <- paste0(pData(hs_expt)[["etnia"]], "_",
                                        pData(hs_expt)[["sex"]])
pData(hs_expt)[["etnia_sex"]] <- as.factor(pData(hs_expt)[["etnia_sex"]])
table(pData(hs_expt)[["etnia_sex"]])

## Set a factor of samples which are visit 1 vs. 2/3.
v1_samples <- pData(hs_expt)[["visitnumber"]] == "1"
pData(hs_expt)[v1_samples, "visitbipart"] <- "v1"
pData(hs_expt)[!v1_samples, "visitbipart"] <- "vother"
table(pData(hs_expt)[["visitbipart"]])

save(list = "hs_expt", file = glue("rda/tmrc3_hs_expt_all_sanitized_demographics-v{ver}.rda"))
```

## Sanity check, clinical outcome vs. demographics

Check to make sure the demographics agrees with the metadata.  I _really_ want
this block to return NULL upon completion.

```{r}
two_columns <- pData(hs_expt)[, c("finaloutcome", "ef_lc_estado_final_estudio")]
undef_idx <- two_columns[[2]] == "undefined"
two_columns <- two_columns[!undef_idx, ]
two_columns[["rewritten"]] <- "undef"
cure_idx <- two_columns[[2]] == 0
two_columns[cure_idx, "rewritten"] <- "cure"
fail_idx <- two_columns[[2]] == 1
two_columns[fail_idx, "rewritten"] <- "failure"
lost_idx <- two_columns[[2]] == 2
two_columns[lost_idx, "rewritten"] <- "lost"
same_idx <- two_columns[["finaloutcome"]] == two_columns[["rewritten"]]
broken <- two_columns[!same_idx, ]
broken
```

## Summarize: Collect sample numbers before filtering

There are some metadata factors for which I think it will be nice to
see the numbers before and after our filters.  The following shows how
many samples we have of the primary types before filtering.

```{r}
dim(pData(hs_expt))
table(pData(hs_expt)$drug)
table(pData(hs_expt)$clinic)
table(pData(hs_expt)$finaloutcome)
table(pData(hs_expt)$typeofcells)
table(pData(hs_expt)$visitnumber)
summary(as.numeric(pData(hs_expt)$eb_lc_tiempo_evolucion))
summary(as.numeric(pData(hs_expt)$eb_lc_tto_mcto_glucan_dosis))
summary(as.numeric(pData(hs_expt)$v3_lc_ejey_lesion_mm_1))
summary(as.numeric(pData(hs_expt)$v3_lc_lesion_area_1))
summary(as.numeric(pData(hs_expt)$v3_lc_ejex_ulcera_mm_1))
table(pData(hs_expt)$eb_lc_sexo)
table(pData(hs_expt)$eb_lc_etnia)
summary(as.numeric(pData(hs_expt)$edad))
table(pData(hs_expt)$eb_lc_peso)
table(pData(hs_expt)$eb_lc_estatura)
table(pData(hs_expt)$ef_lc_estado_final_estudio)
table(pData(hs_expt)$finaloutcome)

unique(pData(hs_expt)$codigo_paciente)
length(unique(pData(hs_expt)$codigo_paciente))
```

Get the sex of the unique patients.

```{r}
unique_people <- unique(pData(hs_expt)[["codigo_paciente"]])
unique_people
length(unique_people)

df <- pData(hs_expt)
people <- df[["codigo_paciente"]]
first_indices <- order(people)[!duplicated(sort(people))]
table(df[first_indices, "eb_lc_sexo"])
table(df[first_indices, "finaloutcome"])
```

## Define desired colors for the various subsets

There are lots of ways which we will categorize the data, here are
some potential color choices for them.

```{r}
color_choices <- list(
    "cf" = list(
      "cure" = "#998EC3",
      "failure" = "#F1A340"),
    "cflost" = list(
      "cure" = "#998EC3",
      "failure" = "#F1A340",
      "lost" = "#343434"),
    "type_visit" = list(
      "monocytes_v1" = "#DD1C77",
      "monocytes_v2" = "#C994C7",
      "monocytes_v3" = "#E7E1EF",
      "eosinophils_v1" = "#31A354",
      "eosinophils_v2" = "#ADDD8E",
      "eosinophils_v3" = "#F7FCD9",
      "neutrophils_v1" = "#3182BD",
      "neutrophils_v2" = "#9ECAE1",
      "neutrophils_v3" = "#DEEBF7",
      "biopsy_v1" = "#D95F0E"),
    "type" = list(
      "monocytes" = "#DD1C77",
      "eosinophils" = "#31A354",
      "neutrophils" = "#3182BD",
      "biopsy" = "#D95F0E"),
    "two_visit" = list(
      "first" = "#33EE33",
      "later" = "#023302"),
    "visit" = list(
      "v1" = "#33EE33",
      "v2" = "#11AA11",
      "v3" = "#023302"),
    "visit2" = list(
      "1" = "#33EE33",
      "2" = "#11AA11",
      "3" = "#023302"),
    "visitbi" = list(
      "v1" = "#BB0000",
      "vother" = "#0000BB"),
    "labs" = list(
      "Brazil" = "#FFC300",
      "Colombia" = "#525252"),
    "clinic" = list(
      "Tumaco" = "#3182AA",
      "Cali" = "#C994AA"),
    "clinic_cf" = list(
      "Tumaco_cure" = "#7670B3",
      "Tumaco_failure" = "#E7298A",
      "Cali_cure" = "#1B9E77",
      "Cali_failure" = "#D95F02"),
    "ethnicity" = list(
      "afrocol" = "#4293CE",
      "indigena" = "#BDBDBD",
      "mestiza" = "#FEB24C"),
    "clinic_etnia" = list(
      "Cali_afrocol" = "#3182BD",
      "Cali_indigena" = "#636363",
      "Cali_mestiza" = "#F03B20",
      "Tumaco_afrocol" = "#9ECAE1",
      "Tumaco_indigena" = "#BDBDBD",
      "Tumaco_mestiza" = "#FEB24C"),
    "sex" = list(
      "female" = "#3182BD",
      "male" = "#C994C7"),
    "clinic_sex" = list(
      "Cali_female" = "#0000CC",
      "Tumaco_female" = "#3182BD",
      "Cali_male" = "#E7298A",
      "Tumaco_male" = "#C994C7"),
    "cf_type" = list(
      "cure_biopsy" = "#D95F0E",
      "failure_biopsy" = "#FEC44F",
      "cure_monocytes" = "#DD1C77",
      "failure_monocytes" = "#C994C7",
      "cure_eosinophils" = "#31A354",
      "failure_eosinophils" = "#ADDD8E",
      "cure_neutrophils" = "#3182BD",
      "failure_neutrophils" = "#9ECAE1"),
    "parasite" = list(
      "notapplicable" = "#000000",
      "lvbraziliensis" = "#AA0000",
      "lvguyanensis" = "#660066",
      "lvpanamensis" = "#3160AC"))
data_structures <- c(data_structures, "color_choices")
```

# Define the starting data

Given the starting point above, we will start extracting groups of
samples of interest.

## Set our initial coverage goal

The first set of samples removed from the data are those with too many
missing genes.

## Figure S2: Non-zero genes before sample filtering

```{r}
all_nz <- plot_nonzero(hs_expt)
pp(file = "figures/S2_nonzero_all_samples.svg")
all_nz$plot
dev.off()
all_nz
```

### Save CPM, all samples

We should save a cpm and rpkm copy of the raw data for every subset in
order to make it easier for folks to work with WGCNA and other tools
(STRING).

```{r}
cpm_data <- normalize_expt(hs_expt, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/all_samples-v{ver}.xlsx"))

rpkm_data <- normalize_expt(hs_expt, filter = TRUE, convert = "rpkm",
                            column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/hs_expt_rpkm-v{ver}.csv"))
```

## Subset: Filter out problematic samples

To my eyes, there are 3 or 4 samples which are likely candidates for
removal.  In addition, we will remove samples which were lost during
the treatment and/or ones which were used in other experiments but
included in the TMRC3 sample sheet (thus the 'notapplicable' or
'null').

I think this is stated elsewhere: variables prefixed with 'tc' are
Tumaco and Cali; 't' or 'c' correspond.

```{r}
tc_nolost <- subset_expt(hs_expt, subset = "finaloutcome!='lost'")
tc_nomilt <- subset_expt(tc_nolost, subset = "drug=='antimony'")
plot_legend(tc_nomilt)
nomilt_nonzero <- plot_nonzero(tc_nomilt, plot_labels = TRUE)
nomilt_nonzero
pp(file = "figures/figs2_nonzero.svg")
nomilt_nonzero
dev.off()
tc_valid <- subset_expt(tc_nomilt, nonzero = 11000) %>%
  set_expt_colors(colors = color_choices[["cf"]])

save(list = "tc_valid", file = glue("rda/tmrc3_tc_valid-v{ver}.rda"))
data_structures <- c(data_structures, "tc_valid")
```

# Visualize the sample breakdown

```{r}
clinic_type_outcome_sankey <- plot_meta_sankey(
  tc_valid, factors = c("clinic", "typeofcells", "finaloutcome"),
  drill_down = TRUE, color_choices = color_choices)
clinic_type_outcome_sankey

clinic_ethnicity_outcome_sankey <- plot_meta_sankey(
  tc_valid, factors = c("clinic", "etnia", "finaloutcome"),
  drill_down = TRUE, color_choices = color_choices)
clinic_ethnicity_outcome_sankey

clinic_sex_outcome_sankey <- plot_meta_sankey(
  tc_valid, factors = c("clinic", "sex", "finaloutcome"),
  drill_down = TRUE, color_choices = color_choices)
clinic_sex_outcome_sankey
```

## Figure XX + 1: Non-zero genes after sample filtering

The following plot is essentially identical to the previous with two
exceptions:

1.  The samples with too few genes (11,000 currently) are gone.
2.  The samples are colored by cure(purple)/fail(yellow)

```{r}
nz_post <- plot_nonzero(tc_valid)
nz_post
pp(file = "figures/figS2v2_post.svg")
nz_post
dev.off()
```

## Summarize: Tally samples after filtering

We need to keep track of how many of each sample type is lost when we
do our various filters.  Thus I am repeating the same set of tallies.
This will likely happen one more time, following the removal of
samples which came from Cali.

```{r}
table(pData(tc_valid)[["drug"]])
table(pData(tc_valid)[["clinic"]])
table(pData(tc_valid)[["finaloutcome"]])
table(pData(tc_valid)[["typeofcells"]])
table(pData(tc_valid)[["visit"]])
summary(as.numeric(pData(tc_valid)[["eb_lc_tiempo_evolucion"]]))
summary(as.numeric(pData(tc_valid)[["eb_lc_tto_mcto_glucan_dosis"]]))
summary(as.numeric(pData(tc_valid)[["v3_lc_ejey_lesion_mm_1"]]))
summary(as.numeric(pData(tc_valid)[["v3_lc_lesion_area_1"]]))
summary(as.numeric(pData(tc_valid)[["v3_lc_ejex_ulcera_mm_1"]]))
table(pData(tc_valid)[["eb_lc_sexo"]])
table(pData(tc_valid)[["eb_lc_etnia"]])
summary(as.numeric(pData(tc_valid)[["edad"]]))
table(pData(tc_valid)[["eb_lc_peso"]])
table(pData(tc_valid)[["eb_lc_estatura"]])

length(unique(pData(tc_valid)[["codigo_paciente"]]))
```

## Split the data by cure/fail for some counting

We now have two datasets which are comprised of only the
cure/fail samples.  It is worth considering if there are ways we
can leverage this separation.

```{r}
tc_cure <- subset_expt(tc_valid, subset = "finaloutcome=='cure'")
data_structures <- c(data_structures, "tc_cure")

table(pData(tc_cure)[["visitnumber"]])

tc_fail <- subset_expt(tc_valid, subset = "finaloutcome=='failure'")
data_structures <- c(data_structures, "tc_fail")

table(pData(tc_fail)[["visitnumber"]])
```

# Print out cell type and visit numbers

```{r}
all_types <- table(pData(tc_valid)[["typeofcells"]])
all_types
all_times <- table(pData(tc_valid)[["visitnumber"]])
all_times
```

### Write valid samples

We completed the initial filter of the data.  Write it out.  This
data should include all samples except those we explicitly removed due
to poor coverage, lost-status, or non-canonical cell type (the pilot
study).

```{r}
cpm_data <- normalize_expt(tc_valid, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/all_valid_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tc_valid, filter = TRUE, convert = "rpkm",
                            column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tc_all_valid_samples-v{ver}.csv"))
```

## Extract cell types

Extract the samples of each cell type, these will be the basis for a
significant number of the clinical outcome comparisons.

```{r}
## Note, I will save these data structures in a little bit when I
## decide how I want to set the conditions, batches, and colors.
tc_eosinophils <- subset_expt(tc_valid, subset = "typeofcells=='eosinophils'")
tc_monocytes <- subset_expt(tc_valid, subset = "typeofcells=='monocytes'")
tc_neutrophils <- subset_expt(tc_valid, subset = "typeofcells=='neutrophils'")
```

### Write cell type cpm

The following writes out rpkm/cpm values for each celltype from both
clinics(tc prefix).

```{r}
cpm_data <- normalize_expt(tc_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/tc_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tc_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tc_eosinophils-v{ver}.csv"))

cpm_data <- normalize_expt(tc_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/tc_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tc_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tc_monocytes-v{ver}.csv"))

cpm_data <- normalize_expt(tc_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/tc_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tc_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tc_neutrophils-v{ver}.csv"))
```

## Tumaco and Cali samples by visit

The prefixes of the following datastructures state that these are from
both clinics (tc) and a specific visit (v1/v2/v3).  The rpkm/cpm
values will be written immediately thereafter.

```{r}
tcv1_samples <- subset_expt(tc_valid, subset = "visitnumber=='1'")
save(list = "tcv1_samples", file = glue("rda/tmrc3_tcv1_samples-v{ver}.rda"))
data_structures <- c(data_structures, "tcv1_samples")

tcv2_samples <- subset_expt(tc_valid, subset = "visitnumber=='2'")
save(list = "tcv2_samples", file = glue("rda/tmrc3_tcv2_samples-v{ver}.rda"))
data_structures <- c(data_structures, "tcv2_samples")

tcv3_samples <- subset_expt(tc_valid, subset = "visitnumber=='3'")
save(list = "tcv3_samples", file = glue("rda/tmrc3_tcv3_samples-v{ver}.rda"))
data_structures <- c(data_structures, "tcv3_samples")
```

### Cali and Tumaco by visit, write cpm

```{r}
cpm_data <- normalize_expt(tcv1_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/tcv1_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tcv1_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tcv1_samples-v{ver}.csv"))

cpm_data <- normalize_expt(tcv2_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/tcv2_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tcv2_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tcv2_samples-v{ver}.csv"))

cpm_data <- normalize_expt(tcv3_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/3_cali_and_tumaco/tcv3_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tcv3_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tcv3_samples-v{ver}.csv"))
```

# Tumaco and Cali data structures

All data structures which start with the prefix 'tc' are Tumaco and
Cali.  Those with 't' are only Tumaco, 'c' are only Cali.

```{r}
tc_clinical <- tc_valid %>%
  set_expt_conditions(fact = "finaloutcome", colors = color_choices[["cf"]]) %>%
  set_expt_batches(fact = "typeofcells")
save(list = "tc_clinical", file = glue("rda/tmrc3_tc_clinical-v{ver}.rda"))
data_structures <- c(data_structures, "tc_clinical")
```

## Subset: Monocytes by clinic

For some of the following data structures, we will be concatenating
various metadata factors of interest, usually the final outcome and
one other metadatum.

## Subset Monocytes

In the following block I am writing out the Monocytes in a couple
different ways, once with the clinical outcome as the primary factor,
and again using the concatenation of clinic/outcome.

```{r}
clinic_cf <- paste0(pData(tc_monocytes)[["clinic"]], "_",
                    pData(tc_monocytes)[["finaloutcome"]])
table(clinic_cf)
tc_monocytes <- set_expt_conditions(
  tc_monocytes, fact = clinic_cf, colors = color_choices[["clinic_cf"]]) %>%
  set_expt_batches(fact = "visitnumber")
save(list = "tc_monocytes", file = glue("rda/tmrc3_tc_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tc_monocytes")

tcv1_monocytes <- subset_expt(tc_monocytes, subset = "visitnumber=='1'")
save(list = "tcv1_monocytes", file = glue("rda/tmrc3_tcv1_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tcv1_monocytes")
tcv2_monocytes <- subset_expt(tc_monocytes, subset = "visitnumber=='2'")
save(list = "tcv2_monocytes", file = glue("rda/tmrc3_tcv2_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tcv2_monocytes")
tcv3_monocytes <- subset_expt(tc_monocytes, subset = "visitnumber=='3'")
save(list = "tcv3_monocytes", file = glue("rda/tmrc3_tcv3_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tcv3_monocytes")
```

## Subset: Biopsies

Followed by the Biopsy samples...

```{r}
tc_clinical_nobiop <- subset_expt(tc_clinical, subset = "typeofcells!='biopsy'") %>%
  set_expt_colors(color_choices[["cf"]])
save(list = "tc_clinical_nobiop", file = glue("rda/tmrc3_tc_clinical_nobiop-v{ver}.rda"))
data_structures <- c(data_structures, "tc_clinical_nobiop")

tc_biopsies <- tc_valid %>%
  set_expt_conditions(fact = "clinic") %>%
  set_expt_batches(fact = "finaloutcome") %>%
  subset_expt(subset = "typeofcells=='biopsy'")
tc_cf <- paste0(pData(tc_biopsies)[["clinic"]], "_",
                pData(tc_biopsies)[["finaloutcome"]])
table(tc_cf)
tc_biopsies <- set_expt_conditions(
  tc_biopsies, fact = tc_cf, colors = color_choices[["clinic_cf"]]) %>%
  set_expt_batches(fact = "visitnumber")
save(list = "tc_biopsies", file = glue("rda/tmrc3_tc_biopsies-v{ver}.rda"))
data_structures <- c(data_structures, "tc_biopsies")
```

## Subset: Eosinophils

... and the Eosinophils.  These are noteworthy because they have fewer
fails than some other cohorts.

```{r}
clinic_cf <- paste0(pData(tc_eosinophils)[["clinic"]], "_",
                    pData(tc_eosinophils)[["finaloutcome"]])
table(clinic_cf)
tc_eosinophils <- set_expt_conditions(
  tc_eosinophils, fact = clinic_cf, colors = color_choices[["clinic_cf"]]) %>%
  set_expt_batches(fact = "visitnumber")
save(list = "tc_eosinophils", file = glue("rda/tmrc3_tc_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tc_eosinophils")

tcv1_eosinophils <- subset_expt(tc_eosinophils, subset = "visitnumber=='1'")
save(list = "tcv1_eosinophils", file = glue("rda/tmrc3_tcv1_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tcv1_eosinophils")
tcv2_eosinophils <- subset_expt(tc_eosinophils, subset = "visitnumber=='2'")
save(list = "tcv2_eosinophils", file = glue("rda/tmrc3_tcv2_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tcv2_eosinophils")
tcv3_eosinophils <- subset_expt(tc_eosinophils, subset = "visitnumber=='3'")
save(list = "tcv3_eosinophils", file = glue("rda/tmrc3_tcv3_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tcv3_eosinophils")
```

## Subset: Neutrophils by clinic

Followed by the Neutrophils...

```{r}
clinic_cf <- paste0(pData(tc_neutrophils)[["clinic"]], "_",
                    pData(tc_neutrophils)[["finaloutcome"]])
table(clinic_cf)
tc_neutrophils <- set_expt_conditions(
  tc_neutrophils, fact = clinic_cf, colors = color_choices[["clinic_cf"]]) %>%
  set_expt_batches(fact = "visitnumber")
save(list = "tc_neutrophils", file = glue("rda/tmrc3_tc_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tc_neutrophils")

tcv1_neutrophils <- subset_expt(tc_neutrophils, subset = "visitnumber=='1'")
save(list = "tcv1_neutrophils", file = glue("rda/tmrc3_tcv1_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tcv1_neutrophils")
tcv2_neutrophils <- subset_expt(tc_neutrophils, subset = "visitnumber=='2'")
save(list = "tcv2_neutrophils", file = glue("rda/tmrc3_tcv2_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tcv2_neutrophils")
tcv3_neutrophils <- subset_expt(tc_neutrophils, subset = "visitnumber=='3'")
save(list = "tcv3_neutrophils", file = glue("rda/tmrc3_tcv3_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tcv3_neutrophils")
```

# Summarize: Tabulate sample numbers

Here is an outline of the samples in their current state:

** Note ** : Something changed ~ 20240510 which is causing this to
fail, commenting it out for the moment.

<! ---

* By type: `r ncol(exprs(tc_valid))`, Cure: `r sum(pData(tc_valid)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tc_valid)[["finaloutcome"]] == "failure")`
    *  Biopsy: `r all_types[["biopsy"]]`, Cure: `r sum(pData(tc_biopsies)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tc_biopsies)[["finaloutcome"]] == "failure")`
        * All biopsy samples are visit 1.
    *  Eosinophils: `r all_types[["eosinophils"]]`, Cure: `r sum(pData(tc_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tc_eosinophils)[["finaloutcome"]] == "failure")`
        * V1 Eosinophils: `r nrow(pData(tcv1_eosinophils))`, Cure: `r sum(pData(tcv1_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv1_eosinophils)[["finaloutcome"]] == "failure")`
        * V2 Eosinophils: `r nrow(pData(tcv2_eosinophils))`, Cure: `r sum(pData(tcv2_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv2_eosinophils)[["finaloutcome"]] == "failure")`
        * V3 Eosinophils: `r nrow(pData(tcv3_eosinophils))`, Cure: `r sum(pData(tcv3_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv3_eosinophils)[["finaloutcome"]] == "failure")`
    *  Monocytes: `r all_types[["monocytes"]]`, Cure: `r sum(pData(tc_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tc_monocytes)[["finaloutcome"]] == "failure")`
        * V1 Monocytes: `r nrow(pData(tcv1_monocytes))`, Cure: `r sum(pData(tcv1_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv1_monocytes)[["finaloutcome"]] == "failure")`
        * V2 Monocytes: `r nrow(pData(tcv2_monocytes))`, Cure: `r sum(pData(tcv2_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv2_monocytes)[["finaloutcome"]] == "failure")`
        * V3 Monocytes: `r nrow(pData(tcv3_monocytes))`, Cure: `r sum(pData(tcv3_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv3_monocytes)[["finaloutcome"]] == "failure")`
    *  Neutrophils: `r all_types[["neutrophils"]]`, Cure: `r sum(pData(tc_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tc_neutrophils)[["finaloutcome"]] == "failure")`
        * V1 Neutrophils: `r nrow(pData(tcv1_neutrophils))`, Cure: `r sum(pData(tcv1_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv1_neutrophils)[["finaloutcome"]] == "failure")`
        * V2 Neutrophils: `r nrow(pData(tcv2_neutrophils))`, Cure: `r sum(pData(tcv2_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv2_neutrophils)[["finaloutcome"]] == "failure")`
* V3 Neutrophils: `r nrow(pData(tcv3_neutrophils))`, Cure: `r sum(pData(tcv3_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tcv3_neutrophils)[["finaloutcome"]] == "failure")`

--->



## By sex

```{r}
tc_sex <- set_expt_conditions(tc_valid, fact = "sex", colors = color_choices[["sex"]])
save(list = "tc_sex", file = glue("rda/tmrc3_tc_sex-v{ver}.rda"))
data_structures <- c(data_structures, "tc_sex")
```

# Dataset: Only Tumaco samples

Our recent discussions have settled one big question regarding which
samples to use:  We will limit our analyses to only those samples from
Tumaco.

The following block will therefore set that group as our default for
future analyses.  We will essentially repeat all of the above sample
separations for the Tumaco-only cohort.

## All cell types

```{r}
t_clinical <- tc_clinical %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "t_clinical", file = glue("rda/tmrc3_t_clinical-v{ver}.rda"))
data_structures <- c(data_structures, "t_clinical")

cpm_data <- normalize_expt(t_clinical, convert = "cpm")
t_clinical_cpm_variance <- variance_expt(cpm_data)
## Add variance by gene for all samples excluding biopsies (I assume also excluding Cali).
## This is intended to address a query from Maria Adelaida on 20230111
```

### Write out the Tumaco Clinical with variance

```{r}
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_clinical-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_clinical, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
                 file = glue("rpkm/t_clinical-v{ver}.csv"))
cpm_data <- normalize_expt(t_clinical, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_clinical_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_clinical, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_clinical_sva-v{ver}.csv"))
```

# Perform write_expt on the Tumaco clinical samples

```{r}
written_expt <- write_expt(
    t_clinical,
    excel = glue("excel/t_clinical-v{ver}.xlsx"))
```

## Without biopsies

```{r}
t_clinical_nobiop <- subset_expt(t_clinical, subset = "typeofcells!='biopsy'")
save(list = "t_clinical_nobiop", file = glue("rda/tmrc3_t_clinical_nobiop-v{ver}.rda"))
data_structures <- c(data_structures, "t_clinical_nobiop")
cpm_data <- normalize_expt(t_clinical_nobiop, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_clinical_nobiop-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_clinical, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_clinical_sva-v{ver}.csv"))

cpm_data <- normalize_expt(t_clinical_nobiop, convert = "cpm",
                           filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_clinical_nobiop_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_clinical_nobiop, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_clinical_nobiop_sva-v{ver}.csv"))
```

### Pull out the top-n most variant genes

I added a few columns to the gene annotations reflecting the
variance/stdev/mean expression of each gene.  This really highlights
the extraordinary degree to which genes are changing in the data...

```{r}
top_expt <- normalize_expt(t_clinical_cpm_variance, filter = TRUE) %>%
  variance_expt()
top_fd <- fData(top_expt)
top_ex <- exprs(top_expt)
idx <- order(top_fd[["exprs_gene_stdev"]], decreasing = TRUE)
top_df <- top_fd[idx, ]
written <- write_xlsx(
    data = head(top_fd, n = 100),
    excel = glue("excel/t_clinical_cpm_stdev_top100-v{ver}.xlsx"))

bottom_expt <- normalize_expt(t_clinical_cpm_variance, filter = "simple", threshold = 500) %>%
  variance_expt()
bottom_fd <- fData(bottom_expt)
bottom_ex <- exprs(bottom_expt)
idx <- order(bottom_fd[["exprs_gene_stdev"]])
bottom <- bottom_fd[idx, ]
written <- write_xlsx(
    data = head(bottom_fd, n = 100),
    excel = glue("excel/t_clinical_cpm_stdev_bottom100-v{ver}.xlsx"))
```

## Summarize: Collect Tumaco sample numbers.

At least in theory, everything which follows will be using the above
'clinical' data structure.  Thus, let us count it up and get a sense
of what we will work with.

```{r}
table(pData(t_clinical)$drug)
table(pData(t_clinical)$clinic)
table(pData(t_clinical)$finaloutcome)
table(pData(t_clinical)$typeofcells)
table(pData(t_clinical)$visit)
summary(as.numeric(pData(t_clinical)$eb_lc_tiempo_evolucion))
summary(as.numeric(pData(t_clinical)$eb_lc_tto_mcto_glucan_dosis))
summary(as.numeric(pData(t_clinical)$v3_lc_ejey_lesion_mm_1))
summary(as.numeric(pData(t_clinical)$v3_lc_lesion_area_1))
summary(as.numeric(pData(t_clinical)$v3_lc_ejex_ulcera_mm_1))
table(pData(t_clinical)$eb_lc_sexo)
table(pData(t_clinical)$eb_lc_etnia)
summary(as.numeric(pData(t_clinical)$edad))
table(pData(t_clinical)$eb_lc_peso)
table(pData(t_clinical)$eb_lc_estatura)

length(unique(pData(t_clinical)[["codigo_paciente"]]))
only_cure <- pData(t_clinical)[["finaloutcome"]] == "cure"
c_meta <- pData(t_clinical)[only_cure, ]
length(unique(c_meta[["codigo_paciente"]]))
only_fail <- pData(t_clinical)[["finaloutcome"]] == "failure"
f_meta <- pData(t_clinical)[only_fail, ]
length(unique(f_meta[["codigo_paciente"]]))
```

## Split the data by cure/fail for some counting

```{r}
t_cure <- subset_expt(t_clinical, subset = "finaloutcome=='cure'")
table(pData(t_cure)$visitnumber)

t_fail <- subset_expt(t_clinical, subset = "finaloutcome=='failure'")
table(pData(t_fail)$visitnumber)
```

## Subset: Create Tumaco-only and cell-type specific data

I previously made a bunch of data subsets by visit, cell type, etc.
So let us overwrite them all with versions which contain only the
Tumaco samples.

### Tumaco biopsies

The is a copy of the cell type extractions above, except we lead off
with the removal of the Cali samples.

```{r}
t_biopsies <- tc_biopsies %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "t_biopsies", file = glue("rda/tmrc3_t_biopsies-v{ver}.rda"))
data_structures <- c(data_structures, "t_biopsies")
cpm_data <- normalize_expt(t_biopsies, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_biopsies-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_biopsies, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_biopsies-v{ver}.csv"))

cpm_data <- normalize_expt(t_biopsies, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_biopsies_sva-v{ver}.xlsx"))
```

### Tumaco Eosinophils

```{r}
t_eosinophils <- tc_eosinophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "t_eosinophils", file = glue("rda/tmrc3_t_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "t_eosinophils")
cpm_data <- normalize_expt(t_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_eosinophils-v{ver}.xlsx"))
cpm_data <- normalize_expt(t_eosinophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_eosinophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_eosinophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_eosinophils_sva-v{ver}.csv"))
```

### Tumaco subsets monocytes

```{r}
t_monocytes <- tc_monocytes %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "t_monocytes", file = glue("rda/tmrc3_t_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "t_monocytes")
cpm_data <- normalize_expt(t_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(t_monocytes, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_monocytes_sva-v{ver}.csv"))
```

### Tumaco subsets neutrophils

```{r}
t_neutrophils <- tc_neutrophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "t_neutrophils", file = glue("rda/tmrc3_t_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "t_neutrophils")
cpm_data <- normalize_expt(t_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
                 file = glue("rpkm/t_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(t_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(t_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/t_neutrophils_sva-v{ver}.csv"))
```

### Tumaco Visit 1 samples

```{r}
tv1_samples <- tcv1_samples %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv1_samples", file = glue("rda/tmrc3_tv1_samples-v{ver}.rda"))
data_structures <- c(data_structures, "tv1_samples")
cpm_data <- normalize_expt(tv1_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_samples-v{ver}.csv"))
cpm_data <- normalize_expt(tv1_samples, convert="cpm", filter=TRUE, batch="svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_samples_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_samples, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_samples_sva-v{ver}.csv"))
```

### Tumaco Visit 2 samples

```{r}
tv2_samples <- tcv2_samples %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv2_samples", file = glue("rda/tmrc3_tv2_samples-v{ver}.rda"))
data_structures <- c(data_structures, "tv2_samples")
cpm_data <- normalize_expt(tv2_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_samples-v{ver}.csv"))
cpm_data <- normalize_expt(tv2_samples, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_samples_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_samples, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_samples_sva-v{ver}.csv"))
```

### Tumaco visit 3 samples

```{r}
tv3_samples <- tcv3_samples %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv3_samples", file = glue("rda/tmrc3_tv3_samples-v{ver}.rda"))
data_structures <- c(data_structures, "tv3_samples")
cpm_data <- normalize_expt(tv3_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_samples-v{ver}.csv"))
cpm_data <- normalize_expt(tv3_samples, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_samples_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_samples, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_samples_sva-v{ver}.csv"))
```

### Tumaco visit 1 Eosinophils

```{r}
tv1_eosinophils <- tcv1_eosinophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv1_eosinophils", file = glue("rda/tmrc3_tv1_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tv1_eosinophils")
cpm_data <- normalize_expt(tv1_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_eosinophils-v{ver}.csv"))
cpm_data <- normalize_expt(tv1_eosinophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_eosinophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_eosinophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_eosinophils_sva-v{ver}.csv"))
```

### Tumaco visit 2 samples

```{r}
tv2_eosinophils <- tcv2_eosinophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv2_eosinophils", file = glue("rda/tmrc3_tv2_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tv2_eosinophils")
cpm_data <- normalize_expt(tv2_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_eosinophils-v{ver}.csv"))
cpm_data <- normalize_expt(tv2_eosinophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_eosinophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_eosinophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_eosinophils_sva-v{ver}.csv"))
```

### Tumaco visit 3 eosinophil

```{r}
tv3_eosinophils <- tcv3_eosinophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv3_eosinophils", file = glue("rda/tmrc3_tv3_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "tv3_eosinophils")
cpm_data <- normalize_expt(tv3_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_eosinophils-v{ver}.csv"))
cpm_data <- normalize_expt(tv3_eosinophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_eosinophils_sva-v{ver}.xlsx"))
## Note, the cbcb filter left behind a sufficient number of zeros that it confused cpm.
rpkm_data <- normalize_expt(tv3_eosinophils, filter = "simple", batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_eosinophils_sva-v{ver}.csv"))
```

### Tumaco visit 1 monocytes

```{r}
tv1_monocytes <- tcv1_monocytes %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv1_monocytes", file = glue("rda/tmrc3_tv1_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tv1_monocytes")
cpm_data <- normalize_expt(tv1_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(tv1_monocytes, convert="cpm", filter=TRUE, batch="svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_monocytes_sva-v{ver}.csv"))
```

### Tumaco visit 2 monocytes

```{r}
tv2_monocytes <- tcv2_monocytes %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv2_monocytes", file = glue("rda/tmrc3_tv2_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tv2_monocytes")
cpm_data <- normalize_expt(tv2_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(tv2_monocytes, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_monocytes_sva-v{ver}.csv"))
```

### Tumaco visit 3 monocytes

```{r}
tv3_monocytes <- tcv3_monocytes %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv3_monocytes", file = glue("rda/tmrc3_tv3_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "tv3_monocytes")
cpm_data <- normalize_expt(tv3_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(tv3_monocytes, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_monocytes_sva-v{ver}.csv"))
```

### Tumaco visit 1 neutrophils

```{r}
tv1_neutrophils <- tcv1_neutrophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv1_neutrophils", file = glue("rda/tmrc3_tv1_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tv1_neutrophils")
cpm_data <- normalize_expt(tv1_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(tv1_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv1_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv1_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv1_neutrophils_sva-v{ver}.csv"))
```

### Tumaco visit 2 neutrophils

```{r}
tv2_neutrophils <- tcv2_neutrophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv2_neutrophils", file = glue("rda/tmrc3_tv2_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tv2_neutrophils")
cpm_data <- normalize_expt(tv2_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(tv2_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv2_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv2_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv2_neutrophils_sva-v{ver}.csv"))
```

### Tumaco visit 3 neutrophils

```{r}
tv3_neutrophils <- tcv3_neutrophils %>%
  subset_expt(subset = "clinic=='Tumaco'")
save(list = "tv3_neutrophils", file = glue("rda/tmrc3_tv3_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "tv3_neutrophils")
cpm_data <- normalize_expt(tv3_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(tv3_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/tv3_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(tv3_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/tv3_neutrophils_sva-v{ver}.csv"))
```

# Dataset: Only Cali samples

Even though we are only considering the Tumaco samples, one might wish
to perform comparisons among the Cali samples.  The following blocks
therefore separate them out in a similar fashion.

```{r}
c_clinical <- tc_clinical %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "c_clinical", file = glue("rda/tmrc3_c_clinical-v{ver}.rda"))
data_structures <- c(data_structures, "c_clinical")

cpm_data <- normalize_expt(c_clinical, convert = "cpm")
c_clinical_cpm_variance <- variance_expt(cpm_data)

written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_clinical-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_clinical, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
                 file = glue("rpkm/t_clinical-v{ver}.csv"))
cpm_data <- normalize_expt(c_clinical, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_clinical_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_clinical, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_clinical_sva-v{ver}.csv"))

c_clinical_nobiop <- subset_expt(c_clinical, subset = "typeofcells!='biopsy'")
save(list = "c_clinical_nobiop", file = glue("rda/tmrc3_c_clinical_nobiop-v{ver}.rda"))
data_structures <- c(data_structures, "c_clinical_nobiop")
cpm_data <- normalize_expt(c_clinical_nobiop, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_clinical_nobiop-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_clinical, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_clinical_sva-v{ver}.csv"))

cpm_data <- normalize_expt(c_clinical_nobiop, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_clinical_nobiop_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_clinical_nobiop, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_clinical_nobiop_sva-v{ver}.csv"))
```

### Pull out the top-n most variant genes

```{r}
top_expt <- normalize_expt(c_clinical_cpm_variance, filter = TRUE) %>%
  variance_expt()
top_fd <- fData(top_expt)
top_ex <- exprs(top_expt)
idx <- order(top_fd[["exprs_gene_stdev"]], decreasing = TRUE)
top_df <- top_fd[idx, ]
written <- write_xlsx(
    data = head(top_fd, n = 100),
    excel = glue("excel/c_clinical_cpm_stdev_top100-v{ver}.xlsx"))

bottom_expt <- normalize_expt(c_clinical_cpm_variance, filter = "simple", threshold = 500) %>%
  variance_expt()
bottom_fd <- fData(bottom_expt)
bottom_ex <- exprs(bottom_expt)
idx <- order(bottom_fd[["exprs_gene_stdev"]])
bottom <- bottom_fd[idx, ]
written <- write_xlsx(
    data = head(bottom_fd, n = 100),
    excel = glue("excel/c_clinical_cpm_stdev_bottom100-v{ver}.xlsx"))
```

## Summarize: Collect Cali sample numbers.

This is copy/pasted from above and prints out how many of each sample
type there are of each category.

```{r}
table(pData(c_clinical)$drug)
table(pData(c_clinical)$clinic)
table(pData(c_clinical)$finaloutcome)
table(pData(c_clinical)$typeofcells)
table(pData(c_clinical)$visit)
summary(as.numeric(pData(c_clinical)$eb_lc_tiempo_evolucion))
summary(as.numeric(pData(c_clinical)$eb_lc_tto_mcto_glucan_dosis))
summary(as.numeric(pData(c_clinical)$v3_lc_ejey_lesion_mm_1))
summary(as.numeric(pData(c_clinical)$v3_lc_lesion_area_1))
summary(as.numeric(pData(c_clinical)$v3_lc_ejex_ulcera_mm_1))
table(pData(c_clinical)$eb_lc_sexo)
table(pData(c_clinical)$eb_lc_etnia)
summary(as.numeric(pData(c_clinical)$edad))
table(pData(c_clinical)$eb_lc_peso)
table(pData(c_clinical)$eb_lc_estatura)

length(unique(pData(c_clinical)[["codigo_paciente"]]))
only_cure <- pData(c_clinical)[["finaloutcome"]] == "cure"
c_meta <- pData(c_clinical)[only_cure, ]
length(unique(c_meta[["codigo_paciente"]]))
only_fail <- pData(c_clinical)[["finaloutcome"]] == "failure"
f_meta <- pData(c_clinical)[only_fail, ]
length(unique(f_meta[["codigo_paciente"]]))
```

## Split the data by cure/fail for some counting

```{r}
cali_cure <- subset_expt(c_clinical, subset = "finaloutcome=='cure'")
table(pData(cali_cure)$visitnumber)

cali_fail <- subset_expt(c_clinical, subset = "finaloutcome=='failure'")
table(pData(cali_fail)$visitnumber)
```

## Subset: Create Cali-specific versions of the various structures

I previously made a bunch of data subsets by visit, cell type, etc.
So let us overwrite them all with versions which contain only the
Cali samples.

There is no going back to the Cali samples after this block unless we
regenerate the data from the original expressionsets.

### Cali biopsies

```{r}
c_biopsies <- tc_biopsies %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "c_biopsies", file = glue("rda/tmrc3_c_biopsies-v{ver}.rda"))
data_structures <- c(data_structures, "c_biopsies")
cpm_data <- normalize_expt(c_biopsies, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_biopsies-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_biopsies, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_biopsies-v{ver}.csv"))

cpm_data <- normalize_expt(c_biopsies, convert = "cpm", filter = TRUE)
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_biopsies_nosva-v{ver}.xlsx"))
```

### Cali Eosinophils

```{r}
c_eosinophils <- tc_eosinophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "c_eosinophils", file = glue("rda/tmrc3_c_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "c_eosinophils")
cpm_data <- normalize_expt(c_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_eosinophils-v{ver}.xlsx"))
cpm_data <- normalize_expt(c_eosinophils, convert = "cpm", filter = TRUE)
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_eosinophils_nosva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_eosinophils, filter = TRUE) %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_eosinophils_nosva-v{ver}.csv"))
```

### Cali subsets monocytes

```{r}
c_monocytes <- tc_monocytes %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "c_monocytes", file = glue("rda/tmrc3_c_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "c_monocytes")
cpm_data <- normalize_expt(c_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(c_monocytes, convert = "cpm", filter = TRUE)
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_monocytes_nosva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_monocytes_sva-v{ver}.csv"))
```

### Cali subsets neutrophils

```{r}
c_neutrophils <- tc_neutrophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "c_neutrophils", file = glue("rda/tmrc3_c_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "c_neutrophils")
cpm_data <- normalize_expt(c_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_neutrophil-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
                 file = glue("rpkm/c_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(c_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(c_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/c_neutrophils_sva-v{ver}.csv"))
```

### Cali Visit 1 samples

```{r}
cv1_samples <- tcv1_samples %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv1_samples", file = glue("rda/tmrc3_cv1_samples-v{ver}.rda"))
data_structures <- c(data_structures, "cv1_samples")
cpm_data <- normalize_expt(cv1_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_samples-v{ver}.csv"))
cpm_data <- normalize_expt(cv1_samples, convert =  "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_samples_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_samples, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_samples_sva-v{ver}.csv"))
```

### Cali Visit 2 samples

```{r}
cv2_samples <- tcv2_samples %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv2_samples", file = glue("rda/tmrc3_cv2_samples-v{ver}.rda"))
data_structures <- c(data_structures, "cv2_samples")
cpm_data <- normalize_expt(cv2_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_samples-v{ver}.csv"))
cpm_data <- normalize_expt(cv2_samples, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_samples_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_samples, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_samples_sva-v{ver}.csv"))
```

### Cali visit 3 samples

```{r}
cv3_samples <- tcv3_samples %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv3_samples", file = glue("rda/tmrc3_cv3_samples-v{ver}.rda"))
data_structures <- c(data_structures, "cv3_samples")
cpm_data <- normalize_expt(cv3_samples, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_samples-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_samples, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_samples-v{ver}.csv"))
cpm_data <- normalize_expt(cv3_samples, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_samples_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_samples, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_samples_sva-v{ver}.csv"))
```

### Cali visit 1 Eosinophils

```{r}
cv1_eosinophils <- tcv1_eosinophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv1_eosinophils", file = glue("rda/tmrc3_cv1_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "cv1_eosinophils")
cpm_data <- normalize_expt(cv1_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_eosinophils-v{ver}.csv"))
cpm_data <- normalize_expt(cv1_eosinophils, convert = "cpm", filter = TRUE)
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_eosinophils_nosva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_eosinophils, filter = TRUE) %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_eosinophils_nosva-v{ver}.csv"))
```

### Cali visit 2 samples

```{r}
cv2_eosinophils <- tcv2_eosinophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv2_eosinophils", file = glue("rda/tmrc3_cv2_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "cv2_eosinophils")
cpm_data <- normalize_expt(cv2_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_eosinophils-v{ver}.csv"))
cpm_data <- normalize_expt(cv2_eosinophils, convert = "cpm", filter = TRUE)
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_eosinophils_nosva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_eosinophils, filter = TRUE) %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_eosinophils_nosva-v{ver}.csv"))
```

### Cali visit 3 eosinophil

```{r}
cv3_eosinophils <- tcv3_eosinophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv3_eosinophils", file = glue("rda/tmrc3_cv3_eosinophils-v{ver}.rda"))
data_structures <- c(data_structures, "cv3_eosinophils")
cpm_data <- normalize_expt(cv3_eosinophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_eosinophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_eosinophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_eosinophils-v{ver}.csv"))
cpm_data <- normalize_expt(cv3_eosinophils, convert = "cpm", filter = TRUE)
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_eosinophils_nosva-v{ver}.xlsx"))
## Note, the cbcb filter left behind a sufficient number of zeros that it confused cpm.
rpkm_data <- normalize_expt(cv3_eosinophils, filter = "simple") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_eosinophils_nosva-v{ver}.csv"))
```

### Cali visit 1 monocytes

```{r}
cv1_monocytes <- tcv1_monocytes %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv1_monocytes", file = glue("rda/tmrc3_cv1_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "cv1_monocytes")
cpm_data <- normalize_expt(cv1_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(cv1_monocytes, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_monocytes_sva-v{ver}.csv"))
```

### Cali visit 2 monocytes

```{r}
cv2_monocytes <- tcv2_monocytes %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv2_monocytes", file = glue("rda/tmrc3_cv2_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "cv2_monocytes")
cpm_data <- normalize_expt(cv2_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(cv2_monocytes, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_monocytes_sva-v{ver}.csv"))
```

### Cali visit 3 monocytes

```{r}
cv3_monocytes <- tcv3_monocytes %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv3_monocytes", file = glue("rda/tmrc3_cv3_monocytes-v{ver}.rda"))
data_structures <- c(data_structures, "cv3_monocytes")
cpm_data <- normalize_expt(cv3_monocytes, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_monocytes-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_monocytes, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_monocytes-v{ver}.csv"))
cpm_data <- normalize_expt(cv3_monocytes, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_monocytes_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_monocytes, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_monocytes_sva-v{ver}.csv"))
```

### Cali visit 1 neutrophils

```{r}
cv1_neutrophils <- tcv1_neutrophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv1_neutrophils", file = glue("rda/tmrc3_cv1_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "cv1_neutrophils")
cpm_data <- normalize_expt(cv1_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(cv1_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv1_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv1_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv1_neutrophils_sva-v{ver}.csv"))
```

### Cali visit 2 neutrophils

```{r}
cv2_neutrophils <- tcv2_neutrophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv2_neutrophils", file = glue("rda/tmrc3_cv2_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "cv2_neutrophils")
cpm_data <- normalize_expt(cv2_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(cv2_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv2_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv2_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv2_neutrophils_sva-v{ver}.csv"))
```

### Cali visit 3 neutrophils

```{r}
cv3_neutrophils <- tcv3_neutrophils %>%
  subset_expt(subset = "clinic=='Cali'")
save(list = "cv3_neutrophils", file = glue("rda/tmrc3_cv3_neutrophils-v{ver}.rda"))
data_structures <- c(data_structures, "cv3_neutrophils")
cpm_data <- normalize_expt(cv3_neutrophils, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_neutrophils-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_neutrophils, filter = TRUE, convert = "rpkm",
                            na_to_zero = TRUE, column = "mean_cds_len")
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_neutrophils-v{ver}.csv"))
cpm_data <- normalize_expt(cv3_neutrophils, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/cv3_neutrophils_sva-v{ver}.xlsx"))
rpkm_data <- normalize_expt(cv3_neutrophils, filter = TRUE, batch = "svaseq") %>%
  normalize_expt(convert = "rpkm", column = "mean_cds_len", na_to_zero = TRUE)
write.csv(x = as.data.frame(exprs(rpkm_data)),
          file = glue("rpkm/cv3_neutrophils_sva-v{ver}.csv"))
```

### Tumaco compare visits

Is this redundant?

```{r}
tc_visit <- set_expt_conditions(
  tc_clinical, fact = "visitnumber", colors = color_choices[["visit2"]]) %>%
  set_expt_batches(fact = "finaloutcome") %>%
  subset_expt(subset = "typeofcells!='biopsy'")
save(list = "tc_visit", file = glue("rda/tmrc3_tc_visit-v{ver}.rda"))
data_structures <- c(data_structures, "tc_visit")

tc_v1vs <- tc_visit
pData(tc_v1vs)[["visit"]] <- as.character(pData(tc_v1vs)[["visitnumber"]])
v1_samples <- pData(tc_v1vs)[["visit"]] == "1"
pData(tc_v1vs)[v1_samples, "visit"] <- "first"
pData(tc_v1vs)[!v1_samples, "visit"] <- "later"
pData(tc_v1vs)[["visit"]] <- as.factor(pData(tc_v1vs)[["visit"]])
tc_v1vs <- set_expt_conditions(
  tc_v1vs, fact = "visit", colors = color_choices[["two_visits"]])
save(list = "tc_v1vs", file = glue("rda/tmrc3_tc_v1vs-v{ver}.rda"))
data_structures <- c(data_structures, "tc_v1vs")
t_v1vs <- subset_expt(tc_v1vs, subset = "clinic == 'Tumaco'")
save(list = "t_v1vs", file = glue("rda/tmrc3_t_v1vs-v{ver}.rda"))
data_structures <- c(data_structures, "t_v1vs")

t_visit <- set_expt_conditions(
  t_clinical, fact = "visitnumber", colors = color_choices[["visit2"]]) %>%
  set_expt_batches(fact = "finaloutcome") %>%
  subset_expt(subset = "typeofcells!='biopsy'")
save(list = "t_visit", file = glue("rda/tmrc3_t_visit-v{ver}.rda"))
data_structures <- c(data_structures, "t_visit")
cpm_data <- normalize_expt(t_visit, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_visit-v{ver}.xlsx"))
cpm_data <- normalize_expt(t_visit, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/t_visit_sva-v{ver}.xlsx"))
```

### Visit cure/fail contrasts

```{r}
visit_cf_expt_factor <- paste0("v", pData(t_clinical)[["visitnumber"]],
                               pData(t_clinical)[["finaloutcome"]])
t_visitcf <- set_expt_conditions(t_clinical, fact = visit_cf_expt_factor)
save(list = "t_visitcf", file = glue("rda/tmrc3_t_visitcf-v{ver}.rda"))
data_structures <- c(data_structures, "t_visitcf")

t_visitcf_eosinophil <- subset_expt(t_visitcf, subset = "typeofcells=='eosinophils'")
save(list = "t_visitcf_eosinophil", file = glue("rda/tmrc3_t_visitcf_eosinophil-v{ver}.rda"))
data_structures <- c(data_structures, "t_visitcf_eosinophil")

t_visitcf_monocyte <- subset_expt(t_visitcf, subset = "typeofcells=='monocytes'")
save(list = "t_visitcf_monocyte", file = glue("rda/tmrc3_t_visitcf_monocyte-v{ver}.rda"))
data_structures <- c(data_structures, "t_visitcf_monocyte")

t_visitcf_neutrophil <- subset_expt(t_visitcf, subset = "typeofcells=='neutrophils'")
save(list = "t_visitcf_neutrophil", file = glue("rda/tmrc3_t_visitcf_neutrophil-v{ver}.rda"))
data_structures <- c(data_structures, "t_visitcf_neutrophil")
```


### Cali compare visits

Is this redundant?

```{r}
c_visit <- set_expt_conditions(c_clinical, fact = "visitnumber") %>%
  set_expt_batches(fact = "finaloutcome") %>%
  subset_expt(subset = "typeofcells!='biopsy'")
save(list = "c_visit", file = glue("rda/tmrc3_c_visit-v{ver}.rda"))
data_structures <- c(data_structures, "c_visit")
cpm_data <- normalize_expt(c_visit, convert = "cpm")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_tumaco/c_visit-v{ver}.xlsx"))
cpm_data <- normalize_expt(c_visit, convert = "cpm", filter = TRUE, batch = "svaseq")
written <- write_xlsx(
    exprs(cpm_data),
    excel = glue("cpm/4_cali/c_visit_sva-v{ver}.xlsx"))
```

### Visit cure/fail contrasts

```{r}
visit_cf_expt_factor <- paste0("v", pData(c_clinical)[["visitnumber"]],
                               pData(c_clinical)[["finaloutcome"]])
c_visitcf <- set_expt_conditions(c_clinical, fact = visit_cf_expt_factor)
save(list = "c_visitcf", file = glue("rda/tmrc3_c_visitcf-v{ver}.rda"))
data_structures <- c(data_structures, "c_visitcf")

c_visitcf_eosinophil <- subset_expt(c_visitcf, subset = "typeofcells=='eosinophils'")
save(list = "c_visitcf_eosinophil", file = glue("rda/tmrc3_c_visitcf_eosinophil-v{ver}.rda"))
data_structures <- c(data_structures, "c_visitcf_eosinophil")

c_visitcf_monocyte <- subset_expt(c_visitcf, subset = "typeofcells=='monocytes'")
save(list = "c_visitcf_monocyte", file = glue("rda/tmrc3_c_visitcf_monocyte-v{ver}.rda"))
data_structures <- c(data_structures, "c_visitcf_monocyte")

c_visitcf_neutrophil <- subset_expt(c_visitcf, subset = "typeofcells=='neutrophils'")
save(list = "c_visitcf_neutrophil", file = glue("rda/tmrc3_c_visitcf_neutrophil-v{ver}.rda"))
data_structures <- c(data_structures, "c_visitcf_neutrophil")
```

# Summarize: Tabulate sample numbers

Here is an outline of the samples in their current state:

## Tumaco

* By type: `r ncol(exprs(t_clinical))`, Cure: `r sum(pData(t_clinical)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(t_clinical)[["finaloutcome"]] == "failure")`
    *  Biopsy: `r all_types[["biopsy"]]`, Cure: `r sum(pData(t_biopsies)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(t_biopsies)[["finaloutcome"]] == "failure")`
        * All biopsy samples are visit 1.
    *  Eosinophils: `r all_types[["eosinophils"]]`, Cure: `r sum(pData(t_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(t_eosinophils)[["finaloutcome"]] == "failure")`
        * V1 Eosinophils: `r nrow(pData(tv1_eosinophils))`, Cure: `r sum(pData(tv1_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv1_eosinophils)[["finaloutcome"]] == "failure")`
        * V2 Eosinophils: `r nrow(pData(tv2_eosinophils))`, Cure: `r sum(pData(tv2_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv2_eosinophils)[["finaloutcome"]] == "failure")`
        * V3 Eosinophils: `r nrow(pData(tv3_eosinophils))`, Cure: `r sum(pData(tv3_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv3_eosinophils)[["finaloutcome"]] == "failure")`
    *  Monocytes: `r all_types[["monocytes"]]`, Cure: `r sum(pData(t_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(t_monocytes)[["finaloutcome"]] == "failure")`
        * V1 Monocytes: `r nrow(pData(tv1_monocytes))`, Cure: `r sum(pData(tv1_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv1_monocytes)[["finaloutcome"]] == "failure")`
        * V2 Monocytes: `r nrow(pData(tv2_monocytes))`, Cure: `r sum(pData(tv2_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv2_monocytes)[["finaloutcome"]] == "failure")`
        * V3 Monocytes: `r nrow(pData(tv3_monocytes))`, Cure: `r sum(pData(tv3_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv3_monocytes)[["finaloutcome"]] == "failure")`
    *  Neutrophils: `r all_types[["neutrophils"]]`, Cure: `r sum(pData(t_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(t_neutrophils)[["finaloutcome"]] == "failure")`
        * V1 Neutrophils: `r nrow(pData(tv1_neutrophils))`, Cure: `r sum(pData(tv1_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv1_neutrophils)[["finaloutcome"]] == "failure")`
        * V2 Neutrophils: `r nrow(pData(tv2_neutrophils))`, Cure: `r sum(pData(tv2_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv2_neutrophils)[["finaloutcome"]] == "failure")`
        * V3 Neutrophils: `r nrow(pData(tv3_neutrophils))`, Cure: `r sum(pData(tv3_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(tv3_neutrophils)[["finaloutcome"]] == "failure")`


```{r}
ncol(exprs(t_clinical))
sum(pData(t_clinical)[["finaloutcome"]] == "cure")
sum(pData(t_clinical)[["finaloutcome"]] == "failure")

all_types[["biopsy"]]
sum(pData(t_biopsies)[["finaloutcome"]] == "cure")
sum(pData(t_biopsies)[["finaloutcome"]] == "failure")

all_types[["eosinophils"]]
sum(pData(t_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(t_eosinophils)[["finaloutcome"]] == "failure")

nrow(pData(tv1_eosinophils))
sum(pData(tv1_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(tv1_eosinophils)[["finaloutcome"]] == "failure")

nrow(pData(tv2_eosinophils))
sum(pData(tv2_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(tv2_eosinophils)[["finaloutcome"]] == "failure")

nrow(pData(tv3_eosinophils))
sum(pData(tv3_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(tv3_eosinophils)[["finaloutcome"]] == "failure")

all_types[["monocytes"]]
sum(pData(t_monocytes)[["finaloutcome"]] == "cure")
sum(pData(t_monocytes)[["finaloutcome"]] == "failure")

nrow(pData(tv1_monocytes))
sum(pData(tv1_monocytes)[["finaloutcome"]] == "cure")
sum(pData(tv1_monocytes)[["finaloutcome"]] == "failure")

nrow(pData(tv2_monocytes))
sum(pData(tv2_monocytes)[["finaloutcome"]] == "cure")
sum(pData(tv2_monocytes)[["finaloutcome"]] == "failure")

nrow(pData(tv3_monocytes))
sum(pData(tv3_monocytes)[["finaloutcome"]] == "cure")
sum(pData(tv3_monocytes)[["finaloutcome"]] == "failure")

all_types[["neutrophils"]]
sum(pData(t_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(t_neutrophils)[["finaloutcome"]] == "failure")

nrow(pData(tv1_neutrophils))
sum(pData(tv1_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(tv1_neutrophils)[["finaloutcome"]] == "failure")

nrow(pData(tv2_monocytes))
sum(pData(tv2_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(tv2_neutrophils)[["finaloutcome"]] == "failure")

nrow(pData(tv3_neutrophils))
sum(pData(tv3_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(tv3_neutrophils)[["finaloutcome"]] == "failure")
```

## Cali

* By type: `r ncol(exprs(c_clinical))`, Cure: `r sum(pData(c_clinical)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(c_clinical)[["finaloutcome"]] == "failure")`
    *  Biopsy: `r all_types[["biopsy"]]`, Cure: `r sum(pData(c_biopsies)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(c_biopsies)[["finaloutcome"]] == "failure")`
        * All biopsy samples are visit 1.
    *  Eosinophils: `r all_types[["eosinophils"]]`, Cure: `r sum(pData(c_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(c_eosinophils)[["finaloutcome"]] == "failure")`
        * V1 Eosinophils: `r nrow(pData(cv1_eosinophils))`, Cure: `r sum(pData(cv1_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv1_eosinophils)[["finaloutcome"]] == "failure")`
        * V2 Eosinophils: `r nrow(pData(cv2_eosinophils))`, Cure: `r sum(pData(cv2_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv2_eosinophils)[["finaloutcome"]] == "failure")`
        * V3 Eosinophils: `r nrow(pData(cv3_eosinophils))`, Cure: `r sum(pData(cv3_eosinophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv3_eosinophils)[["finaloutcome"]] == "failure")`
    *  Monocytes: `r all_types[["monocytes"]]`, Cure: `r sum(pData(c_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(c_monocytes)[["finaloutcome"]] == "failure")`
        * V1 Monocytes: `r nrow(pData(cv1_monocytes))`, Cure: `r sum(pData(cv1_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv1_monocytes)[["finaloutcome"]] == "failure")`
        * V2 Monocytes: `r nrow(pData(cv2_monocytes))`, Cure: `r sum(pData(cv2_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv2_monocytes)[["finaloutcome"]] == "failure")`
        * V3 Monocytes: `r nrow(pData(cv3_monocytes))`, Cure: `r sum(pData(cv3_monocytes)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv3_monocytes)[["finaloutcome"]] == "failure")`
    *  Neutrophils: `r all_types[["neutrophils"]]`, Cure: `r sum(pData(c_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(c_neutrophils)[["finaloutcome"]] == "failure")`
        * V1 Neutrophils: `r nrow(pData(cv1_neutrophils))`, Cure: `r sum(pData(cv1_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv1_neutrophils)[["finaloutcome"]] == "failure")`
        * V2 Neutrophils: `r nrow(pData(cv2_neutrophils))`, Cure: `r sum(pData(cv2_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv2_neutrophils)[["finaloutcome"]] == "failure")`
        * V3 Neutrophils: `r nrow(pData(cv3_neutrophils))`, Cure: `r sum(pData(cv3_neutrophils)[["finaloutcome"]] == "cure")`, Fail: `r sum(pData(cv3_neutrophils)[["finaloutcome"]] == "failure")`

```{r}
ncol(exprs(c_clinical))
sum(pData(c_clinical)[["finaloutcome"]] == "cure")
sum(pData(c_clinical)[["finaloutcome"]] == "failure")

#all_types[["biopsy"]]
sum(pData(c_biopsies)[["finaloutcome"]] == "cure")
sum(pData(c_biopsies)[["finaloutcome"]] == "failure")

#all_types[["eosinophils"]]
sum(pData(c_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(c_eosinophils)[["finaloutcome"]] == "failure")

nrow(pData(cv1_eosinophils))
sum(pData(cv1_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(cv1_eosinophils)[["finaloutcome"]] == "failure")

nrow(pData(cv2_eosinophils))
sum(pData(cv2_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(cv2_eosinophils)[["finaloutcome"]] == "failure")

nrow(pData(cv3_eosinophils))
sum(pData(cv3_eosinophils)[["finaloutcome"]] == "cure")
sum(pData(cv3_eosinophils)[["finaloutcome"]] == "failure")

#all_types[["monocytes"]]
sum(pData(c_monocytes)[["finaloutcome"]] == "cure")
sum(pData(c_monocytes)[["finaloutcome"]] == "failure")

nrow(pData(cv1_monocytes))
sum(pData(cv1_monocytes)[["finaloutcome"]] == "cure")
sum(pData(cv1_monocytes)[["finaloutcome"]] == "failure")

nrow(pData(cv2_monocytes))
sum(pData(cv2_monocytes)[["finaloutcome"]] == "cure")
sum(pData(cv2_monocytes)[["finaloutcome"]] == "failure")

nrow(pData(cv3_monocytes))
sum(pData(cv3_monocytes)[["finaloutcome"]] == "cure")
sum(pData(cv3_monocytes)[["finaloutcome"]] == "failure")

#all_types[["neutrophils"]]
sum(pData(c_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(c_neutrophils)[["finaloutcome"]] == "failure")

nrow(pData(cv1_neutrophils))
sum(pData(cv1_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(cv1_neutrophils)[["finaloutcome"]] == "failure")

nrow(pData(cv2_monocytes))
sum(pData(cv2_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(cv2_neutrophils)[["finaloutcome"]] == "failure")

nrow(pData(cv3_neutrophils))
sum(pData(cv3_neutrophils)[["finaloutcome"]] == "cure")
sum(pData(cv3_neutrophils)[["finaloutcome"]] == "failure")
```

# An external dataset

One potentially interesting comparisons point, especially vis a vis
the Biopsy samples, is to compare with another set of recent
L. braziliensis infected biopsy samples.  The following sets up an
expressionset using the raw data from bioproject PRJNA525604.

```{r}
new_sample_sheet <- sm(
  suppressWarnings(gather_preprocessing_metadata(
    "sample_sheets/scott_sra_samples.xlsx", specification = make_rnaseq_spec())))

external_cf <- create_expt("sample_sheets/scott_sra_samples_modified.xlsx",
                           gene_info = hs_annot, file_column = "hisatcounttable") %>%
  set_expt_conditions(fact = "treatmentoutcome") %>%
    set_expt_batches(fact = "sex") %>%
    subset_expt(subset = "condition!='na'")
pData(external_cf)[["ParasiteSpecies"]] <- "lvbraziliensis"
save(list = "external_cf", file = glue("rda/tmrc3_external_cf-v{ver}.rda"))
data_structures <- c(data_structures, "external_cf")

biopsy_cf <- set_expt_conditions(tc_biopsies, fact = "finaloutcome")
pData(biopsy_cf)[["lab"]] <- "Colombia"
pData(external_cf)[["lab"]] <- "Brazil"
pData(external_cf)[["finaloutcome"]] <- pData(external_cf)[["treatmentoutcome"]]

tmrc3_external <- combine_expts(external_cf, biopsy_cf) %>%
  set_expt_conditions(fact = "lab", colors = color_choices[["labs"]]) %>%
  set_expt_batches(fact = "finaloutcome")

save(list = "tmrc3_external", file = glue("rda/tmrc3_external-v{ver}.rda"))
data_structures <- c(data_structures, "tmrc3_external")

tmrc3_external_species <- set_expt_conditions(
  tmrc3_external, fact = "ParasiteSpecies", colors = color_choices[["parasite"]]) %>%
  set_expt_batches(fact = "lab")
save(list = "tmrc3_external_species", file = glue("rda/tmrc3_external_sepcies-v{ver}.rda"))
data_structures <- c(data_structures, "tmrc3_external_species")

## In my singularity container, this fails with a pthread problem, not doing quant here.
#tt <- normalize_expt(tmrc3_external_species, filter = TRUE, transform = "log2",
#                     convert = "cpm", norm = "quant")
tt <- normalize_expt(tmrc3_external_species, filter = TRUE, transform = "log2",
                     convert = "cpm", norm = "tmm")
plot_pca(tt)

confused <- set_expt_conditions(
  hs_expt, fact = "ParasiteSpecies", colors = color_choices[["parasite"]])
```

## Ethnicity

```{r t_etnia_expt}
tc_etnia_expt <- set_expt_conditions(
  tc_clinical, fact = "etnia", colors = color_choices[["ethnicity"]])
save(list = "tc_etnia_expt", file = glue("rda/tmrc3_tc_etnia-v{ver}.rda"))
data_structures <- c(data_structures, "tc_etnia_expt")

t_etnia_expt <- set_expt_conditions(
  t_clinical, fact = "etnia", colors = color_choices[["ethnicity"]])
save(list = "t_etnia_expt", file = glue("rda/tmrc3_t_etnia-v{ver}.rda"))
data_structures <- c(data_structures, "t_etnia_expt")
```

# Save all data structures into one big pile

```{r}
save(list = data_structures, file = glue("rda/tmrc3_data_structures-v{ver}.rda"))
```

```{r, eval=FALSE}
tmp <- loadme(filename = savefile)
```
