---
title: "TMRC3 `r Sys.getenv('VERSION')`: Differential Expression analyses, Tumaco only."
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
runtime: shiny
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: zenburn
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body .main-container {
  max-width: 1600px;
}
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
  font-size: 16px
}
</style>

```{r options, include=FALSE}
library(hpgltools)
library(dplyr)
library(forcats)
library(glue)
## library(DOSE)  ## If I cannot get the darn importClassesFrom working right just library it...

knitr::opts_knit$set(progress = TRUE, verbose = TRUE, width = 90, echo = TRUE)
knitr::opts_chunk$set(
  error = TRUE, fig.width = 8, fig.height = 8, fig.retina = 2,
  out.width = "100%", dev = "png",
  dev.args = list(png = list(type = "cairo-png")))
old_options <- options(digits = 4, stringsAsFactors = FALSE, knitr.duplicate.label = "allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size = 12))
ver <- Sys.getenv("VERSION")
previous_file <- ""
rundate <- format(Sys.Date(), format = "%Y%m%d")

##tmp <- try(sm(loadme(filename = gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = previous_file))))
rmd_file <- glue("tmrc3_differential_expression_tumaco_{ver}.Rmd")
savefile <- gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = rmd_file)
loaded <- load(file = glue("rda/tmrc3_data_structures-v{ver}.rda"))
xlsx_prefix <- "analyses/4_tumaco"
cf_prefix <- glue("{xlsx_prefix}/DE_Cure_vs_Fail")
```

# Changelog

* 202309: Disabled GSVA analyses until/unless we get permission to
  include the mSigDB 7.5.1 release.  I will simplify the filenames so
  that one may easily drop in a downloaded copy of the data and run
  those blocks.  Until then, I guess you will have to trust me when I
  say those blocks all work?
* 202309: Moved all GSEA analyses to 04lrt_gsea_gsva.Rmd
* 202309 next day: Moving GSEA back because it adds too much
  complexity to save/reload the DE results for gProfiler and friends.
* Still hunting for messed up colors, changed input data to match new version.

# Introduction

The various differential expression analyses of the data generated in tmrc3_datasets
will occur in this document.

## Define contrasts for DE analyses

Each of the following lists describes the set of contrasts that I
think are interesting for the various ways one might consider the
TMRC3 dataset.  The variables are named according to the assumed data
with which they will be used, thus tc_cf_contrasts is expected to be
used for the Tumaco+Cali data and provide a series of cure/fail
comparisons which (to the extent possible) across both locations.  In
every case, the name of the list element will be used as the contrast
name, and will thus be seen as the sheet name in the output xlsx
file(s); the two pieces of the character vector value are the
numerator and denominator of the associated contrast.

## GSEA

The GSEA analyses will follow each DE analysis during this document.

Most (all?) of the GSEA analyses used in this paper were done via
gProfiler rather than goseq/clusterProfiler/topGO/GOstats.  Primarily
because it is so easy to invoke gprofiler.

```{r}
clinic_contrasts <- list(
  "clinics" = c("Cali", "Tumaco"))
## In some cases we have no Cali failure samples, so there remain only 2
## contrasts that are likely of interest
tc_cf_contrasts <- list(
  "tumaco" = c("Tumacofailure", "Tumacocure"),
  "cure" = c("Tumacocure", "Calicure"))
## In other cases, we have cure/fail for both places.
clinic_cf_contrasts <- list(
  "cali" = c("Califailure", "Calicure"),
  "tumaco" = c("Tumacofailure", "Tumacocure"),
  "cure" = c("Tumacocure", "Calicure"),
  "fail" = c("Tumacofailure", "Califailure"))
cf_contrast <- list(
  "outcome" = c("Tumacofailure", "Tumacocure"))
t_cf_contrast <- list(
  "outcome" = c("failure", "cure"))
visitcf_contrasts <- list(
  "v1cf" = c("v1failure", "v1cure"),
  "v2cf" = c("v2failure", "v2cure"),
  "v3cf" = c("v3failure", "v3cure"))
visit_contrasts <- list(
  "v2v1" = c("c2", "c1"),
  "v3v1" = c("c3", "c1"),
  "v3v2" = c("c3", "c2"))
visit_v1later <- list(
  "later_vs_first" = c("later", "first"))
celltypes <- list(
  "eo_mono" = c("eosinophils", "monocytes"),
  "ne_mono" = c("neutrophils", "monocytes"),
  "eo_ne" = c("eosinophils", "neutrophils"))
ethnicity_contrasts <- list(
  "mestizo_indigenous" = c("mestiza", "indigena"),
  "mestizo_afrocol" = c("mestiza", "afrocol"),
  "indigenous_afrocol" = c("indigena", "afrocol"))
```

# Only Tumaco samples

Start over, this time with only the samples from Tumaco.  We currently
are assuming these will prove to be the only analyses used for final
interpretation.  This is primarily because we have insufficient
failed treatment samples from Cali.

## All samples

Start by considering all Tumaco cell types.  Note that in this case we
only use SVA, primarily because I am not certain what would be an
appropriate batch factor, perhaps visit?

```{r}
t_cf_clinical_de_sva <- all_pairwise(t_clinical, model_batch = "svaseq", filter = TRUE)
t_cf_clinical_de_sva
t_cf_clinical_table_sva <- combine_de_tables(
  t_cf_clinical_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/t_clinical_cf_tables_sva-v{ver}.xlsx"))
t_cf_clinical_table_sva
t_cf_clinical_table_sva[["plots"]][["outcome"]][["deseq_ma_plots"]]
t_cf_clinical_sig_sva <- extract_significant_genes(
  t_cf_clinical_table_sva,
  excel = glue("{cf_prefix}/All_Samples/t_clinical_cf_sig_sva-v{ver}.xlsx"))
t_cf_clinical_sig_sva

dim(t_cf_clinical_sig_sva$deseq$ups[[1]])
dim(t_cf_clinical_sig_sva$deseq$downs[[1]])
```

## gProfiler search of all Tumaco samples

The following gProfiler searches use the all_gprofiler() function
instead of simple_gprofiler().  As a result, the results are separated
by {contrast}_{direction}.  Thus 'outcome_down'.

The same plots are available as the previous gProfiler searches, but
in many of the following runs, I used the dotplot() function to get a
slightly different view of the results.

```{r}
t_cf_clinical_gp <- all_gprofiler(t_cf_clinical_sig_sva)
t_cf_clinical_gp

written <- write_gprofiler_data(
  t_cf_clinical_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/clinical_cure_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_clinical_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/clinical_cure_down-v{ver}.xlsx"))
## Wikipathways of the up c/f genes
enrichplot::dotplot(t_cf_clinical_gp[["outcome_up"]][["WP_enrich"]])

## Transcription factor database of the up c/f genes
enrichplot::dotplot(t_cf_clinical_gp[["outcome_up"]][["TF_enrich"]])

## Reactome of the up c/f genes
enrichplot::dotplot(t_cf_clinical_gp[["outcome_up"]][["REAC_enrich"]])

## GO of the down c/f genes
enrichplot::dotplot(t_cf_clinical_gp[["outcome_down"]][["GO_enrich"]])
t_cf_clinical_gp[["outcome_up"]][["pvalue_plots"]][["BP"]]

## Reactome of the down c/f genes
enrichplot::dotplot(t_cf_clinical_gp[["outcome_down"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_clinical_gp[["outcome_up"]][["GO_enrich"]])
```

# Visit comparisons

Later in this document I do a bunch of visit/cf comparisons.  In this
block I want to explicitly only compare v1 to other visits.  This is
something I did quite a lot in the 2019 datasets, but never actually
moved to this document.

```{r}
tv1_vs_later <- all_pairwise(t_v1vs, model_batch = "svaseq", filter = TRUE)
tv1_vs_later
tv1_vs_later_table <- combine_de_tables(
  tv1_vs_later, keepers = visit_v1later,
  excel = glue("{xlsx_prefix}/DE_Visits/tv1_vs_later_tables-v{ver}.xlsx"))
tv1_vs_later_table
tv1_vs_later_sig <- extract_significant_genes(
  tv1_vs_later_table,
  excel = glue("{xlsx_prefix}/DE_Visits/tv1_vs_later_sig-v{ver}.xlsx"))
tv1_vs_later_sig
```

#### GSEA: V1 vs other visits.

```{r}
tv1later_gp <- all_gprofiler(tv1_vs_later_sig)
tv1later_gp

tv1later_gp[[1]]$pvalue_plots$BP
tv1later_gp[[2]]$pvalue_plots$BP
```

# Sex comparison

```{r}
t_sex <- subset_expt(tc_sex, subset = "clinic == 'Tumaco'")
t_sex
t_sex_de <- all_pairwise(t_sex, model_batch = "svaseq", filter = TRUE)
t_sex_de
t_sex_table <- combine_de_tables(
  t_sex_de, excel = glue("{xlsx_prefix}/GSEA/t_sex_table-v{ver}.xlsx"))
t_sex_table
t_sex_sig <- extract_significant_genes(
  t_sex_table, excel = glue("{xlsx_prefix}/GSEA/t_sex_sig-v{ver}.xlsx"))
t_sex_sig
```

#### GSEA: Comparing Tumaco M/F

Let us see if we observe general male/female differences in the data.
This has an important caveat: there are few female failures in the
dataset and so the results may reflect that.

```{r}
t_sex_gp <- all_gprofiler(t_sex_sig)
t_sex_gp

written <- write_gprofiler_data(
  t_sex_gp[[1]],
  excel = glue("{xlsx_prefix}/GSEA/sex_female_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_sex_gp[[2]],
  excel = glue("{xlsx_prefix}/GSEA/sex_male_up-v{ver}.xlsx"))
```

```{r}
tc_sex_cure <- subset_expt(tc_sex, subset = "finaloutcome=='cure'")
t_sex_cure <- subset_expt(tc_sex_cure, subset = "clinic == 'Tumaco'")

t_sex_cure_de <- all_pairwise(t_sex_cure, model_batch = "svaseq", filter = TRUE)
t_sex_cure_de
t_sex_cure_table <- combine_de_tables(
  t_sex_cure_de, excel = glue("{xlsx_prefix}/DE_Sex/t_sex_cure_table-v{ver}.xlsx"))
t_sex_cure_table
t_sex_cure_sig <- extract_significant_genes(
  t_sex_cure_table, excel = glue("{xlsx_prefix}/DE_Sex/t_sex_cure_sig-v{ver}.xlsx"))
t_sex_cure_sig
```

##### GSEA: Tumaco comparisons of M/F

```{r}
t_sex_cure_gp <- all_gprofiler(t_sex_cure_sig)
t_sex_cure_gp
```

# Ethnicity comparisons

```{r}
t_ethnicity_de <- all_pairwise(t_etnia_expt, model_batch = "svaseq", filter = TRUE)
t_ethnicity_de
t_ethnicity_table <- combine_de_tables(
  keepers = ethnicity_contrasts,
  t_ethnicity_de, excel = glue("{xlsx_prefix}/DE_Ethnicity/t_ethnicity_table-v{ver}.xlsx"))
t_ethnicity_table
t_ethnicity_sig <- extract_significant_genes(
  t_ethnicity_table, excel = glue("{xlsx_prefix}/DE_Ethnicity/t_ethnicity_sig-v{ver}.xlsx"))
t_ethnicity_sig
```

#### GSEA: Ethnicity differences

Performed once with both clinics and again with only Tumaco.

```{r}
t_ethnicity_gp <- all_gprofiler(t_ethnicity_sig)
t_ethnicity_gp

written <- write_gprofiler_data(
  t_ethnicity_gp[[1]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_mi_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_ethnicity_gp[[2]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_mi_down-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_ethnicity_gp[[3]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_ma_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_ethnicity_gp[[4]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_ma_down-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_ethnicity_gp[[5]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_ia_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_ethnicity_gp[[6]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_ia_down-v{ver}.xlsx"))
```

#### GSEA: Ethnicity differences among cures

```{r}
written <- write_gprofiler_data(
  t_ethnicity_gp[["indigenous_afrocol_down"]],
  excel = glue("{xlsx_prefix}/GSEA/ethnicity_afrocol_down-v{ver}.xlsx"))
```

### Separate the Tumaco data by visit

One of the most compelling ideas in the data is the opportunity to
find genes in the first visit which may help predict the likelihood
that a person will respond well to treatment.  The following block
will therefore look at cure/fail from Tumaco at visit 1.

#### Cure/Fail, Tumaco Visit 1

```{r}
t_cf_clinical_v1_de_sva <- all_pairwise(tv1_samples, model_batch = "svaseq", filter = TRUE)
t_cf_clinical_v1_de_sva
t_cf_clinical_v1_table_sva <- combine_de_tables(
  t_cf_clinical_v1_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Visits/t_clinical_v1_cf_tables_sva-v{ver}.xlsx"))
t_cf_clinical_v1_table_sva
t_cf_clinical_v1_sig_sva <- extract_significant_genes(
  t_cf_clinical_v1_table_sva,
  excel = glue("{cf_prefix}/Visits/t_clinical_v1_cf_sig_sva-v{ver}.xlsx"))
t_cf_clinical_v1_sig_sva

dim(t_cf_clinical_v1_sig_sva$deseq$ups[[1]])
dim(t_cf_clinical_v1_sig_sva$deseq$downs[[1]])
```

##### GSEA: V1 gProfiler comparison

It looks like there are very few groups in the visit 1 significant genes.

```{r}
t_cf_clinical_v1_sig_sva_gp <- all_gprofiler(t_cf_clinical_v1_sig_sva)
t_cf_clinical_v1_sig_sva_gp

## Wikipathways of the up c/f genes
enrichplot::dotplot(t_cf_clinical_v1_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_clinical_v1_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

#### Cure/Fail, Tumaco Visit 2

The visit 2 and visit 3 samples are interesting because they provide
an opportunity to see if we can observe changes in response in the
middle and end of treatment...

```{r}
t_cf_clinical_v2_de_sva <- all_pairwise(tv2_samples, model_batch = "svaseq", filter = TRUE)
t_cf_clinical_v2_de_sva
t_cf_clinical_v2_table_sva <- combine_de_tables(
  t_cf_clinical_v2_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Visits/t_clinical_v2_cf_tables_sva-v{ver}.xlsx"))
t_cf_clinical_v2_table_sva
t_cf_clinical_v2_sig_sva <- extract_significant_genes(
  t_cf_clinical_v2_table_sva,
  excel = glue("{cf_prefix}/Visits/t_clinical_v2_cf_sig_sva-v{ver}.xlsx"))
t_cf_clinical_v2_sig_sva

dim(t_cf_clinical_v2_sig_sva$deseq$ups[[1]])
dim(t_cf_clinical_v2_sig_sva$deseq$downs[[1]])
```

##### GSEA: Visit 2 gProfiler searches

Up: 74 GO, 4 KEGG, 6 reactome, 4 WP, 56 TF, 1 miRNA, 0 HP/HPA/CORUM.
Down:  19 GO, 1 KEGG, 1 HP, 2 HPA, 0 reactome/wp/tf/corum

```{r}
t_cf_clinical_v2_sig_sva_gp <- all_gprofiler(t_cf_clinical_v2_sig_sva)
t_cf_clinical_v2_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_clinical_v2_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_clinical_v2_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_clinical_v2_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_clinical_v2_down-v{ver}.xlsx"))

## Wikipathways of the up c/f genes
enrichplot::dotplot(t_cf_clinical_v2_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_clinical_v2_sig_sva_gp[["outcome_up"]][["REAC_enrich"]])
enrichplot::dotplot(t_cf_clinical_v2_sig_sva_gp[["outcome_up"]][["TF_enrich"]])

enrichplot::dotplot(t_cf_clinical_v2_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

#### Cure/Fail, Tumaco Visit 3

```{r}
t_cf_clinical_v3_de_sva <- all_pairwise(tv3_samples, model_batch = "svaseq", filter = TRUE)
t_cf_clinical_v3_de_sva
t_cf_clinical_v3_table_sva <- combine_de_tables(
  t_cf_clinical_v3_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Visits/t_clinical_v3_cf_tables_sva-v{ver}.xlsx"))
t_cf_clinical_v3_table_sva
t_cf_clinical_v3_sig_sva <- extract_significant_genes(
  t_cf_clinical_v3_table_sva,
  excel = glue("{cf_prefix}/Visits/t_clinical_v3_cf_sig_sva-v{ver}.xlsx"))
t_cf_clinical_v3_sig_sva

dim(t_cf_clinical_v3_sig_sva$deseq$ups[[1]])
dim(t_cf_clinical_v3_sig_sva$deseq$downs[[1]])
```

##### GSEA: Visit 3 gProfiler searches

Up: 120 genes; 141 GO, 1 KEGG, 5 Reactome, 2 WP, 30 TF, 1 miRNA, 0 HPA/CORUM/HP
Down: 62 genes; 30 GO, 2 KEGG, 1 Reactome, 0 WP/TF/miRNA/HPA/CORUM/HP,

```{r}
t_cf_clinical_v3_sig_sva_gp <- all_gprofiler(t_cf_clinical_v3_sig_sva)
t_cf_clinical_v3_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_clinical_v3_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_clinical_v3_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_clinical_v3_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_clinical_v3_down-v{ver}.xlsx"))

## Wikipathways of the up c/f genes
enrichplot::dotplot(t_cf_clinical_v3_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_clinical_v3_sig_sva_gp[["outcome_up"]][["REAC_enrich"]])
enrichplot::dotplot(t_cf_clinical_v3_sig_sva_gp[["outcome_up"]][["TF_enrich"]])

enrichplot::dotplot(t_cf_clinical_v3_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

### Repeat no biopsies

The biopsy samples are problematic for a few reasons, so let us repeat
without them.

```{r}
t_cf_clinical_nobiop_de_sva <- all_pairwise(t_clinical_nobiop,
                                            model_batch = "svaseq", filter = TRUE)
t_cf_clinical_nobiop_de_sva
t_cf_clinical_nobiop_table_sva <- combine_de_tables(
  t_cf_clinical_nobiop_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/No_Biopsies/t_clinical_nobiop_cf_tables_sva-v{ver}.xlsx"))
t_cf_clinical_nobiop_table_sva
t_cf_clinical_nobiop_sig_sva <- extract_significant_genes(
  t_cf_clinical_nobiop_table_sva,
  excel = glue("{cf_prefix}/No_Biopsies/t_clinical_nobiop_cf_sig_sva-v{ver}.xlsx"))
t_cf_clinical_nobiop_sig_sva

dim(t_cf_clinical_nobiop_sig_sva$deseq$ups[[1]])
dim(t_cf_clinical_nobiop_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler of clinical samples without biopsies

Up: 137 genes; 88 GO, 0 KEGG, 6 Reactome, 1 WP, 46 TF, 1 miRNA, 0 others
Down: 73 genes; 78 GO, 1 KEGG, 1 Reactome, 9 TF, 0 others

```{r}
t_cf_clinical_nobiop_sig_sva_gp <- all_gprofiler(t_cf_clinical_nobiop_sig_sva)
t_cf_clinical_nobiop_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_clinical_nobiop_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_clinical_nobiop_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_clinical_nobiop_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_clinical_nobiop_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_clinical_nobiop_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_clinical_nobiop_sig_sva_gp[["outcome_up"]][["TF_enrich"]])

enrichplot::dotplot(t_cf_clinical_nobiop_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

### By cell type

Now let us switch our view to each individual cell type collected.
The hope here is that we will be able to learn some cell-specific
differences in the response for people who did(not) respond well.

#### Cure/Fail, Biopsies

```{r}
t_cf_biopsy_de_sva <- all_pairwise(t_biopsies, model_batch = "svaseq", filter = TRUE)
t_cf_biopsy_de_sva
t_cf_biopsy_table_sva <- combine_de_tables(
  t_cf_biopsy_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Biopsies/t_biopsy_cf_tables_sva-v{ver}.xlsx"))
t_cf_biopsy_table_sva
t_cf_biopsy_sig_sva <- extract_significant_genes(
  t_cf_biopsy_table_sva,
  excel = glue("{cf_prefix}/Biopsies/t_cf_biopsy_sig_sva-v{ver}.xlsx"))
t_cf_biopsy_sig_sva

dim(t_cf_biopsy_sig_sva$deseq$ups[[1]])
dim(t_cf_biopsy_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Biopsies

Up: 17 genes; 74 GO, 3 KEGG, 1 Reactome, 3 WP, 1 TF, 0 others
Down: 11 genes; 2 GO, 0 others

FIXME: Add write_gprofiler_data() to other stanzas.

```{r}
t_cf_biopsy_sig_sva_gp <- all_gprofiler(t_cf_biopsy_sig_sva)
t_cf_biopsy_sig_sva_gp

t_cf_biopsy_gp_up_written <- write_gprofiler_data(
  t_cf_biopsy_sig_sva_gp[[1]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_biopsy_sig_sva_up.xlsx"))
t_cf_biopsy_gp_down_written <- write_gprofiler_data(
  t_cf_biopsy_sig_sva_gp[[2]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_biopsy_sig_sva_down.xlsx"))

enrichplot::dotplot(t_cf_biopsy_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_biopsy_sig_sva_gp[["outcome_up"]][["WP_enrich"]])

enrichplot::dotplot(t_cf_biopsy_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

#### Cure/Fail, Monocytes

Same question, but this time looking at monocytes.  In addition, this
comparison was done twice, once using SVA and once using visit as a
batch factor.

```{r}
t_cf_monocyte_de_sva <- all_pairwise(t_monocytes, model_batch = "svaseq",
                                     filter = TRUE)
t_cf_monocyte_de_sva
t_cf_monocyte_tables_sva <- combine_de_tables(
  t_cf_monocyte_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_cf_tables_sva-v{ver}.xlsx"))
t_cf_monocyte_tables_sva
t_cf_monocyte_sig_sva <- extract_significant_genes(
  t_cf_monocyte_tables_sva,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_cf_sig_sva-v{ver}.xlsx"))
t_cf_monocyte_sig_sva

dim(t_cf_monocyte_sig_sva$deseq$ups[[1]])
dim(t_cf_monocyte_sig_sva$deseq$downs[[1]])

t_cf_monocyte_de_batchvisit <- all_pairwise(t_monocytes, model_batch = TRUE, filter = TRUE)
t_cf_monocyte_de_batchvisit
t_cf_monocyte_tables_batchvisit <- combine_de_tables(
  t_cf_monocyte_de_batchvisit, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_cf_tables_batchvisit-v{ver}.xlsx"))
t_cf_monocyte_tables_batchvisit
t_cf_monocyte_sig_batchvisit <- extract_significant_genes(
  t_cf_monocyte_tables_batchvisit,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_cf_sig_batchvisit-v{ver}.xlsx"))
t_cf_monocyte_sig_batchvisit

dim(t_cf_monocyte_sig_batchvisit$deseq$ups[[1]])
dim(t_cf_monocyte_sig_batchvisit$deseq$downs[[1]])
```

##### GSEA: gProfiler Monocytes

Now that I am looking back over these results, I am not compeltely
certain why I only did the gprofiler search for the sva data...

Up: 60 genes; 12 GO, 1 KEGG, 1 WP, 4 TF, 0 others
Down: 53 genes; 26 GO, 1 KEGG, 1 Reactome, 2 TF, 0 others

```{r}
t_cf_monocyte_sig_sva_gp <- all_gprofiler(t_cf_monocyte_sig_sva)
t_cf_monocyte_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_monocyte_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_monocyte_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_monocyte_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_monocyte_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_monocyte_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_monocyte_sig_sva_gp[["outcome_up"]][["TF_enrich"]])

enrichplot::dotplot(t_cf_monocyte_sig_sva_gp[["outcome_down"]][["GO_enrich"]])

t_cf_monocyte_sig_batch_gp <- all_gprofiler(t_cf_monocyte_sig_batchvisit)
t_cf_monocyte_sig_batch_gp

written <- write_gprofiler_data(
  t_cf_monocyte_sig_batch_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_monocyte_batch_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_monocyte_sig_batch_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_monocyte_batch_down-v{ver}.xlsx"))
enrichplot::dotplot(t_cf_monocyte_sig_batch_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_monocyte_sig_batch_gp[["outcome_up"]][["HP_enrich"]])
```

### Individual visits, Monocytes

Now focus in on the monocyte samples on a per-visit basis.

#### Visit 1

```{r}
t_cf_monocyte_v1_de_sva <- all_pairwise(tv1_monocytes, model_batch = "svaseq", filter = TRUE)
t_cf_monocyte_v1_de_sva
t_cf_monocyte_v1_tables_sva <- combine_de_tables(
  t_cf_monocyte_v1_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_v1_cf_tables_sva-v{ver}.xlsx"))
t_cf_monocyte_v1_tables_sva
t_cf_monocyte_v1_sig_sva <- extract_significant_genes(
  t_cf_monocyte_v1_tables_sva,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_v1_cf_sig_sva-v{ver}.xlsx"))
t_cf_monocyte_v1_sig_sva

dim(t_cf_monocyte_v1_sig_sva$deseq$ups[[1]])
dim(t_cf_monocyte_v1_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Monocytes by visit, V1

V1: Up: 14 genes; No categories
V1: Down: 52 genes; 20 GO, 5 TF

```{r}
t_cf_monocyte_v1_sig_sva_gp <- all_gprofiler(t_cf_monocyte_v1_sig_sva)
t_cf_monocyte_v1_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_monocyte_v1_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_monocyte_v1_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_monocyte_v1_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_monocyte_v1_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_monocyte_v1_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

#### Visit 2

```{r}
t_cf_monocyte_v2_de_sva <- all_pairwise(tv2_monocytes, model_batch = "svaseq", filter = TRUE)
t_cf_monocyte_v2_de_sva
t_cf_monocyte_v2_tables_sva <- combine_de_tables(
  t_cf_monocyte_v2_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_v2_cf_tables_sva-v{ver}.xlsx"))
t_cf_monocyte_v2_tables_sva
t_cf_monocyte_v2_sig_sva <- extract_significant_genes(
  t_cf_monocyte_v2_tables_sva,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_v2_cf_sig_sva-v{ver}.xlsx"))
t_cf_monocyte_v2_sig_sva

dim(t_cf_monocyte_v2_sig_sva$deseq$ups[[1]])
dim(t_cf_monocyte_v2_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Monocytes by visit V1

Insufficient data to use in gProfiler.

V2: Up: 1 gene
V2: Down: 0 genes.


#### Visit 3

```{r}
t_cf_monocyte_v3_de_sva <- all_pairwise(tv3_monocytes, model_batch = "svaseq", filter = TRUE)
t_cf_monocyte_v3_de_sva
t_cf_monocyte_v3_tables_sva <- combine_de_tables(
  t_cf_monocyte_v3_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_v3_cf_tables_sva-v{ver}.xlsx"))
t_cf_monocyte_v3_tables_sva
t_cf_monocyte_v3_sig_sva <- extract_significant_genes(
  t_cf_monocyte_v3_tables_sva,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_v3_cf_sig_sva-v{ver}.xlsx"))
t_cf_monocyte_v3_sig_sva

dim(t_cf_monocyte_v3_sig_sva$deseq$ups[[1]])
dim(t_cf_monocyte_v3_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Monocytes by visit, V3

V3: Up: 4 genes.
V3: Down: 0 genes.

#### Monocytes: Compare sva to batch-in-model

```{r}
sva_aucc <- calculate_aucc(t_cf_monocyte_tables_sva[["data"]][[1]],
                           tbl2 = t_cf_monocyte_tables_batchvisit[["data"]][[1]],
                           py = "deseq_adjp", ly = "deseq_logfc")
sva_aucc

shared_ids <- rownames(t_cf_monocyte_tables_sva[["data"]][[1]]) %in%
  rownames(t_cf_monocyte_tables_batchvisit[["data"]][[1]])
first <- t_cf_monocyte_tables_sva[["data"]][[1]][shared_ids, ]
second <- t_cf_monocyte_tables_batchvisit[["data"]][[1]][rownames(first), ]
cor.test(first[["deseq_logfc"]], second[["deseq_logfc"]])
```

### Neutrophil samples

Switch context to the Neutrophils, once again repeat the analysis
using SVA and visit as a batch factor.

```{r}
t_cf_neutrophil_de_sva <- all_pairwise(t_neutrophils, model_batch = "svaseq", filter = TRUE)
t_cf_neutrophil_de_sva
t_cf_neutrophil_tables_sva <- combine_de_tables(
  t_cf_neutrophil_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_cf_tables_sva-v{ver}.xlsx"))
t_cf_neutrophil_tables_sva
t_cf_neutrophil_sig_sva <- extract_significant_genes(
  t_cf_neutrophil_tables_sva,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_cf_sig_sva-v{ver}.xlsx"))
t_cf_neutrophil_sig_sva

dim(t_cf_neutrophil_sig_sva$deseq$ups[[1]])
dim(t_cf_neutrophil_sig_sva$deseq$downs[[1]])

t_cf_neutrophil_de_batchvisit <- all_pairwise(t_neutrophils, model_batch = TRUE, filter = TRUE)
t_cf_neutrophil_de_batchvisit
t_cf_neutrophil_tables_batchvisit <- combine_de_tables(
  t_cf_neutrophil_de_batchvisit, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_cf_tables_batchvisit-v{ver}.xlsx"))
t_cf_neutrophil_tables_batchvisit
t_cf_neutrophil_sig_batchvisit <- extract_significant_genes(
  t_cf_neutrophil_tables_batchvisit,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_cf_sig_batchvisit-v{ver}.xlsx"))
t_cf_neutrophil_sig_batchvisit

dim(t_cf_neutrophil_sig_batchvisit$deseq$ups[[1]])
dim(t_cf_neutrophil_sig_batchvisit$deseq$downs[[1]])
```

##### GSEA: gProfiler Neutrophils

Up: 84 genes; 5 GO, 2 Reactome, 3 TF, no others.
Down: 29 genes: 12 GO, 1 Reactome, 1 TF, 1 miRNA, 11 HP, 0 others

```{r}
t_cf_neutrophil_sig_sva_gp <- all_gprofiler(t_cf_neutrophil_sig_sva)
t_cf_neutrophil_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_neutrophil_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_neutrophil_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_neutrophil_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_neutrophil_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_neutrophil_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_neutrophil_sig_sva_gp[["outcome_up"]][["TF_enrich"]])

enrichplot::dotplot(t_cf_neutrophil_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
## enrichplot::dotplot(t_cf_neutrophil_sig_sva_gp[["outcome_down"]][["HP_enrich"]])
```

#### Neutrophils by visit

When I did this with the monocytes, I split it up into multiple blocks
for each visit.  This time I am just going to run them all together.

```{r}
visitcf_factor <- paste0("v", pData(t_neutrophils)[["visitnumber"]],
                         pData(t_neutrophils)[["finaloutcome"]])
t_neutrophil_visitcf <- set_expt_conditions(t_neutrophils, fact=visitcf_factor)

t_cf_neutrophil_visits_de_sva <- all_pairwise(t_neutrophil_visitcf, model_batch = "svaseq",
                                              filter = TRUE)
t_cf_neutrophil_visits_de_sva
t_cf_neutrophil_visits_tables_sva <- combine_de_tables(
  t_cf_neutrophil_visits_de_sva, keepers = visitcf_contrasts,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_visitcf_tables_sva-v{ver}.xlsx"))
t_cf_neutrophil_visits_tables_sva
t_cf_neutrophil_visits_sig_sva <- extract_significant_genes(
  t_cf_neutrophil_visits_tables_sva,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_visitcf_sig_sva-v{ver}.xlsx"))
t_cf_neutrophil_visits_sig_sva

dim(t_cf_neutrophil_visits_sig_sva$deseq$ups[[1]])
dim(t_cf_neutrophil_visits_sig_sva$deseq$downs[[1]])

t_cf_neutrophil_v1_de_sva <- all_pairwise(tv1_neutrophils, model_batch = "svaseq", filter = TRUE)
t_cf_neutrophil_v1_de_sva
t_cf_neutrophil_v1_tables_sva <- combine_de_tables(
  t_cf_neutrophil_v1_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_v1_cf_tables_sva-v{ver}.xlsx"))
t_cf_neutrophil_v1_tables_sva
t_cf_neutrophil_v1_sig_sva <- extract_significant_genes(
  t_cf_neutrophil_v1_tables_sva,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_v1_cf_sig_sva-v{ver}.xlsx"))
t_cf_neutrophil_v1_sig_sva

dim(t_cf_neutrophil_v1_sig_sva$deseq$ups[[1]])
dim(t_cf_neutrophil_v1_sig_sva$deseq$downs[[1]])

t_cf_neutrophil_v2_de_sva <- all_pairwise(tv2_neutrophils, model_batch = "svaseq", filter = TRUE)
t_cf_neutrophil_v2_de_sva
t_cf_neutrophil_v2_tables_sva <- combine_de_tables(
  t_cf_neutrophil_v2_de_sva,
  keepers = cf_contrast,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_v2_cf_tables_sva-v{ver}.xlsx"))
t_cf_neutrophil_v2_tables_sva
t_cf_neutrophil_v2_sig_sva <- extract_significant_genes(
  t_cf_neutrophil_v2_tables_sva,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_v2_cf_sig_sva-v{ver}.xlsx"))
t_cf_neutrophil_v2_sig_sva

dim(t_cf_neutrophil_v2_sig_sva$deseq$ups[[1]])
dim(t_cf_neutrophil_v2_sig_sva$deseq$downs[[1]])

t_cf_neutrophil_v3_de_sva <- all_pairwise(tv3_neutrophils, model_batch = "svaseq", filter = TRUE)
t_cf_neutrophil_v3_de_sva
t_cf_neutrophil_v3_tables_sva <- combine_de_tables(
  t_cf_neutrophil_v3_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_v3_cf_tables_sva-v{ver}.xlsx"))
t_cf_neutrophil_v3_tables_sva
t_cf_neutrophil_v3_sig_sva <- extract_significant_genes(
  t_cf_neutrophil_v3_tables_sva,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_v3_cf_sig_sva-v{ver}.xlsx"))
t_cf_neutrophil_v3_sig_sva

dim(t_cf_neutrophil_v3_sig_sva$deseq$ups[[1]])
dim(t_cf_monocyte_v3_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Neutrophils V1

V1: Up: 5 genes
V1: Down: 8 genes; 14 GO.

```{r}
t_cf_neutrophil_v1_sig_sva_gp <- all_gprofiler(t_cf_neutrophil_v1_sig_sva)
t_cf_neutrophil_v1_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_neutrophil_v1_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_neutrophil_v1_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_neutrophil_v1_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_neutrophil_v1_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_neutrophil_v1_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
```

##### GSEA: gProfiler Neutrophils by visit, V2

Up: 5 genes; 3 GO, 10 TF.
Down: 1 gene.


#### Neutrophils: Compare sva to batch-in-model

```{r}
sva_aucc <- calculate_aucc(t_cf_neutrophil_tables_sva[["data"]][[1]],
                           tbl2 = t_cf_neutrophil_tables_batchvisit[["data"]][[1]],
                           py = "deseq_adjp", ly = "deseq_logfc")
sva_aucc

shared_ids <- rownames(t_cf_neutrophil_tables_sva[["data"]][[1]]) %in%
  rownames(t_cf_neutrophil_tables_batchvisit[["data"]][[1]])
first <- t_cf_neutrophil_tables_sva[["data"]][[1]][shared_ids, ]
second <- t_cf_neutrophil_tables_batchvisit[["data"]][[1]][rownames(first), ]
cor.test(first[["deseq_logfc"]], second[["deseq_logfc"]])
```

### Eosinophils

This time, with feeling!  Repeating the same set of tasks with the
eosinophil samples.

```{r}
t_cf_eosinophil_de_sva <- all_pairwise(t_eosinophils, model_batch = "svaseq", filter = TRUE)
t_cf_eosinophil_de_sva
t_cf_eosinophil_tables_sva <- combine_de_tables(
  t_cf_eosinophil_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_cf_tables_sva-v{ver}.xlsx"))
t_cf_eosinophil_tables_sva
t_cf_eosinophil_sig_sva <- extract_significant_genes(
  t_cf_eosinophil_tables_sva,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_cf_sig_sva-v{ver}.xlsx"))
t_cf_eosinophil_sig_sva

dim(t_cf_eosinophil_sig_sva$deseq$ups[[1]])
dim(t_cf_eosinophil_sig_sva$deseq$downs[[1]])

t_cf_eosinophil_de_batchvisit <- all_pairwise(t_eosinophils, model_batch = TRUE, filter = TRUE)
t_cf_eosinophil_de_batchvisit
t_cf_eosinophil_tables_batchvisit <- combine_de_tables(
  t_cf_eosinophil_de_batchvisit, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_cf_tables_batchvisit-v{ver}.xlsx"))
t_cf_eosinophil_tables_batchvisit
t_cf_eosinophil_sig_batchvisit <- extract_significant_genes(
  t_cf_eosinophil_tables_batchvisit,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_cf_sig_batchvisit-v{ver}.xlsx"))
t_cf_eosinophil_sig_batchvisit

dim(t_cf_eosinophil_sig_batchvisit$deseq$ups[[1]])
dim(t_cf_eosinophil_sig_batchvisit$deseq$downs[[1]])

visitcf_factor <- paste0("v", pData(t_eosinophils)[["visitnumber"]],
                         pData(t_eosinophils)[["finaloutcome"]])
t_eosinophil_visitcf <- set_expt_conditions(t_eosinophils, fact = visitcf_factor)

t_cf_eosinophil_visits_de_sva <- all_pairwise(t_eosinophil_visitcf, model_batch = "svaseq",
                                              filter = TRUE)
t_cf_eosinophil_visits_de_sva
t_cf_eosinophil_visits_tables_sva <- combine_de_tables(
  t_cf_eosinophil_visits_de_sva, keepers = visitcf_contrasts,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_visitcf_tables_sva-v{ver}.xlsx"))
t_cf_eosinophil_visits_tables_sva
t_cf_eosinophil_visits_sig_sva <- extract_significant_genes(
  t_cf_eosinophil_visits_tables_sva,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_visitcf_sig_sva-v{ver}.xlsx"))
t_cf_eosinophil_visits_sig_sva

dim(t_cf_eosinophil_visits_sig_sva$deseq$ups[[1]])
dim(t_cf_eosinophil_visits_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Eosinophils

Up: 116 genes; 123 GO, 2 KEGG, 7 Reactome, 5 WP, 69 TF, 1 miRNA, 0 others
Down:  74 genes; 5 GO, 1 Reactome, 4 TF, 0 others

```{r}
t_cf_eosinophil_sig_sva_gp <- all_gprofiler(t_cf_eosinophil_sig_sva)
t_cf_eosinophil_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_eosinophil_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_eosinophil_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["REAC_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["WP_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["TF_enrich"]])

enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_down"]][["TF_enrich"]])
```

#### C/F celltype volcano plots with specific labels

### Figure 4B: Volcano plots

Note to self, when I rendered the html, stupid R ran out of temp files
and so did not actually print the darn html document, as a result I
modified the render function to try to make sure there is a clean
directory in which to work; testing now.  If it continues to not work,
I will need to remove some of the images created in this document.

```{r}
num_color <- color_choices[["clinic_cf"]][["Tumaco_failure"]]
den_color <- color_choices[["clinic_cf"]][["Tumaco_cure"]]
wanted_genes <- c("FI44L", "IFI27", "PRR5", "PRR5-ARHGAP8", "RHCE",
                  "FBXO39", "RSAD2", "SMTNL1", "USP18", "AFAP1")

cf_monocyte_table <- t_cf_monocyte_tables_sva[["data"]][["outcome"]]
cf_monocyte_volcano <- plot_volcano_condition_de(
  cf_monocyte_table, "outcome", label = wanted_genes,
  fc_col = "deseq_logfc", p_col = "deseq_adjp", line_position = NULL,
  color_high = num_color, color_low = den_color, label_size = 6)
pp(file = glue("images/cf_monocyte_volcano_labeled-v{ver}.svg"))
cf_monocyte_volcano$plot
dev.off()
cf_monocyte_volcano$plot

cf_eosinophil_table <- t_cf_eosinophil_tables_sva[["data"]][["outcome"]]
cf_eosinophil_volcano <- plot_volcano_condition_de(
  cf_eosinophil_table, "outcome", label = wanted_genes,
  fc_col = "deseq_logfc", p_col = "deseq_adjp", line_position = NULL,
  color_high = num_color, color_low = den_color, label_size = 6)
pp(file = glue("images/cf_eosinophil_volcano_labeled-v{ver}.svg"))
cf_eosinophil_volcano$plot
dev.off()
cf_eosinophil_volcano$plot

cf_neutrophil_table <- t_cf_neutrophil_tables_sva[["data"]][["outcome"]]
cf_neutrophil_volcano <- plot_volcano_condition_de(
  cf_neutrophil_table, "outcome", label = wanted_genes,
  fc_col = "deseq_logfc", p_col = "deseq_adjp", line_position = NULL,
  color_high = num_color, color_low = den_color, label_size = 6)
pp(file = glue("images/cf_neutrophil_volcano_labeled-v{ver}.svg"))
cf_neutrophil_volcano$plot
dev.off()
cf_neutrophil_volcano$plot
```

### Eosinophil time comparisons

```{r}
t_cf_eosinophil_v1_de_sva <- all_pairwise(tv1_eosinophils, model_batch = "svaseq", filter = TRUE)
t_cf_eosinophil_v1_de_sva
t_cf_eosinophil_v1_tables_sva <- combine_de_tables(
  t_cf_eosinophil_v1_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_v1_cf_tables_sva-v{ver}.xlsx"))
t_cf_eosinophil_v1_tables_sva
t_cf_eosinophil_v1_sig_sva <- extract_significant_genes(
  t_cf_eosinophil_v1_tables_sva,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_v1_cf_sig_sva-v{ver}.xlsx"))
t_cf_eosinophil_v1_tables_sva

dim(t_cf_eosinophil_v1_sig_sva$deseq$ups[[1]])
dim(t_cf_eosinophil_v1_sig_sva$deseq$downs[[1]])

t_cf_eosinophil_v2_de_sva <- all_pairwise(tv2_eosinophils, model_batch = "svaseq", filter = TRUE)
t_cf_eosinophil_v2_de_sva
t_cf_eosinophil_v2_tables_sva <- combine_de_tables(
  t_cf_eosinophil_v2_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_v2_cf_tables_sva-v{ver}.xlsx"))
t_cf_eosinophil_v2_tables_sva
t_cf_eosinophil_v2_sig_sva <- extract_significant_genes(
  t_cf_eosinophil_v2_tables_sva,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_v2_cf_sig_sva-v{ver}.xlsx"))
t_cf_eosinophil_v2_sig_sva

dim(t_cf_eosinophil_v2_sig_sva$deseq$ups[[1]])
dim(t_cf_eosinophil_v2_sig_sva$deseq$downs[[1]])

t_cf_eosinophil_v3_de_sva <- all_pairwise(tv3_eosinophils, model_batch = "svaseq", filter = TRUE)
t_cf_eosinophil_v3_de_sva
t_cf_eosinophil_v3_tables_sva <- combine_de_tables(
  t_cf_eosinophil_v3_de_sva, keepers = cf_contrast,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_v3_cf_tables_sva-v{ver}.xlsx"))
t_cf_eosinophil_v3_tables_sva
t_cf_eosinophil_v3_sig_sva <- extract_significant_genes(
  t_cf_eosinophil_v3_tables_sva,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_v3_cf_sig_sva-v{ver}.xlsx"))
t_cf_eosinophil_v3_sig_sva

dim(t_cf_eosinophil_v3_sig_sva$deseq$ups[[1]])
dim(t_cf_eosinophil_v3_sig_sva$deseq$downs[[1]])
```

##### GSEA: gProfiler Eosinophils V1

Up: 13 genes, no hits.
Down: 19 genes; 11 GO, 1 Reactome, 1 TF

```{r}
t_cf_eosinophil_v1_sig_sva_gp <- all_gprofiler(t_cf_eosinophil_v1_sig_sva)
t_cf_eosinophil_v1_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_eosinophil_v1_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_v1_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_eosinophil_v1_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_v1_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_down"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_down"]][["TF_enrich"]])
```

##### GSEA: gProfiler Eosinophils V2

Up: 9 genes; 23 GO, 2 KEGG, 2 Reactome, 4 WP
Down: 4 genes; no hits

```{r}
t_cf_eosinophil_v2_sig_sva_gp <- all_gprofiler(t_cf_eosinophil_v2_sig_sva)
t_cf_eosinophil_v2_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_eosinophil_v2_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_v2_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_eosinophil_v2_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_v2_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["WP_enrich"]])
```

#### gProfiler: Eosinophils V3

Up: 68 genes; 95 GO, 2 KEGG, 12 Reactome, 3 WP, 63 TF, 1 miRNA
Down: 29 genes; 3 GO, 1 WP, 1 TF, 3 miRNA

```{r}
t_cf_eosinophil_v3_sig_sva_gp <- all_gprofiler(t_cf_eosinophil_v3_sig_sva)
t_cf_eosinophil_v3_sig_sva_gp

written <- write_gprofiler_data(
  t_cf_eosinophil_v3_sig_sva_gp[["outcome_up"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_v3_up-v{ver}.xlsx"))
written <- write_gprofiler_data(
  t_cf_eosinophil_v3_sig_sva_gp[["outcome_down"]],
  excel = glue("{xlsx_prefix}/GSEA/t_cf_eosinophil_v3_down-v{ver}.xlsx"))

enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["GO_enrich"]])
enrichplot::dotplot(t_cf_eosinophil_sig_sva_gp[["outcome_up"]][["WP_enrich"]])
```

#### Eosinophils: Compare sva to batch-in-visit

```{r}
sva_aucc <- calculate_aucc(t_cf_eosinophil_tables_sva[["data"]][[1]],
                           tbl2 = t_cf_eosinophil_tables_batchvisit[["data"]][[1]],
                           py = "deseq_adjp", ly = "deseq_logfc")
sva_aucc

shared_ids <- rownames(t_cf_eosinophil_tables_sva[["data"]][[1]]) %in%
  rownames(t_cf_eosinophil_tables_batchvisit[["data"]][[1]])
first <- t_cf_eosinophil_tables_sva[["data"]][[1]][shared_ids, ]
second <- t_cf_eosinophil_tables_batchvisit[["data"]][[1]][rownames(first), ]
cor.test(first[["deseq_logfc"]], second[["deseq_logfc"]])
```

#### Compare monocyte CF, neutrophil CF, eosinophil CF

```{r}
t_mono_neut_sva_aucc <- calculate_aucc(t_cf_monocyte_tables_sva[["data"]][["outcome"]],
                                       tbl2 = t_cf_neutrophil_tables_sva[["data"]][["outcome"]],
                                       py = "deseq_adjp", ly = "deseq_logfc")
t_mono_neut_sva_aucc

t_mono_eo_sva_aucc <- calculate_aucc(t_cf_monocyte_tables_sva[["data"]][["outcome"]],
                                     tbl2 = t_cf_eosinophil_tables_sva[["data"]][["outcome"]],
                                     py = "deseq_adjp", ly = "deseq_logfc")
t_mono_eo_sva_aucc

t_neut_eo_sva_aucc <- calculate_aucc(t_cf_neutrophil_tables_sva[["data"]][["outcome"]],
                                     tbl2 = t_cf_eosinophil_tables_sva[["data"]][["outcome"]],
                                     py = "deseq_adjp", ly = "deseq_logfc")
t_neut_eo_sva_aucc
```

### By visit

For these contrasts, we want to see fail_v1 vs. cure_v1, fail_v2
vs. cure_v2 etc.  As a result, we will need to juggle the data
slightly and add another set of contrasts.

#### Cure/Fail by visits, all cell types

```{r}
t_visit_cf_all_de_sva <- all_pairwise(t_visitcf, model_batch = "svaseq", filter = TRUE)
t_visit_cf_all_de_sva
t_visit_cf_all_tables_sva <- combine_de_tables(
  t_visit_cf_all_de_sva, keepers = visitcf_contrasts,
  excel = glue("{cf_prefix}/t_all_visitcf_tables_sva-v{ver}.xlsx"))
t_visit_cf_all_tables_sva
t_visit_cf_all_sig_sva <- extract_significant_genes(
  t_visit_cf_all_tables_sva,
  excel = glue("{cf_prefix}/t_all_visitcf_sig_sva-v{ver}.xlsx"))
t_visit_cf_all_sig_sva
```

##### GSEA: Tumaco all samples C/F

I am adding every gprofiler call to this document, but I have a strong
feeling that this one is redundant.

```{r}
t_visit_cf_all_gp <- all_gprofiler(t_visit_cf_all_sig_sva)
t_visit_cf_all_gp
```

#### Cure/Fail by visit, Monocytes

```{r}
visitcf_factor <- paste0("v", pData(t_monocytes)[["visitnumber"]], "_",
                         pData(t_monocytes)[["finaloutcome"]])
t_monocytes_visitcf <- set_expt_conditions(t_monocytes, fact = visitcf_factor)

t_visit_cf_monocyte_de_sva <- all_pairwise(t_monocytes_visitcf, model_batch = "svaseq",
                                           filter = TRUE)
t_visit_cf_monocyte_de_sva
t_visit_cf_monocyte_tables_sva <- combine_de_tables(
  t_visit_cf_monocyte_de_sva, keepers = visitcf_contrasts,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_visitcf_tables_sva-v{ver}.xlsx"))
t_visit_cf_monocyte_tables_sva
t_visit_cf_monocyte_sig_sva <- extract_significant_genes(
  t_visit_cf_monocyte_tables_sva,
  excel = glue("{cf_prefix}/Monocytes/t_monocyte_visitcf_sig_sva-v{ver}.xlsx"))
t_visit_cf_monocyte_sig_sva

t_v1fc_deseq_ma <- t_visit_cf_monocyte_tables_sva[["plots"]][["v1cf"]][["deseq_ma_plots"]][["plot"]]
dev <- pp(file = "images/monocyte_cf_de_v1_maplot.png")
t_v1fc_deseq_ma
closed <- dev.off()
t_v1fc_deseq_ma

t_v2fc_deseq_ma <- t_visit_cf_monocyte_tables_sva[["plots"]][["v2cf"]][["deseq_ma_plots"]][["plot"]]
dev <- pp(file = "images/monocyte_cf_de_v2_maplot.png")
t_v2fc_deseq_ma
closed <- dev.off()
t_v2fc_deseq_ma

t_v3fc_deseq_ma <- t_visit_cf_monocyte_tables_sva[["plots"]][["v3cf"]][["deseq_ma_plots"]][["plot"]]
dev <- pp(file = "images/monocyte_cf_de_v3_maplot.png")
t_v3fc_deseq_ma
closed <- dev.off()
t_v3fc_deseq_ma
```

One query from Alejandro is to look at the genes shared up/down across
visits.  I am not entirely certain we have enough samples for this to
work, but let us find out.

I am thinking this is a good place to use the AUCC curves I learned
about thanks to Julie Cridland.

Note that the following is all monocyte samples, this should therefore
potentially be moved up and a version of this with only the Tumaco
samples put here?

```{r}
v1cf <- t_visit_cf_monocyte_tables_sva[["data"]][["v1cf"]]
v2cf <- t_visit_cf_monocyte_tables_sva[["data"]][["v2cf"]]
v3cf <- t_visit_cf_monocyte_tables_sva[["data"]][["v3cf"]]

v1_sig <- c(
  rownames(t_visit_cf_monocyte_sig_sva[["deseq"]][["ups"]][["v1cf"]]),
  rownames(t_visit_cf_monocyte_sig_sva[["deseq"]][["downs"]][["v1cf"]]))
length(v1_sig)

v2_sig <- c(
  rownames(t_visit_cf_monocyte_sig_sva[["deseq"]][["ups"]][["v2cf"]]),
  rownames(t_visit_cf_monocyte_sig_sva[["deseq"]][["downs"]][["v2cf"]]))
length(v2_sig)

v3_sig <- c(
  rownames(t_visit_cf_monocyte_sig_sva[["deseq"]][["ups"]][["v2cf"]]),
  rownames(t_visit_cf_monocyte_sig_sva[["deseq"]][["downs"]][["v2cf"]]))
length(v3_sig)

t_monocyte_visit_aucc_v2v1 <- calculate_aucc(v1cf, tbl2 = v2cf,
                                             py = "deseq_adjp", ly = "deseq_logfc")
dev <- pp(file = "images/monocyte_visit_v2v1_aucc.png")
t_monocyte_visit_aucc_v2v1[["plot"]]
closed <- dev.off()
t_monocyte_visit_aucc_v2v1[["plot"]]

t_monocyte_visit_aucc_v3v1 <- calculate_aucc(v1cf, tbl2 = v3cf,
                                             py = "deseq_adjp", ly = "deseq_logfc")
dev <- pp(file = "images/monocyte_visit_v3v1_aucc.png")
t_monocyte_visit_aucc_v3v1[["plot"]]
closed <- dev.off()
t_monocyte_visit_aucc_v3v1[["plot"]]
```

#### Cure/Fail by visit, Neutrophils

```{r}
visitcf_factor <- paste0("v", pData(t_neutrophils)[["visitnumber"]], "_",
                         pData(t_neutrophils)[["finaloutcome"]])
t_neutrophil_visitcf <- set_expt_conditions(t_neutrophils, fact = visitcf_factor)

t_visit_cf_neutrophil_de_sva <- all_pairwise(t_neutrophil_visitcf, model_batch = "svaseq",
                                             filter = TRUE)
t_visit_cf_neutrophil_de_sva
t_visit_cf_neutrophil_tables_sva <- combine_de_tables(
  t_visit_cf_neutrophil_de_sva, keepers = visitcf_contrasts,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_visitcf_tables_sva-v{ver}.xlsx"))
t_visit_cf_neutrophil_tables_sva
t_visit_cf_neutrophil_sig_sva <- extract_significant_genes(
  t_visit_cf_neutrophil_tables_sva,
  excel = glue("{cf_prefix}/Neutrophils/t_neutrophil_visitcf_sig_sva-v{ver}.xlsx"))
t_visit_cf_neutrophil_sig_sva
```

#### Cure/Fail by visit, Eosinophils

```{r}
visitcf_factor <- paste0("v", pData(t_eosinophils)[["visitnumber"]], "_",
                         pData(t_eosinophils)[["finaloutcome"]])
t_eosinophil_visitcf <- set_expt_conditions(t_eosinophils, fact = visitcf_factor)

t_visit_cf_eosinophil_de_sva <- all_pairwise(t_eosinophil_visitcf, model_batch = "svaseq",
                                             filter = TRUE)
t_visit_cf_eosinophil_de_sva
t_visit_cf_eosinophil_tables_sva <- combine_de_tables(
  t_visit_cf_eosinophil_de_sva, keepers = visitcf_contrasts,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_visitcf_tables_sva-v{ver}.xlsx"))
t_visit_cf_eosinophil_tables_sva
t_visit_cf_eosinophil_sig_sva <- extract_significant_genes(
  t_visit_cf_eosinophil_tables_sva,
  excel = glue("{cf_prefix}/Eosinophils/t_eosinophil_visitcf_sig_sva-v{ver}.xlsx"))
t_visit_cf_eosinophil_sig_sva
```

## Persistence in visit 3

Having put some SL read mapping information in the sample sheet, Maria
Adelaida added a new column using it with the putative persistence
state on a per-sample basis.  One question which arised from that:
what differences are observable between the persistent yes vs. no
samples on a per-cell-type basis among the visit 3 samples.

### Setting up

First things first, create the datasets.

```{r}
persistence_expt <- subset_expt(t_clinical, subset = "persistence=='Y'|persistence=='N'") %>%
  subset_expt(subset = 'visitnumber==3') %>%
  set_expt_conditions(fact = 'persistence')

## persistence_biopsy <- subset_expt(persistence_expt, subset = "typeofcells=='biopsy'")
persistence_monocyte <- subset_expt(persistence_expt, subset = "typeofcells=='monocytes'")
persistence_neutrophil <- subset_expt(persistence_expt, subset = "typeofcells=='neutrophils'")
persistence_eosinophil <- subset_expt(persistence_expt, subset = "typeofcells=='eosinophils'")
```

### Take a look

See if there are any patterns which look usable.

```{r}
## All
persistence_norm <- normalize_expt(persistence_expt, transform = "log2", convert = "cpm",
                                   norm = "quant", filter = TRUE)
plot_pca(persistence_norm)$plot
persistence_nb <- normalize_expt(persistence_expt, transform = "log2", convert = "cpm",
                                 batch = "svaseq", filter = TRUE)
plot_pca(persistence_nb)$plot

## Biopsies
##persistence_biopsy_norm <- normalize_expt(persistence_biopsy, transform = "log2", convert = "cpm",
##                                   norm = "quant", filter = TRUE)
##plot_pca(persistence_biopsy_norm)$plot
## Insufficient data

## Monocytes
persistence_monocyte_norm <- normalize_expt(persistence_monocyte, transform = "log2", convert = "cpm",
                                            norm = "quant", filter = TRUE)
plot_pca(persistence_monocyte_norm)$plot
persistence_monocyte_nb <- normalize_expt(persistence_monocyte, transform = "log2", convert = "cpm",
                                          batch = "svaseq", filter = TRUE)
plot_pca(persistence_monocyte_nb)$plot

## Neutrophils
persistence_neutrophil_norm <- normalize_expt(persistence_neutrophil, transform = "log2", convert = "cpm",
                                              norm = "quant", filter = TRUE)
plot_pca(persistence_neutrophil_norm)$plot
persistence_neutrophil_nb <- normalize_expt(persistence_neutrophil, transform = "log2", convert = "cpm",
                                            batch = "svaseq", filter = TRUE)
plot_pca(persistence_neutrophil_nb)$plot

## Eosinophils
persistence_eosinophil_norm <- normalize_expt(persistence_eosinophil, transform = "log2", convert = "cpm",
                                              norm = "quant", filter = TRUE)
plot_pca(persistence_eosinophil_norm)$plot
persistence_eosinophil_nb <- normalize_expt(persistence_eosinophil, transform = "log2", convert = "cpm",
                                            batch = "svaseq", filter = TRUE)
plot_pca(persistence_eosinophil_nb)$plot
```

### persistence DE

```{r}
persistence_de_sva <- all_pairwise(persistence_expt, filter = TRUE, model_batch = "svaseq")
persistence_de_sva
persistence_table_sva <- combine_de_tables(
  persistence_de_sva,
  excel = glue("{xlsx_prefix}/DE_Persistence/persistence_all_de_sva-v{ver}.xlsx"))
persistence_table_sva
persistence_monocyte_de_sva <- all_pairwise(persistence_monocyte, filter = TRUE, model_batch = "svaseq")
persistence_monocyte_de_sva
persistence_monocyte_table_sva <- combine_de_tables(
  persistence_monocyte_de_sva,
  excel = glue("{xlsx_prefix}/DE_Persistence/persistence_monocyte_de_sva-v{ver}.xlsx"))
persistence_monocyte_table_sva

persistence_neutrophil_de_sva <- all_pairwise(persistence_neutrophil, filter = TRUE, model_batch = "svaseq")
persistence_neutrophil_de_sva
persistence_neutrophil_table_sva <- combine_de_tables(
  persistence_neutrophil_de_sva,
  excel = glue("{xlsx_prefix}/DE_Persistence/persistence_neutrophil_de_sva-v{ver}.xlsx"))
persistence_neutrophil_table_sva

## There are insufficient samples (1) in the 'N' category.
##persistence_eosinophil_de_sva <- all_pairwise(persistence_eosinophil, filter = TRUE, model_batch = "svaseq")
##persistence_eosinophil_de_sva
##persistence_eosinophil_table_sva <- combine_de_tables(
##  persistence_eosinophil_de_sva,
##  excel = glue("{xlsx_prefix}/DE_Persistence/persistence_eosinophil_de_sva-v{ver}.xlsx"))
```

## Comparing visits without regard to cure/fail

### All cell types

```{r}
t_visit_all_de_sva <- all_pairwise(t_visit, filter = TRUE, model_batch = "svaseq")
t_visit_all_de_sva
t_visit_all_table_sva <- combine_de_tables(
  t_visit_all_de_sva, keepers = visit_contrasts,
  excel = glue("{xlsx_prefix}/DE_Visits/t_all_visit_tables_sva-v{ver}.xlsx"))
t_visit_all_table_sva
t_visit_all_sig_sva <- extract_significant_genes(
  t_visit_all_table_sva,
  excel = glue("{xlsx_prefix}/DE_Visits/t_all_visit_sig_sva-v{ver}.xlsx"))
t_visit_all_sig_sva
```

### Monocyte samples

```{r}
t_visit_monocytes <- set_expt_conditions(t_monocytes, fact = "visitnumber")

t_visit_monocyte_de_sva <- all_pairwise(t_visit_monocytes, filter = TRUE, model_batch = "svaseq")
t_visit_monocyte_de_sva
t_visit_monocyte_table_sva <- combine_de_tables(
  t_visit_monocyte_de_sva, keepers = visit_contrasts,
  excel = glue("{xlsx_prefix}/DE_Visits/Monocytes/t_monocyte_visit_tables_sva-v{ver}.xlsx"))
t_visit_monocyte_table_sva
t_visit_monocyte_sig_sva <- extract_significant_genes(
  t_visit_monocyte_table_sva,
  excel = glue("{xlsx_prefix}/DE_Visits/Monocytes/t_monocyte_visit_sig_sva-v{ver}.xlsx"))
t_visit_monocyte_sig_sva
```

### Neutrophil samples

```{r}
t_visit_neutrophils <- set_expt_conditions(t_neutrophils, fact = "visitnumber")

t_visit_neutrophil_de_sva <- all_pairwise(t_visit_neutrophils, filter = TRUE, model_batch = "svaseq")
t_visit_neutrophil_de_sva
t_visit_neutrophil_table_sva <- combine_de_tables(
  t_visit_neutrophil_de_sva, keepers = visit_contrasts,
  excel = glue("{xlsx_prefix}/DE_Visits/Neutrophils/t_neutrophil_visit_table_sva-v{ver}.xlsx"))
t_visit_neutrophil_table_sva
t_visit_neutrophil_sig_sva <- extract_significant_genes(
  t_visit_neutrophil_table_sva,
  excel = glue("{xlsx_prefix}/DE_Visits/Neutrophils/t_neutrophil_visit_sig_sva-v{ver}.xlsx"))
t_visit_neutrophil_sig_sva
```

### Eosinophil samples

```{r}
t_visit_eosinophils <- set_expt_conditions(t_eosinophils, fact="visitnumber")

t_visit_eosinophil_de <- all_pairwise(t_visit_eosinophils, filter = TRUE, model_batch = "svaseq")
t_visit_eosinophil_de
t_visit_eosinophil_table <- combine_de_tables(
  t_visit_eosinophil_de, keepers = visit_contrasts,
  excel = glue("{xlsx_prefix}/DE_Visits/Eosinophils/t_eosinophil_visit_table_sva-v{ver}.xlsx"))
t_visit_eosinophil_table
t_visit_eosinophil_sig <- extract_significant_genes(
  t_visit_eosinophil_table,
  excel = glue("{xlsx_prefix}/DE_Visits/Eosinophils/t_eosinophil_visit_sig_sva-v{ver}.xlsx"))
## No significant genes observed.
```

# Explore ROC

Alejandro showed some ROC curves for eosinophil data showing
sensitivity vs. specificity of a couple genes which were observed in
v1 eosinophils vs. all-times eosinophils across cure/fail.  I am
curious to better understand how this was done and what utility it
might have in other contexts.

To that end, I want to try something similar myself. In order to
properly perform the analysis with these various tools, I need to
reconfigure the data in a pretty specific format:

1.  Single df with 1 row per set of observations (sample in this case
I think)
2.  The outcome column(s) need to be 1 (or more?) metadata factor(s)
(cure/fail or a paste0 of relevant queries (eo_v1_cure,
eo_v123_cure, etc)
3.  The predictor column(s) are the measurements (rpkm of 1 or more
genes), 1 column each gene.

If I intend to use this for our tx data, I will likely need a utility
function to create the properly formatted input df.

For the purposes of my playing, I will choose three genes from the
eosinophil C/F table, one which is significant, one which is not, and
an arbitrary.

The input genes will therefore be chosen from the data structure:
t_cf_eosinophil_tables_sva:

ENSG00000198178, ENSG00000179344, ENSG00000182628

```{r}
eo_rpkm <- normalize_expt(tv1_eosinophils, convert = "rpkm", column = "cds_length")
```

# An external dataset

This paper is DOI:10.1126/scitranslmed.aax4204

Variable gene expression and parasite load predict treatment outcome in cutaneous leishmaniasis.

One query from Maria Adelaida is to see how this data fits with ours.
I have read this paper a couple of times now and I get confused on a
couple of points every time, which I will explain in a moment.  The
expermental design is key to my confusion and key to what I think is
being missed in our interpretation of the results:

1.  The PCA is not cure vs. fail but healthy skin vs. CL lesion.

## Only the Scott data

```{r}
external_cf

external_norm <- normalize_expt(external_cf, filter = TRUE, norm = "quant",
                                convert = "cpm", transform = "log2")
plot_pca(external_norm)
external_nb <- normalize_expt(external_cf, filter = TRUE, batch = "svaseq",
                                convert = "cpm", transform = "log2")
plot_pca(external_nb)

external_de <- all_pairwise(external_cf, filter = TRUE, model_batch = "svaseq")
external_de
external_table <- combine_de_tables(external_de, excel = "excel/scott_table.xlsx")
external_table
external_sig <- extract_significant_genes(external_table, excel = "excel/scott_sig.xlsx")
external_sig

external_top100 <- extract_significant_genes(external_table, n = 100)
external_up <- external_top100[["deseq"]][["ups"]][["failure_vs_cure"]]
external_down <- external_top100[["deseq"]][["downs"]][["failure_vs_cure"]]
```

# An explicit comparison of methods.

I think I am getting a significantly different result from Scott, so I
am going to do an explicit side-by-side comparison of our results at
each step.  In order to do this, I am using the capsule they kindly
provided with their publication.

I am copy/pasting material from their publication with some
modification which I will note as I go.

Here is their block 'r packages'

*Note/Spoiler alert*: It actually turns out our results are basically
relatively similar, I just didn't understand what comparisons are
actually in paper vs those I have primary interest.  In addition, we
handled gene IDs differently (gene card vs. EnsemblID) which has a
surprisingly big effect.

Oh, I just realized that when I did these analyses, I did them in a
completely separate tree and compared the results post-facto.  This
assumption remains in this document and therefore is unlikely to work
properly in the containerized environment I am attempting to create.
Given that the primary goal of this section is to show to myself that
I compared the two datasets as thoroughly as I could, perhaps I should
just disable them for the container and allow the reader to perform
the exercise de-novo.

```{r, eval=FALSE}
library(tidyverse)
library(ggthemes)
library(reshape2)
library(edgeR)
library(patchwork)
library(vegan)
library(DT)
library(tximport)
library(gplots)
library(FinCal)
library(ggrepel)
library(gt)
library(ggExtra)
library(EnsDb.Hsapiens.v86)
library(stringr)
library(cowplot)
library(ggpubr)
```

I have a separate tree in which I copied the capsule and data.  I
performed exactly their steps kallisto quant steps within it and put
the output data into the same place within it.  I did change the
commands slightly because I downloaded the files from SRA and so don't
have them with names like 'host_CL01', but instead 'PRJNA...'.  But
the samples are in the same order, so I sent the output files to the
same final filenames.  Here is an example from the first sample:

```{bash first_sample, eval=FALSE}
cd preprocessing
module add kallisto
kallisto index -i Homo_sapiens.GRCh38.cdna.all.Index Homo_sapiens.GRCh38.cdna.all.fa
# Map reads to the indexed reference transcriptome for HOST
# first the healthy subjects (HS)
export LESS = '--buffers 0 -B'
kallisto quant -i Homo_sapiens.GRCh38.cdna.all.Index -o host_HS01 -t 24 -b 60 \
         --single -l 250 -s 30 <(less SRR8668755/*-trimmed.fastq.xz) 2>host_HS01.log 1>&2 &
```

# Block 'sample_info'

I am going to change the path very slightly in the following block
simply because I put the capsule in a separate directory and do not
want to copy it here.  Otherwise it is unmodified.  Also, the function
gt::tab_header() annoys the crap out of me.

```{r, eval=FALSE}
import <- read_tsv("../scott_2019/capsule-6534016/data/studydesign.txt")
import %>% dplyr::filter(disease == "cutaneous") %>%
  dplyr::select(-2) %>%  gt() %>%
  tab_header(title = md("Clinical metadata from patients with cutaneous leishmaniasis (CL)"),
             subtitle = md("`(n=21)`")) %>%  cols_align(align = "center", columns = TRUE)
targets.lesion <- import
targets.onlypatients <- targets.lesion[8:28,] # only CL lesions (n=21)

# Making factors that will be used for pairwise comparisons:
# HS vs. CL lesions as a factor:
disease.lesion <- factor(targets.lesion$disease)
# Cure vs. Failure lesions as a factor:
treatment.lesion <- factor(targets.onlypatients$treatment_outcome)
```

# Importing the data and annotations

They did use a slightly different annotation set, Ensembl revision 86.
Once again I am modifying the paths slightly to reflect where I put
the capsule.

```{r, eval=FALSE}
# capturing Ensembl transcript IDs (tx) and gene symbols ("gene_name") from
# EnsDb.Hsapiens.v86 annotation package
Tx <- as.data.frame(transcripts(EnsDb.Hsapiens.v86,
                                columns=c(listColumns(EnsDb.Hsapiens.v86, "tx"),
                                          "gene_name")))

Tx <- dplyr::rename(Tx, target_id = tx_id)
row.names(Tx) <- NULL
Tx <- Tx[,c(6,12)]

# getting file paths for Kallisto outputs
paths.all <- file.path("../scott_2019/capsule-6534016/data/readMapping/human", targets.lesion$sample, "abundance.h5")
paths.patients <- file.path("../scott_2019/capsule-6534016/data/readMapping/human", targets.onlypatients$sample, "abundance.h5")

# importing .h5 Kallisto data and collapsing transcript-level data to genes
Txi.lesion.coding <- tximport(paths.all,
                              type = "kallisto",
                              tx2gene = Tx,
                              txOut = FALSE,
                              ignoreTxVersion = TRUE,
                              countsFromAbundance = "lengthScaledTPM")

# importing againg, but this time just the CL patients
Txi.lesion.coding.onlypatients <- tximport(paths.patients,
                                           type = "kallisto",
                                           tx2gene = Tx,
                                           txOut = FALSE,
                                           ignoreTxVersion = TRUE,
                                           countsFromAbundance = "lengthScaledTPM")
```

# Filtering and normalization

The block 'visualizationDatasets' follows unchanged.  In the next
block I will add another plot or perhaps 2

```{r, eval=FALSE}
# First make a DGEList from the counts:
Txi.lesion.coding.DGEList <- DGEList(Txi.lesion.coding$counts)
colnames(Txi.lesion.coding.DGEList$counts) <- targets.lesion$sample
colnames(Txi.lesion.coding$counts) <- targets.lesion$sample

Txi.lesion.coding.DGEList.OP <- DGEList(Txi.lesion.coding.onlypatients$counts)
colnames(Txi.lesion.coding.DGEList.OP) <- targets.onlypatients$sample

# Convert to counts per million:
Txi.lesion.coding.DGEList.cpm <- edgeR::cpm(Txi.lesion.coding.DGEList, log = TRUE)
Txi.lesion.coding.DGEList.OP.cpm <- edgeR::cpm(Txi.lesion.coding.DGEList.OP, log = TRUE)

keepers.coding <- rowSums(Txi.lesion.coding.DGEList.cpm>1)>=7
keepers.coding.OP <- rowSums(Txi.lesion.coding.DGEList.OP.cpm>1)>=7

Txi.lesion.coding.DGEList.filtered <- Txi.lesion.coding.DGEList[keepers.coding,]
Txi.lesion.coding.DGEList.OP.filtered <- Txi.lesion.coding.DGEList.OP[keepers.coding.OP,]

# convert back to cpm:
Txi.lesion.coding.DGEList.LogCPM.filtered <- edgeR::cpm(Txi.lesion.coding.DGEList.filtered,
                                                        log=TRUE)
Txi.lesion.coding.DGEList.LogCPM.OP.filtered <- edgeR::cpm(Txi.lesion.coding.DGEList.OP.filtered,
                                                           log=TRUE)

# Normalizing data:
calcNorm1 <- calcNormFactors(Txi.lesion.coding.DGEList.filtered, method = "TMM")
calcNorm2 <- calcNormFactors(Txi.lesion.coding.DGEList.OP.filtered, method = "TMM")

Txi.lesion.coding.DGEList.LogCPM.filtered.norm <- edgeR::cpm(calcNorm1, log=TRUE)
colnames(Txi.lesion.coding.DGEList.LogCPM.filtered.norm) <- targets.lesion$sample
Txi.lesion.coding.DGEList.OP.LogCPM.filtered.norm <- edgeR::cpm(calcNorm2, log=TRUE)
colnames(Txi.lesion.coding.DGEList.OP.LogCPM.filtered.norm) <- targets.onlypatients$sample
# Raw dataset:
V1 <- as.data.frame(Txi.lesion.coding.DGEList.cpm)
colnames(V1) <- targets.lesion$sample
V1 <- melt(V1)
colnames(V1) <- c("sample","expression")

# Filtered dataset:
V1.1 <- as.data.frame(Txi.lesion.coding.DGEList.LogCPM.filtered)
colnames(V1.1) <- targets.lesion$sample
V1.1 <- melt(V1.1)
colnames(V1.1) <- c("sample","expression")

# Filtered-normalized dataset:
V1.1.1 <- as.data.frame(Txi.lesion.coding.DGEList.LogCPM.filtered.norm)
colnames(V1.1.1) <- targets.lesion$sample
V1.1.1 <- melt(V1.1.1)
colnames(V1.1.1) <- c("sample","expression")

# plotting:
ggplot(V1, aes(x=sample, y=expression, fill=sample)) +
  geom_violin(trim = TRUE, show.legend = TRUE) +
  stat_summary(fun.y = "median", geom = "point", shape = 95, size = 10, color = "black") +
  theme_bw() +
  theme(legend.position = "none", axis.title=element_text(size=7),
        axis.title.x=element_blank(), axis.text=element_text(size=5),
        axis.text.x = element_text(angle = 90, hjust = 1),
        plot.title = element_text(size = 7)) +
  ggtitle("Raw dataset") +
  ggplot(V1.1, aes(x=sample, y=expression, fill=sample)) +
  geom_violin(trim = TRUE, show.legend = TRUE) +
  stat_summary(fun.y = "median", geom = "point", shape = 95, size = 10, color = "black") +
  theme_bw() +
  theme(legend.position = "none", axis.title=element_text(size=7),
        axis.title.x=element_blank(), axis.text=element_text(size=5),
        axis.text.x = element_text(angle = 90, hjust = 1),
        plot.title = element_text(size = 7)) +
  ggtitle("Filtered dataset") +
  ggplot(V1.1.1, aes(x=sample, y=expression, fill=sample)) +
  geom_violin(trim = TRUE, show.legend = TRUE) +
  stat_summary(fun.y = "median", geom = "point", shape = 95, size = 10, color = "black") +
  theme_bw() +
  theme(legend.position = "none", axis.title=element_text(size=7),
        axis.title.x=element_blank(), axis.text=element_text(size=5),
        axis.text.x = element_text(angle = 90, hjust = 1),
        plot.title = element_text(size = 7)) +
  ggtitle("Filtered and normalized dataset")
```

# The unfiltered data

The following block in their dataset recreated the matrix without
filtering and will use that for differential expression.  It is a
little hard to follow for me because they subset based on the sample
numbers (8 to 28, which if I am not mistaken just drops the healthy
samples?).

```{r, eval=FALSE}
DataNotFiltered_Norm_OP <- calcNormFactors(Txi.lesion.coding.DGEList[,8:28],
                                           method = "TMM")
DataNotFiltered_Norm_log2CPM_OP <- edgeR::cpm(DataNotFiltered_Norm_OP, log=TRUE)
colnames(DataNotFiltered_Norm_log2CPM_OP) <- targets.onlypatients$sample
CPM_normData_notfiltered_OP <- 2^(DataNotFiltered_Norm_log2CPM_OP)
#uncomment the next line to produce raw data that was uploaded to the Gene Expression Omnibus (GEO) for publication.
#write.table(Txi.lesion.coding$counts, file = "Amorim_GEO_raw.txt", sep = "\t", quote = FALSE)

# Including all the individuals (HS and CL patients) for public domain submission:
DataNotFiltered_Norm <- calcNormFactors(Txi.lesion.coding.DGEList, method = "TMM")
DataNotFiltered_Norm_log2CPM <- edgeR::cpm(DataNotFiltered_Norm, log=TRUE)
colnames(DataNotFiltered_Norm_log2CPM) <- targets.lesion$sample
CPM_normData_notfiltered <- 2^(DataNotFiltered_Norm_log2CPM)
#uncomment the next line to produce the normalized data file that was uploaded to the Gene Expression Omnibus (GEO) for publication.
#write.table(DataNotFiltered_Norm_log2CPM, "Amorim_GEO_normalized.txt", sep = "\t", quote = FALSE)
```

# The scott exploratory analysis

The following block generated a couple of the figures in the paper and
comprise a pretty straightforward PCA.  I am going to make a following
block containing the same image with the cure/fail visualization using
the same method/data.

```{r, eval=FALSE}
pca.res <- prcomp(t(Txi.lesion.coding.DGEList.LogCPM.filtered.norm), scale.=F, retx=T)
pc.var <- pca.res$sdev^2
pc.per <- round(pc.var/sum(pc.var)*100, 1)
data.frame <- as.data.frame(pca.res$x)

# Calculate distance between samples by permanova:
allsamples.dist <- vegdist(t(2^Txi.lesion.coding.DGEList.LogCPM.filtered.norm),
                           method = "bray")

vegan <- adonis2(allsamples.dist~targets.lesion$disease,
                 data=targets.lesion,
                 permutations = 999, method="bray")

targets.lesion$disease
ggplot(data.frame, aes(x=PC1, y=PC2, color=factor(targets.lesion$disease))) +
  geom_point(size=5, shape=20) +
  theme_calc() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        axis.text.x = element_text(size = 15, vjust = 0.5),
        axis.text.y = element_text(size = 15), axis.title = element_text(size = 15),
        legend.position="none") +
  scale_color_manual(values = c("#073F80","#EB512C")) +
  annotate("text", x=-50, y=80, label=paste("Permanova Pr(>F) =",
                                            vegan[1,5]), size=3, fontface="bold") +
  xlab(paste("PC1 -",pc.per[1],"%")) +
  ylab(paste("PC2 -",pc.per[2],"%")) +
  xlim(-200,110)
```

### My most similar pca

```{r}
external_l2cpm <- normalize_expt(external_cf, filter = TRUE,
                                convert = "cpm", transform = "log2")
plot_pca(external_l2cpm, plot_labels = "repel")
```

## Cure/Fail PCA using the same prcomp result

I am just copy/pasting their code again, but changing the color factor
so that cure is purple, failure is red, and na(uninfected) is black.

The following plot should be the first direct comparison point between
the two analysis pipelines.  Thus, if you look back a few block at my
invocation of plot_pca(external_norm), you will see a green/orange
plot which is functionally identical if you note:

1.  The x and y axes are flipped, which ok whatever it is PCA.
2.  I excluded the healthy samples.
3.  I dropped to gene level and used hisat.

With those caveats in mind, it is trivial to find the same
relationshipes in the samples.  E.g. the bottom red/purple individual
samples are in the same relative position as my top orange/green pair.
the same 4 samples are relative x-axis outliers (my right green, their
left purple).  The last 6 samples (my orange, their red) are all in
the relative orientation.

I think I can further prove the similarity of our inputs via a direct
comparison of the datastructures:
Txi.lesion.coding.DGEList.LogCPM.filtered.norm (ugh what a name)
vs. external_cf.  In order to make that comparison, I need to rename
my rows to the genecard IDs and the columns.

```{r, eval=FALSE}
their_norm_exprs <- Txi.lesion.coding.DGEList.LogCPM.filtered.norm

my_hgnc_ids <- make.names(fData(external_cf)[["hgnc_symbol"]], unique = TRUE)
my_renamed <- set_expt_genenames(external_cf, ids = my_hgnc_ids)
my_norm <- normalize_expt(my_renamed, filter = TRUE, transform = "log2", convert = "cpm")
my_norm_exprs <- as.data.frame(exprs(my_norm))

our_exprs <- merge(their_norm_exprs, my_norm_exprs, by = "row.names")
rownames(our_exprs) <- our_exprs[["Row.names"]]
our_exprs[["Row.names"]] <- NULL
dim(our_exprs)

## I fully expected a correlation heatmap of the combined
## data to show a set of paired samples across the board.
## That is absolutely not true.
correlations <- plot_corheat(our_exprs)
correlations$plot
```

```{r, eval=FALSE}
color_fact <- factor(targets.lesion$treatment_outcome)
levels(color_fact)
## Added by atb to see cure/fail on the same dataset
ggplot(data.frame, aes(x=PC1, y=PC2, color=color_fact)) +
  geom_point(size=5, shape=20) +
  theme_calc() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        axis.text.x = element_text(size = 15, vjust = 0.5),
        axis.text.y = element_text(size = 15), axis.title = element_text(size = 15),
        legend.position="none") +
  scale_color_manual(values = c("purple", "red","black")) +
  annotate("text", x=-50, y=80, label=paste("Permanova Pr(>F) =",
                                            vegan[1,5]), size=3, fontface="bold") +
  xlab(paste("PC1 -",pc.per[1],"%")) +
  ylab(paste("PC2 -",pc.per[2],"%")) +
  xlim(-200,110)
```

# DE comparisons

The following is their comparison of healthy tissue vs.  CL lesion and
Failure vs. Cure.  I am going to follow it with my analagous
examination using limma.  Note, each of the pairs of variables created
in the following block is xxx followed by xxx.treat; the former is
healthy vs lesion and the latter is the fail vs cure set.

```{r, eval=FALSE}
# Model matrices:
# CL lesions vs. HS:
design.lesion <- model.matrix(~0 + disease.lesion)
colnames(design.lesion) <- levels(disease.lesion)

# Failure vs. Cure:
design.lesion.treatment <- model.matrix(~0 + treatment.lesion)
colnames(design.lesion.treatment) <- levels(treatment.lesion)

myDGEList.lesion.coding <- DGEList(calcNorm1$counts)
myDGEList.OP.NotFil <- DGEList(CPM_normData_notfiltered_OP)

# Model mean-variance trend and fit linear model to data.
# Use VOOM function from Limma package to model the mean-variance relationship
normData.lesion.coding <- voom(myDGEList.lesion.coding, design.lesion)
normData.OP.NotFil <- voom(myDGEList.OP.NotFil, design.lesion.treatment)

colnames(normData.lesion.coding) <- targets.lesion$sample
colnames(normData.OP.NotFil) <- targets.onlypatients$sample

# fit a linear model to your data
fit.lesion.coding <- lmFit(normData.lesion.coding, design.lesion)
fit.lesion.coding.treatment <- lmFit(normData.OP.NotFil, design.lesion.treatment)

# contrast matrix
contrast.matrix.lesion <- makeContrasts(CL.vs.CON = cutaneous - control,
                                        levels=design.lesion)
contrast.matrix.lesion.treat <- makeContrasts(failure.vs.cure = failure - cure,
                                              levels=design.lesion.treatment)

# extract the linear model fit
fits.lesion.coding <- contrasts.fit(fit.lesion.coding,
                                    contrast.matrix.lesion)
fits.lesion.coding.treat <- contrasts.fit(fit.lesion.coding.treatment,
                                          contrast.matrix.lesion.treat)

# get bayesian stats for your linear model fit
ebFit.lesion.coding <- eBayes(fits.lesion.coding)
ebFit.lesion.coding.treat <- eBayes(fits.lesion.coding.treat)

# TopTable ----
allHits.lesion.coding <- topTable(ebFit.lesion.coding,
                                  adjust ="BH", coef=1,
                                  number=34935, sort.by="logFC")
allHits.lesion.coding.treat <- topTable(ebFit.lesion.coding.treat,
                                        adjust ="BH", coef=1,
                                        number=34776, sort.by="logFC")
myTopHits <- rownames_to_column(allHits.lesion.coding, "geneID")
myTopHits.treat <- rownames_to_column(allHits.lesion.coding.treat, "geneID")

# mutate the format of numeric values:
myTopHits <- mutate(myTopHits, log10Pval = round(-log10(adj.P.Val),2),
                    adj.P.Val = round(adj.P.Val, 2),
                    B = round(B, 2),
                    AveExpr = round(AveExpr, 2),
                    t = round(t, 2),
                    logFC = round(logFC, 2),
                    geneID = geneID)

myTopHits.treat <- mutate(myTopHits.treat, log10Pval = round(-log10(adj.P.Val),2),
                          adj.P.Val = round(adj.P.Val, 2),
                          B = round(B, 2),
                          AveExpr = round(AveExpr, 2),
                          t = round(t, 2),
                          logFC = round(logFC, 2),
                          geneID = geneID)
#save(myTopHits, file = "myTopHits")
#save(myTopHits.treat, file = "myTopHits.treat")
```

# Perform my analagous limma analysis

```{r compare_result, eval=FALSE}
limma_cf <- limma_pairwise(my_renamed, model_batch = FALSE)

my_table <- limma_cf[["all_tables"]][["failure_vs_cure"]]
their_table <- myTopHits.treat

dim(my_table)
dim(myTopHits.treat)
our_table <- merge(my_table, myTopHits.treat, by.x = "row.names", by.y = "geneID")
dim(our_table)
comparison <- plot_linear_scatter(our_table[, c("logFC.x", "logFC.y")])
comparison$scatter
comparison$correlation
comparison$lm_model
```

Ok, so there is a constituitive difference in our results, and it is
significant.  What does that mean for the set of genes observed?

# See if there are shared DE genes

!!NOTE!!  I am using a non-adjusted p-value filter here because I want
to use the same filter they used for the volcano plot.

```{r shared_genes, eval=FALSE}
my_filter <- abs(my_table[["logFC"]]) > 1.0 & my_table[["P.Value"]] <= 0.05
sum(my_filter)
their_filter <- abs(their_table[["logFC"]]) > 1.0 & their_table[["P.Value"]] <= 0.05
sum(their_filter)

my_shared <- rownames(my_table)[my_filter] %in% their_table[their_filter, "geneID"]
sum(my_shared)

shared <- rownames(my_table)[my_filter]
shared[my_shared]
```

## Compare the two datasets directly

```{r scott_external}
only_tmrc3 <- subset_expt(tmrc3_external, subset = "condition=='Colombia'") %>%
  set_expt_conditions(fact = "finaloutcome")
only_tmrc3_de <- all_pairwise(only_tmrc3, model_batch = "svaseq", filter = TRUE)
only_tmrc3_de
only_tmrc3_table <- combine_de_tables(only_tmrc3_de)
only_tmrc3_table
only_tmrc3_top100 <- extract_significant_genes(only_tmrc3_table, n = 100)
only_tmrc3_up <- only_tmrc3_top100[["deseq"]][["ups"]][["failure_vs_cure"]]
only_tmrc3_down <- only_tmrc3_top100[["deseq"]][["downs"]][["failure_vs_cure"]]

tmrc3_external_de <- all_pairwise(tmrc3_external, model_batch = "svaseq", filter = "simple")
tmrc3_external_table <- combine_de_tables(tmrc3_external_de, excel = "excel/tmrc3_scott_biopsies.xlsx")
tmrc3_external_sig <- extract_significant_genes(tmrc3_external_table, excel = "excel/tmrc3_scott_biopsies_sig.xlsx")

tmrc3_external_cf <- set_expt_conditions(tmrc3_external, fact = "finaloutcome") %>%
  set_expt_batches(fact = "lab")
tmrc3_external_cf_norm <- normalize_expt(tmrc3_external_cf, filter = TRUE, norm = "quant", convert = "cpm", transform = "log2")
plot_pca(tmrc3_external_cf_norm)
tmrc3_external_cf_nb <- normalize_expt(tmrc3_external_cf, filter = TRUE, batch = "svaseq", convert = "cpm", transform = "log2")
plot_pca(tmrc3_external_cf_nb)

tmrc3_external_cf_de <- all_pairwise(tmrc3_external_cf, model_batch = "svaseq", filter = TRUE)
tmrc3_external_cf_de
tmrc3_external_cf_table <- combine_de_tables(tmrc3_external_cf_de, excel = "excel/tmrc3_scott_cf_table.xlsx")
tmrc3_external_cf_table
tmrc3_external_cf_sig <- extract_significant_genes(tmrc3_external_cf_table, excel = "excel/tmrc3_scott_cf_sig.xlsx")
tmrc3_external_cf_sig

tmrc3_external_species <- set_expt_conditions(tmrc3_external, fact = "ParasiteSpecies") %>%
  set_expt_colors(color_choices[["parasite"]])
```

# Compare the l2FC values

Let us look at the top/bottom 100 genes of these two datasets and see if they
have any similarities.

Note to self, set up s4 dispatch on compare_de_tables!

```{r}
compared <- compare_de_tables(only_tmrc3_table, external_table, first_table = 1, second_table = 1)
compared$scatter
compared$correlation
```

# Save the results for playing elsewhere

```{r}
save(list = ls(), file = glue("rda/tmrc3_03differential_expression_tumaco-v{ver}.rda"))
```

```{r loadme_after, eval=FALSE}
tmp <- loadme(filename = savefile)
```
