---
title: "TMRC3 `r Sys.getenv('VERSION')`: Differential Expression analyses"
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
runtime: shiny
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: zenburn
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body .main-container {
  max-width: 1600px;
}
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
  font-size: 16px
}
</style>

```{r options, include=FALSE}
library(hpgltools)
library(dplyr)
library(forcats)
library(glue)
library(DOSE)

knitr::opts_knit$set(progress = TRUE, verbose = TRUE, width = 90, echo = TRUE)
knitr::opts_chunk$set(
  error = TRUE, fig.width = 8, fig.height = 8, fig.retina = 2,
  out.width = "100%", dev = "png",
  dev.args = list(png = list(type = "cairo-png")))
old_options <- options(digits = 4, stringsAsFactors = FALSE, knitr.duplicate.label = "allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size = 12))
ver <- Sys.getenv("VERSION")
parallel <- toupper(Sys.getenv("PARALLEL"))
if (parallel == "" || parallel == "TRUE") {
  parallel <- TRUE
} else {
  parallel <- FALSE
}
rundate <- format(Sys.Date(), format = "%Y%m%d")

rmd_file <- glue("03differential_expression_both.Rmd")
savefile <- gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = rmd_file)
loaded <- load(file = glue("rda/tmrc3_data_structures-v{ver}.rda"))
xlsx_prefix <- "analyses/3_cali_and_tumaco"
clinic_prefix <- glue("{xlsx_prefix}/DE_Clinic")
clinic_cf_prefix <- glue("{xlsx_prefix}/DE_Clinic_Cure_Fail")
cf_prefix <- glue("{xlsx_prefix}/DE_Cure_Fail")
visit_prefix <- glue("{xlsx_prefix}/DE_Visits")
sex_prefix <- glue("{xlsx_prefix}/Sex")
eth_prefix <- glue("{xlsx_prefix}/Ethnicity")
gsea_prefix <- glue("{xlsx_prefix}/GSEA")
```

# Changelog

* 202309: Disabled GSVA analyses until/unless we get permission to
  include the mSigDB 7.5.1 release.  I will simplify the filenames so
  that one may easily drop in a downloaded copy of the data and run
  those blocks.  Until then, I guess you will have to trust me when I
  say those blocks all work?
* 202309: Moved all GSEA analyses to 04lrt_gsea_gsva.Rmd
* 202309 next day: Moving GSEA back because it adds too much
  complexity to save/reload the DE results for gProfiler and friends.
* Still hunting for messed up colors, changed input data to match new version.

# Introduction

The various differential expression analyses of the data generated in tmrc3_datasets
will occur in this document.

## Naming conventions

I am going to try to standardize how I name the various data
structures created in this document.  Most of the large data created
are either sets of differential expression analyses, their combined
results, or the set of results deemed 'significant'.

Hopefully by now they all follow these guidelines:

{clinic(s)}_sample-subset}_{primary-question(s)}_{datatype}_{batch-method}

* {clinic}: This is either tc or t for Tumaco and Cali, or just
Tumaco.
* {sample-subset}: Things like 'all' or 'monocytes'.
* {primary-question}: Shorthand name for the primary contrasts
performed, thus 'clinics' would suggest a comparison of Tumaco
vs. Cali.  'visits' would compare v2/v1, etc.
* {datatype}: de, table, sig
* {batch-type}: nobatch, batch{factor}, sva.  {factor} in this
instance should be a column from the metadata.

With this in mind, 'tc_biopsies_clinic_de_sva' should be the Tumaco+Cali
biopsy data after performing the differential expression analyses
comparing the clinics using sva.

I suspect there remain some exceptions and/or errors.

## Define contrasts for DE analyses

Each of the following lists describes the set of contrasts that I
think are interesting for the various ways one might consider the
TMRC3 dataset.  The variables are named according to the assumed data
with which they will be used, thus tc_cf_contrasts is expected to be
used for the Tumaco+Cali data and provide a series of cure/fail
comparisons which (to the extent possible) across both locations.  In
every case, the name of the list element will be used as the contrast
name, and will thus be seen as the sheet name in the output xlsx
file(s); the two pieces of the character vector value are the
numerator and denominator of the associated contrast.

## GSEA

The GSEA analyses will follow each DE analysis during this document.

Most (all?) of the GSEA analyses used in this paper were done via
gProfiler rather than goseq/clusterProfiler/topGO/GOstats.  Primarily
because it is so easy to invoke gprofiler.

```{r}
clinic_contrasts <- list(
  "clinics" = c("cali", "tumaco"))
## In some cases we have no Cali failure samples, so there remain only 2
## contrasts that are likely of interest
tc_cf_contrasts <- list(
  "tumaco" = c("tumacofailure", "tumacocure"),
  "cure" = c("tumacocure", "calicure"))
## In other cases, we have cure/fail for both places.
clinic_cf_contrasts <- list(
  "cali" = c("califailure", "calicure"),
  "tumaco" = c("tumacofailure", "tumacocure"),
  "cure" = c("tumacocure", "calicure"),
  "fail" = c("tumacofailure", "califailure"))
cf_contrast <- list(
  "outcome" = c("tumacofailure", "tumacocure"))
t_cf_contrast <- list(
  "outcome" = c("failure", "cure"))
visitcf_contrasts <- list(
  "v1cf" = c("v1failure", "v1cure"),
  "v2cf" = c("v2failure", "v2cure"),
  "v3cf" = c("v3failure", "v3cure"))
visit_contrasts <- list(
  "v2v1" = c("c2", "c1"),
  "v3v1" = c("c3", "c1"),
  "v3v2" = c("c3", "c2"))
visit_v1later <- list(
  "later_vs_first" = c("later", "first"))
celltypes <- list(
  "eo_mono" = c("eosinophils", "monocytes"),
  "ne_mono" = c("neutrophils", "monocytes"),
  "eo_ne" = c("eosinophils", "neutrophils"))
ethnicity_contrasts <- list(
  "mestizo_indigenous" = c("mestiza", "indigena"),
  "mestizo_afrocol" = c("mestiza", "afrocol"),
  "indigenous_afrocol" = c("indigena", "afrocol"))
```

# Compare samples by clinic

## DE: Compare clinics, all samples

Perform a svaseq-guided comparison of the two clinics.  Ideally this
will give some clue about just how strong the clinic-based batch
effect really is and what its causes are.

```{r}
tc_clinic_type <- tc_valid %>%
  set_expt_conditions(fact = "clinic") %>%
  set_expt_batches(fact = "typeofcells")

table(pData(tc_clinic_type)[["condition"]])
tc_all_clinic_de_sva <- all_pairwise(tc_clinic_type, model_batch = "svaseq",
                                     filter = TRUE, parallel = parallel, methods = methods)
tc_all_clinic_de_sva
tc_all_clinic_de_sva[["deseq"]][["contrasts_performed"]]

tc_all_clinic_table_sva <- combine_de_tables(
  tc_all_clinic_de_sva, keepers = clinic_contrasts,
  excel = glue("{clinic_prefix}/tc_all_clinic_table_sva-v{ver}.xlsx"))
tc_all_clinic_table_sva
tc_all_clinic_sig_sva <- extract_significant_genes(
  tc_all_clinic_table_sva,
  excel = glue("{clinic_prefix}/compare_clinics/tc_clinic_type_sig_sva-v{ver}.xlsx"))
tc_all_clinic_sig_sva
```

### GSEA: comparing the clinics

```{r}
increased_tumaco_categories_up <- simple_gprofiler(
  tc_all_clinic_sig_sva[["deseq"]][["ups"]][["clinics"]],
  excel = glue("{gsea_prefix}/tumaco_cateogies_up-v{ver}.xlsx"))
increased_tumaco_categories_up
increased_tumaco_categories_up[["pvalue_plots"]][["BP"]]

increased_cali_categories <- simple_gprofiler(
  tc_all_clinic_sig_sva[["deseq"]][["downs"]][["clinics"]],
  excel = glue("{gsea_prefix}/cali_cateogies_up-v{ver}.xlsx"))
increased_cali_categories
increased_cali_categories[["pvalue_plots"]][["BP"]]
```

### Visualize clinic differences

Let us take a quick look at the results of the comparison of
Tumaco/Cali

Note: I keep re-introducing an error which causes these (volcano and MA) plots to be
reversed with respect to the logFC values.  Pay careful attention to
these and make sure that they agree with the numbers of genes observed
in the contrast.

```{r}
## Check that up is up
summary(tc_all_clinic_table_sva[["data"]][["clinics"]][["deseq_logfc"]])
## I think we can assume that most genes are down when considering Tumaco/Cali.
sum(tc_all_clinic_table_sva$data$clinics$deseq_logfc < -1.0 &
      tc_all_clinic_table_sva$data$clinics$deseq_adjp < 0.05)
tc_all_clinic_table_sva[["plots"]][["clinics"]][["deseq_vol_plots"]]
## Ok, so it says 1794 up, but that is clearly the down side...  Something is definitely messed up.
## The points are on the correct sides of the plot, but the categories of up/down are reversed.
## Theresa noted that she colors differently, and I think better: left side gets called
## 'increased in denominator', right side gets called 'increased in numerator';
## these two groups are colored according to their condition colors, and everything else is gray.
## I am checking out Theresa's helper_functions.R to get a sense of how she handles this, I think
## I can use a variant of her idea pretty easily:
##  1.  Add a column 'Significance', which is a factor, and contains either 'Not enriched',
##      'Enriched in x', or 'Enriched in y' according to the logfc/adjp.
##  2.  use the significance column for the geom_point color/fill in the volcano plot.
## My change to this idea would be to extract the colors from the input expressionset.
```

There appear to be many more genes which are increased in the Tumaco
samples with respect to the Cali samples.

## DE: Compare clinics, eosinophil samples

The remaining cell types all have pretty strong clinic-based variance;
but I am not certain if it is consistent across cell types.

```{r}
table(pData(tc_eosinophils)[["condition"]])
tc_eosinophils_clinic_de_nobatch <- all_pairwise(tc_eosinophils, parallel = parallel,
                                                 model_batch = FALSE, filter = TRUE,
                                                 methods = methods)
tc_eosinophils_clinic_de_nobatch
tc_eosinophils_clinic_de_nobatch[["deseq"]][["contrasts_performed"]]

tc_eosinophils_clinic_table_nobatch <- combine_de_tables(
  tc_eosinophils_clinic_de_nobatch, keepers = tc_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Eosinophils/tc_eosinophils_clinic_table_nobatch-v{ver}.xlsx"))
tc_eosinophils_clinic_table_nobatch
tc_eosinophils_clinic_sig_nobatch <- extract_significant_genes(
  tc_eosinophils_clinic_table_nobatch,
  excel = glue("{clinic_cf_prefix}/Eosinophils/tc_eosinophils_clinic_sig_nobatch-v{ver}.xlsx"))
tc_eosinophils_clinic_sig_nobatch

tc_eosinophils_clinic_de_sva <- all_pairwise(tc_eosinophils, model_batch = "svaseq",
                                             parallel = parallel, filter = TRUE, methods = methods)
tc_eosinophils_clinic_de_sva
tc_eosinophils_clinic_de_sva[["deseq"]][["contrasts_performed"]]

tc_eosinophils_clinic_table_sva <- combine_de_tables(
  tc_eosinophils_clinic_de_sva, keepers = tc_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Eosinophils/tc_eosinophils_clinic_table_sva-v{ver}.xlsx"))
tc_eosinophils_clinic_table_sva
tc_eosinophils_clinic_sig_sva <- extract_significant_genes(
  tc_eosinophils_clinic_table_sva,
  excel = glue("{clinic_cf_prefix}/Eosinophils/tc_eosinophils_clinic_sig_sva-v{ver}.xlsx"))
tc_eosinophils_clinic_sig_sva
```

## DE: Compare clinics, biopsy samples

Interestingly to me, the biopsy samples appear to have the least
location-based variance.  But we can perform an explicit DE and see
how well that hypothesis holds up.

Note that these data include cure and fail samples for

```{r}
table(pData(tc_biopsies)[["condition"]])
tc_biopsies_clinic_de_sva <- all_pairwise(tc_biopsies, parallel = parallel,
                                          model_batch = "svaseq", filter = TRUE,
                                          methods = methods)
tc_biopsies_clinic_de_sva
tc_biopsies_clinic_de_sva[["deseq"]][["contrasts_performed"]]

tc_biopsies_clinic_table_sva <- combine_de_tables(
  tc_biopsies_clinic_de_sva, keepers = tc_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Biopsies/tc_biopsies_clinic_table_sva-v{ver}.xlsx"))
tc_biopsies_clinic_table_sva
tc_biopsies_clinic_sig_sva <- extract_significant_genes(
  tc_biopsies_clinic_table_sva,
  excel = glue("{clinic_cf_prefix}/Biopsies/tc_biopsies_clinic_sig_sva-v{ver}.xlsx"))
tc_biopsies_clinic_sig_sva
```

## DE: Compare clinics, monocyte samples

At least for the moment, I am only looking at the differences between
no-batch vs. sva across clinics for the monocyte samples.  This
was chosen mostly arbitrarily.

### DE: Compare clinics, monocytes without batch estimation

Our baseline is the comparison of the monocytes samples without batch
in the model or surrogate estimation.  In theory at least, this should
correspond to the PCA plot above when no batch estimation was performed.

```{r}
table(pData(tc_monocytes)[["condition"]])
tc_monocytes_de_nobatch <- all_pairwise(tc_monocytes, model_batch = FALSE,
                                        parallel = parallel, filter = TRUE,
                                        methods = methods)
tc_monocytes_de_nobatch

tc_monocytes_table_nobatch <- combine_de_tables(
  tc_monocytes_de_nobatch, keepers = clinic_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Monocytes/tc_monocytes_clinic_table_nobatch-v{ver}.xlsx"))
tc_monocytes_table_nobatch
tc_monocytes_sig_nobatch <- extract_significant_genes(
  tc_monocytes_table_nobatch,
  excel = glue("{clinic_cf_prefix}/Monocytes/tc_monocytes_clinic_sig_nobatch-v{ver}.xlsx"))
tc_monocytes_sig_nobatch
```

### DE: Compare clinics, monocytes with svaseq

In contrast, the following comparison should give a view of the data
corresponding to the svaseq PCA plot above.  In the best case
scenario, we should therefore be able to see some significane
differences between the Tumaco cure and fail samples.

```{r}
tc_monocytes_de_sva <- all_pairwise(tc_monocytes, model_batch = "svaseq",
                                    parallel = parallel, filter = TRUE,
                                    methods = methods)
tc_monocytes_de_sva

tc_monocytes_table_sva <- combine_de_tables(
  tc_monocytes_de_sva, keepers = clinic_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Monocytes/tc_monocytes_clinic_table_sva-v{ver}.xlsx"))
tc_monocytes_table_sva
tc_monocytes_sig_sva <- extract_significant_genes(
  tc_monocytes_table_sva,
  excel = glue("{clinic_cf_prefix}/Monocytes/tc_monocytes_clinic_sig_sva-v{ver}.xlsx"))
tc_monocytes_sig_sva
```

### DE Compare: How similar are the no-batch vs. SVA results?

The following block shows that these two results are exceedingly
different, sugesting that the Cali cure/fail and Tumaco cure/fail
cannot easily be considered in the same analysis.  I did some playing
around with my calculate_aucc function in this block and found that it
is in some important way broken, at least if one expands the top-n
genes to more than 20% of the number of genes in the data.

```{r}
cali_table <- tc_monocytes_table_nobatch[["data"]][["cali"]]
table <- tc_monocytes_table_nobatch[["data"]][["tumaco"]]

cali_aucc <- calculate_aucc(cali_table, table, px = "deseq_adjp", py = "deseq_adjp",
                            lx = "deseq_logfc", ly = "deseq_logfc")
cali_aucc

cali_table_sva <- tc_monocytes_table_sva[["data"]][["cali"]]
tumaco_table_sva <- tc_monocytes_table_sva[["data"]][["tumaco"]]
cali_aucc_sva <- calculate_aucc(cali_table_sva, tumaco_table_sva, px = "deseq_adjp",
                                py = "deseq_adjp", lx = "deseq_logfc", ly = "deseq_logfc")
cali_aucc_sva
```

## DE: Compare clinics, neutrophil samples

```{r}
tc_neutrophils_de_nobatch <- all_pairwise(tc_neutrophils, parallel = parallel,
                                          model_batch = FALSE, filter = TRUE,
                                          methods = methods)
tc_neutrophils_de_nobatch

tc_neutrophils_table_nobatch <- combine_de_tables(
  tc_neutrophils_de_nobatch, keepers = clinic_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Neutrophils/tc_neutrophils_table_nobatch-v{ver}.xlsx"))
tc_neutrophils_table_nobatch
tc_neutrophils_sig_nobatch <- extract_significant_genes(
  tc_neutrophils_table_nobatch,
  excel = glue("{clinic_cf_prefix}/Neutrophils/tc_neutrophils_sig_nobatch-v{ver}.xlsx"))
tc_neutrophils_sig_nobatch

tc_neutrophils_de_sva <- all_pairwise(tc_neutrophils, parallel = parallel,
                                      model_batch = "svaseq", filter = TRUE,
                                      methods = methods)
tc_neutrophils_de_sva

tc_neutrophils_table_sva <- combine_de_tables(
  tc_neutrophils_de_sva, keepers = clinic_cf_contrasts,
  excel = glue("{clinic_cf_prefix}/Neutrophils/tc_neutrophils_table_sva-v{ver}.xlsx"))
tc_neutrophils_table_sva
tc_neutrophils_sig_sva <- extract_significant_genes(
  tc_neutrophils_table_sva,
  excel = glue("{clinic_cf_prefix}/Neutrophils/tc_neutrophils_sig_sva-v{ver}.xlsx"))
tc_neutrophils_sig_sva
```

## GSEA: Extract clinic-specific genes

Given the above comparisons, we can extract some gene sets which
resulted from those DE analyses and eventually perform some
ontology/KEGG/reactome/etc searches.  This reminds me, I want to make
my extract_significant_ functions to return gene-set data structures
and my various ontology searches to take them as inputs.  This should
help avoid potential errors when extracting up/down genes.

```{r}
clinic_sigenes_up <- rownames(tc_all_clinic_sig_sva[["deseq"]][["ups"]][["clinics"]])
clinic_sigenes_down <- rownames(tc_all_clinic_sig_sva[["deseq"]][["downs"]][["clinics"]])
clinic_sigenes <- c(clinic_sigenes_up, clinic_sigenes_down)

tc_eosinophils_sigenes_up <- rownames(tc_eosinophils_clinic_sig_sva[["deseq"]][["ups"]][["cure"]])
tc_eosinophils_sigenes_down <- rownames(tc_eosinophils_clinic_sig_sva[["deseq"]][["downs"]][["cure"]])
tc_monocytes_sigenes_up <- rownames(tc_monocytes_sig_sva[["deseq"]][["ups"]][["cure"]])
tc_monocytes_sigenes_down <- rownames(tc_monocytes_sig_sva[["deseq"]][["downs"]][["cure"]])
tc_neutrophils_sigenes_up <- rownames(tc_neutrophils_sig_sva[["deseq"]][["ups"]][["cure"]])
tc_neutrophils_sigenes_down <- rownames(tc_neutrophils_sig_sva[["deseq"]][["downs"]][["cure"]])

tc_eosinophils_sigenes <- c(tc_eosinophils_sigenes_up,
                            tc_eosinophils_sigenes_down)
tc_monocytes_sigenes <- c(tc_monocytes_sigenes_up,
                          tc_monocytes_sigenes_down)
tc_neutrophils_sigenes <- c(tc_neutrophils_sigenes_up,
                            tc_neutrophils_sigenes_down)
```

## GSEA: gProfiler of genes deemed up/down when comparing Cali and Tumaco

I was curious to try to understand why the two clinics appear to be so
different vis a vis their PCA/DE; so I thought that gProfiler might
help boil those results down to something more digestible.

### GSEA: Compare clinics, all samples

Note that in the following block I used the function
simple_gprofiler(), but later in this document I will use
all_gprofiler().  The first invocation limits the search to a single
table, while the second will iterate over every result in a pairwise
differential expression analysis.

In this instance, we are looking at the vector of gene IDs deemed
significantly different between the two clinics in either the up or
down direction.

One other thing worth noting, the new version of gProfiler provides
some fun interactive plots.  I will add an example here.

```{r}
tc_eosinophil_gprofiler <- simple_gprofiler(
  tc_eosinophils_sigenes_up,
  excel = glue("{gsea_prefix}/eosinophil_clinics_tumaco_up-v{ver}.xlsx"))
tc_eosinophil_gprofiler

clinic_gp <- simple_gprofiler(
  clinic_sigenes,
  excel = glue("{gsea_prefix}/both_clinics_cali_up-v{ver}.xlsx"))
clinic_gp$pvalue_plots$REAC
clinic_gp$pvalue_plots$BP
clinic_gp$pvalue_plots$TF
clinic_gp$interactive_plots$GO
```

### GSEA: Compare clinics, Eosinophil samples

In the following block, I am looking at the gProfiler over represented
groups observed across clinics in only the Eosinophils.  First I do so
for all genes(up or down), followed by only the up and down groups.
Each of the following will include only the Reactome and GO:BP plots.
These searches did not have too many other hits, excepting the
transcription factor database.

```{r}
tc_eosinophils_gp <- simple_gprofiler(
  tc_eosinophils_sigenes,
  excel = glue("{gsea_prefix}/eosinophil_clinics-v{ver}.xlsx"))
tc_eosinophils_gp
tc_eosinophils_gp$pvalue_plots$REAC
tc_eosinophils_gp$pvalue_plots$BP

tc_eosinophils_up_gp <- simple_gprofiler(
  tc_eosinophils_sigenes_up,
  excel = glue("{gsea_prefix}/eosinophil_clinics_tumaco_up-v{ver}.xlsx"))
tc_eosinophils_up_gp
tc_eosinophils_up_gp$pvalue_plots$REAC

tc_eosinophils_down_gp <- simple_gprofiler(
  tc_eosinophils_sigenes_down,
  excel = glue("{gsea_prefix}/eosinophil_clinics_cali_up-v{ver}.xlsx"))
tc_eosinophils_down_gp
tc_eosinophils_down_gp$pvalue_plots$REAC
```

### GSEA: Compare clinics, Monocyte samples

In the following block I repeated the above query, but this time
looking at the monocyte samples.

```{r}
tc_monocytes_up_gp <- simple_gprofiler(
  tc_monocytes_sigenes,
  excel = glue("{gsea_prefix}/monocyte_clinics-v{ver}.xlsx"))
tc_monocytes_up_gp
tc_monocytes_up_gp$pvalue_plots$REAC
tc_monocytes_up_gp$pvalue_plots$BP

tc_monocytes_down_gp <- simple_gprofiler(
  tc_monocytes_sigenes_down,
  excel = glue("{gsea_prefix}/monocyte_clinics_cali_up-v{ver}.xlsx"))
tc_monocytes_down_gp$pvalue_plots$REAC
tc_monocytes_down_gp$pvalue_plots$BP
```

#### GSEA: Compare clinics, Neutrophil samples

Ibid.  This time looking at the Neutrophils.  Thus the first two
images should be a superset of the second and third pairs of images;
assuming that the genes in the up/down list do not cause the groups to
no longer be significant.  Interestingly, the reactome search did not
return any hits for the increased search.

```{r}
tc_neutrophils_gp <- simple_gprofiler(
  tc_neutrophils_sigenes,
  excel = glue("{gsea_prefix}/neutrophil_clinics-v{ver}.xlsx"))
## tc_neutrophils_gp$pvalue_plots$REAC ## no hits
tc_neutrophils_gp$pvalue_plots$BP
tc_neutrophils_gp$pvalue_plots$TF

tc_neutrophils_up_gp <- simple_gprofiler(
  tc_neutrophils_sigenes_up,
  excel = glue("{gsea_prefix}/neutrophil_clinics_tumaco_up-v{ver}.xlsx"))
## tc_neutrophils_up_gp$pvalue_plots$REAC ## No hits
tc_neutrophils_up_gp$pvalue_plots$BP

tc_neutrophils_down_gp <- simple_gprofiler(
  tc_neutrophils_sigenes_down,
  excel = glue("{gsea_prefix}/neutrophil_clinics_cali_up-v{ver}.xlsx"))
tc_neutrophils_down_gp$pvalue_plots$REAC
tc_neutrophils_down_gp$pvalue_plots$BP
```

# Compare DE: How similar are Tumaco C/F vs. Cali C/F

The following expands the cross-clinic query above to also test the
neutrophils.  Once again, I think it will pretty strongly support the
hypothesis that the two clinics are not compatible.

We are concerned that the clinic-based batch effect may make our
results essentially useless.  One way to test this concern is to
compare the set of genes observed different between the Cali Cure/Fail
vs. the Tumaco Cure/Fail.

```{r}
cali_table_nobatch <- tc_neutrophils_table_nobatch[["data"]][["cali"]]
tumaco_table_nobatch <- tc_neutrophils_table_nobatch[["data"]][["tumaco"]]

cali_merged_nobatch <- merge(cali_table_nobatch, tumaco_table_nobatch, by="row.names")
cor.test(cali_merged_nobatch[, "deseq_logfc.x"], cali_merged_nobatch[, "deseq_logfc.y"])
cali_aucc_nobatch <- calculate_aucc(cali_table_nobatch, tumaco_table_nobatch, px = "deseq_adjp",
                                    py = "deseq_adjp", lx = "deseq_logfc", ly = "deseq_logfc")
cali_aucc_nobatch$plot
```

# Tumaco and Cali, cure vs. fail

In all of the above, we are looking to understand the differences between the two location.
Let us now step back and perform the original question: fail/cure without regard to location.

I performed this query with a few different parameters, notably with(out)
sva and again using each cell type, including biopsies. The main
reasion I am keeping these comparisons is in the relatively weak hope
that there will be sufficient signal in the full dataset that it might
be able to overcome the apparently ridiculous batch effect from the
two clinics.

## All cell types together, with(out) SVA

```{r}
table(pData(tc_valid)[["condition"]])
tc_all_cf_de_sva <- all_pairwise(tc_valid, filter = TRUE, methods = methods,
                                 parallel = parallel, model_batch = "svaseq")
tc_all_cf_table_sva <- combine_de_tables(
  tc_all_cf_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/tc_valid_cf_table_sva-v{ver}.xlsx"))
tc_all_cf_sig_sva <- extract_significant_genes(
  tc_all_cf_table_sva,
  excel = glue("{cf_prefix}/All_Samples/tc_valid_cf_sig_sva-v{ver}.xlsx"))

tc_all_cf_de_batch <- all_pairwise(tc_valid, filter = TRUE, methods = methods,
                                   parallel = parallel, model_batch = TRUE)
tc_all_cf_table_batch <- combine_de_tables(
  tc_all_cf_de_batch, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/tc_valid_cf_table_batch-v{ver}.xlsx"))
tc_all_cf_sig_batch <- extract_significant_genes(
  tc_all_cf_table_batch,
  excel = glue("{cf_prefix}/All_Samples/tc_valid_cf_sig_batch-v{ver}.xlsx"))
```

## All cell types except biopsies

I am not sure if this is the best choice, but I call the set of all
samples excluding biopsies 'clinical'.

```{r}
table(pData(tc_clinical_nobiop)[["condition"]])
tc_clinical_cf_de_sva <- all_pairwise(tc_clinical_nobiop, filter = TRUE,
                                      parallel = parallel, model_batch = "svaseq",
                                      methods = methods)
tc_clinical_cf_de_sva

tc_clinical_cf_table_sva <- combine_de_tables(
  tc_clinical_cf_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Clinical_Samples/tc_clinical_cf_table_sva-v{ver}.xlsx"))
tc_clinical_cf_table_sva

tc_clinical_cf_sig_sva <- extract_significant_genes(
  tc_clinical_cf_table_sva, according_to = "deseq",
  excel = glue("{cf_prefix}/Clinical_Samples/tc_clinical_cf_sig_sva-v{ver}.xlsx"))
tc_clinical_cf_sig_sva

tc_clinical_cf_de_batch <- all_pairwise(tc_clinical_nobiop, filter = TRUE,
                                        parallel = parallel, model_batch = TRUE,
                                        methods = methods)
tc_clinical_cf_de_batch

tc_clinical_cf_table_batch <- combine_de_tables(
  tc_clinical_cf_de_batch, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Clinical_Samples/tc_clinical_cf_table_batch-v{ver}.xlsx"))
tc_clinical_cf_table_batch

tc_clinical_cf_sig_batch <- extract_significant_genes(
  tc_clinical_cf_table_batch, according_to = "deseq",
  excel = glue("{cf_prefix}/Clinical_Samples/tc_clinical_cf_sig_batch-v{ver}.xlsx"))
tc_clinical_cf_sig_batch
```

### A portion of Supplemental Figure 11.

```{r}
num_color <- color_choices[["cf"]][["cure"]]
den_color <- color_choices[["cf"]][["failure"]]
tc_clinical_cf_table <- tc_clinical_cf_table_sva[["data"]][["outcome"]]
tc_clinical_cf_volcano_top10 <- plot_volcano_condition_de(
  tc_clinical_cf_table, "outcome", label = 10,
  fc_col = "deseq_logfc", p_col = "deseq_adjp", line_position = NULL,
  color_high = num_color, color_low = den_color, label_size = 6)
pp(file = "figures/s11c_tc_clinical_cf_volcano_labeled_top10.svg")
tc_clinical_cf_volcano_top10[["plot"]]
dev.off()
tc_clinical_cf_volcano_top10[["plot"]]
```

## Biopsies, with(out) SVA

In the following block, we repeat the same question, but using only
the biopsy samples from both clinics.

```{r}
tc_biopsies_cf <- set_expt_conditions(tc_biopsies, fact = "finaloutcome")
tc_biopsies_cf_de_sva <- all_pairwise(tc_biopsies_cf, filter = TRUE, methods = methods,
                                      parallel = parallel, model_batch = "svaseq")
tc_biopsies_cf_table_sva <- combine_de_tables(
  tc_biopsies_cf_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Biopsies/tc_biopsies_cf_table_sva-v{ver}.xlsx"))
tc_biopsies_cf_sig_sva <- extract_significant_genes(
  tc_biopsies_cf_table_sva,
  excel = glue("{cf_prefix}/All_Samples/tc_biopsies_cf_sig_sva-v{ver}.xlsx"))

tc_biopsies_cf_de_batch <- all_pairwise(tc_biopsies_cf, filter = TRUE, methods = methods,
                                        parallel = parallel, model_batch = TRUE)
tc_biopsies_cf_table_batch <- combine_de_tables(
  tc_biopsies_cf_de_batch, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/tc_biopsies_cf_table_batch-v{ver}.xlsx"))
tc_biopsies_cf_sig_batch <- extract_significant_genes(
  tc_biopsies_cf_table_batch,
  excel = glue("{cf_prefix}/All_Samples/tc_biopsies_cf_sig_batch-v{ver}.xlsx"))
```

## Eosinophils, with(out) SVA

In the following block, we repeat the same question, but using only
the Eosinophil samples from both clinics.

```{r}
tc_eosinophils_cf <- set_expt_conditions(tc_eosinophils, fact = "finaloutcome")
tc_eosinophils_cf_de_sva <- all_pairwise(tc_eosinophils_cf, filter = TRUE, methods = methods,
                                         parallel = parallel, model_batch = "svaseq")
tc_eosinophils_cf_table_sva <- combine_de_tables(
  tc_eosinophils_cf_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Eosinophils/tc_eosinophils_cf_table_sva-v{ver}.xlsx"))
tc_eosinophils_cf_sig_sva <- extract_significant_genes(
  tc_eosinophils_cf_table_sva,
  excel = glue("{cf_prefix}/All_Samples/tc_eosinophils_cf_sig_sva-v{ver}.xlsx"))

tc_eosinophils_cf_de_batch <- all_pairwise(tc_eosinophils_cf, filter = TRUE,
                                           parallel = parallel, model_batch = TRUE,
                                           methods = methods)
tc_eosinophils_cf_table_batch <- combine_de_tables(
  tc_eosinophils_cf_de_batch, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/tc_eosinophils_cf_table_batch-v{ver}.xlsx"))
tc_eosinophils_cf_sig_batch <- extract_significant_genes(
  tc_eosinophils_cf_table_batch,
  excel = glue("{cf_prefix}/All_Samples/tc_eosinophils_cf_sig_batch-v{ver}.xlsx"))
```

## Monocytes, with(out) SVA

Repeat yet again, this time with the monocyte samples.  The idea is to
see if there is a cell type which is particularly good (or bad) at
discriminating the two clinics.

```{r}
tc_monocytes_cf <- set_expt_conditions(tc_monocytes, fact = "finaloutcome")
tc_monocytes_cf_de_sva <- all_pairwise(tc_monocytes_cf, filter = TRUE, methods = methods,
                                       parallel = parallel, model_batch = "svaseq")
tc_monocytes_cf_table_sva <- combine_de_tables(
  tc_monocytes_cf_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Monocytes/tc_monocytes_cf_table_sva-v{ver}.xlsx"))
tc_monocytes_cf_sig_sva <- extract_significant_genes(
  tc_monocytes_cf_table_sva,
  excel = glue("{cf_prefix}/All_Samples/tc_monocytes_cf_sig_sva-v{ver}.xlsx"))

tc_monocytes_cf_de_batch <- all_pairwise(tc_monocytes_cf, filter = TRUE, methods = methods,
                                         parallel = parallel, model_batch = TRUE)
tc_monocytes_cf_table_batch <- combine_de_tables(
  tc_monocytes_cf_de_batch, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/tc_monocytes_cf_table_batch-v{ver}.xlsx"))
tc_monocytes_cf_sig_batch <- extract_significant_genes(
  tc_monocytes_cf_table_batch,
  excel = glue("{cf_prefix}/All_Samples/tc_monocytes_cf_sig_batch-v{ver}.xlsx"))
```

## Neutrophils, with(out) SVA

Last try, this time using the Neutrophil samples.

```{r}
tc_neutrophils_cf <- set_expt_conditions(tc_neutrophils, fact = "finaloutcome")
tc_neutrophils_cf_de_sva <- all_pairwise(tc_neutrophils_cf, parallel = parallel,
                                         filter = TRUE, model_batch = "svaseq",
                                         methods = methods)
tc_neutrophils_cf_table_sva <- combine_de_tables(
  tc_neutrophils_cf_de_sva, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/Neutrophils/tc_neutrophils_cf_table_sva-v{ver}.xlsx"))
tc_neutrophils_cf_sig_sva <- extract_significant_genes(
  tc_neutrophils_cf_table_sva,
  excel = glue("{cf_prefix}/All_Samples/tc_neutrophils_cf_sig_sva-v{ver}.xlsx"))

tc_neutrophils_cf_de_batch <- all_pairwise(tc_neutrophils_cf, filter = TRUE,
                                           parallel = parallel, model_batch = TRUE,
                                           methods = methods)
tc_neutrophils_cf_table_batch <- combine_de_tables(
  tc_neutrophils_cf_de_batch, keepers = t_cf_contrast,
  excel = glue("{cf_prefix}/All_Samples/tc_neutrophils_cf_table_batch-v{ver}.xlsx"))
tc_neutrophils_cf_sig_batch <- extract_significant_genes(
  tc_neutrophils_cf_table_batch,
  excel = glue("{cf_prefix}/All_Samples/tc_neutrophils_cf_sig_batch-v{ver}.xlsx"))
```

# Visit comparisons

Later in this document I do a bunch of visit/cf comparisons.  In this
block I want to explicitly only compare v1 to other visits.  This is
something I did quite a lot in the 2019 datasets, but never actually
moved to this document.

```{r}
v1_vs_later <- all_pairwise(tc_v1vs, model_batch = "svaseq", methods = methods,
                            parallel = parallel, filter = TRUE)
v1_vs_later_table <- combine_de_tables(
  v1_vs_later, keepers = visit_v1later,
  excel = glue("{visit_prefix}/v1_vs_later_tables-v{ver}.xlsx"))
v1_vs_later_sig <- extract_significant_genes(
  v1_vs_later_table,
  excel = glue("{visit_prefix}/v1_vs_later_sig-v{ver}.xlsx"))
```

#### GSEA: V1 vs other visits.

```{r}
v1later_gp <- all_gprofiler(v1_vs_later_sig)
v1later_gp[[1]]$pvalue_plots$REAC
v1later_gp[[2]]$pvalue_plots$REAC
```

# Sex comparison

```{r}
tc_sex_de <- all_pairwise(tc_sex, model_batch = "svaseq", methods = methods,
                          parallel = parallel, filter = TRUE)
tc_sex_table <- combine_de_tables(
  tc_sex_de, excel = glue("{sex_prefix}/tc_sex_table-v{ver}.xlsx"))
tc_sex_sig <- extract_significant_genes(
  tc_sex_table, excel = glue("{sex_prefix}/tc_sex_sig-v{ver}.xlsx"))
tc_sex_gp <- all_gprofiler(tc_sex_sig)
```

```{r}
tc_sex_cure <- subset_expt(tc_sex, subset = "finaloutcome=='cure'")
tc_sex_cure_de <- all_pairwise(tc_sex_cure, model_batch = "svaseq",
                               parallel = parallel, filter = TRUE,
                               methods = methods)
tc_sex_cure_de
tc_sex_cure_table <- combine_de_tables(
  tc_sex_cure_de, excel = glue("{sex_prefix}/tc_sex_cure_table-v{ver}.xlsx"))
tc_sex_cure_table
tc_sex_cure_sig <- extract_significant_genes(
  tc_sex_cure_table, excel = glue("{sex_prefix}/tc_sex_cure_sig-v{ver}.xlsx"))
tc_sex_cure_sig
```

#### GSEA: Sex comparisons both clinics

```{r}
tc_sex_cure_gp <- all_gprofiler(tc_sex_cure_sig)
tc_sex_cure_gp
tc_sex_cure_gp[[1]][["pvalue_plots"]][["BP"]]
tc_sex_cure_gp[[2]][["pvalue_plots"]][["BP"]]
```

# Ethnicity comparisons

```{r}
tc_ethnicity_de <- all_pairwise(tc_etnia_expt, model_batch = "svaseq",
                                parallel = parallel, filter = TRUE,
                                methods = methods)
tc_ethnicity_de
tc_ethnicity_table <- combine_de_tables(
  tc_ethnicity_de, keepers = ethnicity_contrasts,
  excel = glue("{eth_prefix}/tc_ethnicity_table-v{ver}.xlsx"))
tc_ethnicity_table
tc_ethnicity_table[["plots"]][["mestizo_indigenous"]][["deseq_ma_plots"]]
tc_ethnicity_table[["plots"]][["mestizo_afrocol"]][["deseq_ma_plots"]]
tc_ethnicity_table[["plots"]][["indigenous_afrocol"]][["deseq_ma_plots"]]

tc_ethnicity_sig <- extract_significant_genes(
  tc_ethnicity_table, excel = glue("{eth_prefix}/tc_ethnicity_sig-v{ver}.xlsx"))

ethnicity_cure <- subset_expt(tc_etnia_expt, subset = "finaloutcome=='cure'")
ethnicity_cure_de <- all_pairwise(ethnicity_cure, model_batch = "svaseq",
                                  parallel = parallel, filter = TRUE,
                                  methods = methods)
ethnicity_cure_table <- combine_de_tables(
  ethnicity_cure_de, keepers = ethnicity_contrasts,
  excel = glue("{eth_prefix}/ethnicity_cure_table-v{ver}.xlsx"))
ethnicity_cure_table
ethnicity_cure_table[["plots"]][["mestizo_indigenous"]][["deseq_ma_plots"]]
ethnicity_cure_table[["plots"]][["mestizo_afrocol"]][["deseq_ma_plots"]]
ethnicity_cure_table[["plots"]][["indigenous_afrocol"]][["deseq_ma_plots"]]
ethnicity_cure_sig <- extract_significant_genes(
  ethnicity_cure_table, excel = glue("{eth_prefix}/ethnicity_cure_sig-v{ver}.xlsx"))
ethnicity_cure_sig
```

#### GSEA: Ethnicity differences

Performed once with both clinics and again with only Tumaco.

```{r}
tc_ethnicity_gp <- all_gprofiler(tc_ethnicity_sig)
```

```{r}
pander::pander(sessionInfo())
message("This is hpgltools commit: ", get_git_commit())
message("Saving to ", savefile)
# tmp <- sm(saveme(filename = savefile))
```

```{r loadme_after, eval=FALSE}
tmp <- loadme(filename = savefile)
```
